<h1 id="january-26-2011">January 26, 2011</h1>
<ul>
<li>All data updated to the <strong>latest Firehose run</strong> (December 30).</li>
<li><strong>New RNA-Seq expression data</strong> for many tumor types.</li>
<li>Updated LUSC MAF file.</li>
</ul>
<h1 id="january-5-2011">January 5, 2011</h1>
<ul>
<li>Now <strong>19 tumor types</strong>, all data updated to the <strong>latest Firehose run</strong> (November 28).</li>
<li>Support for <strong>microRNA</strong> expression and copy-number (including multiple loci).</li>
<li>New <strong>gene symbol validation</strong> service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</li>
<li>Improved <strong>links to IGV</strong> for visualization of DNA copy-number changes.</li>
<li>Background information from the <a href="http://www.sanger.ac.uk/genetics/CGP/Census/">Sanger Cancer Gene Census</a>.</li>
<li>Two <strong>new <a href="tutorial.jsp">Tutorials</a></strong> to get you quickly started in using the portal.</li>
</ul>
<h1 id="december-5-2011">December 5, 2011</h1>
<p>New data types and updated data:</p>
<ul>
<li>Added support for RNA-Seq expression data (COADREAD, KIRC, LUSC).</li>
<li>Added support for log2 copy-number data (available on the Plots tab).</li>
<li>Updated LAML and LUSC MAF files.</li>
<li>Updated LUSC case lists (reflecting the manuscript sample set).</li>
</ul>
<h1 id="november-15-2011">November 15, 2011</h1>
<p>Another major update, with the following new features and data:</p>
<ul>
<li>In collaboration with Bilkent University, we have added a new <em>network</em> tab to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
<li>New and improved mutation details, with sorting and filtering capabilities.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Links to the <a href="http://www.broadinstitute.org/igv/">Integrative Genomics Viewer (IGV)</a> to view copy-number alterations.</li>
<li>All data updated to the October 2011 TCGA Broad Firehose run.</li>
<li>RPPA data is now available for 2154 samples in six tumor types.</li>
<li>Many new and updated MAF files - now 9 tumor types with mutation data.</li>
</ul>
<p><img alt="GBM Network" src="images/previews/ova_network.png" /></p>
<h1 id="october-18-2011">October 18, 2011</h1>
<p>This release adds the following new data:</p>
<ul>
<li>Based on the September Firehose run with new tumor types (Bladder, Liver, Low grade glioma).</li>
<li>Latest MAF files from the COADREAD and LUSC analysis working groups.</li>
<li>Mutation Assessor links for all missense mutations.</li>
<li>RPPA data for kidney clear cell and endometrial cancer.</li>
<li>GBM case lists for transcriptional subtypes and CIMP.</li>
</ul>
<h1 id="september-12-2011">September 12, 2011</h1>
<p>This is a major update, which includes the following main changes:</p>
<ul>
<li>New tumor types and data: Now ~3500 samples from 12 tumor types, including mutation data for 8 tumor types.</li>
<li>RPPA data support - the Portal currently has data for ovarian and breast cancer.</li>
<li>Access control via <a href="http://openid.net/">OpenID</a>.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<h1 id="april-26-2011">April 26, 2011</h1>
<p>This is the initial release of the cBio TCGA Cancer Genomics Portal. The portal contains data from all completed and ongoing TCGA projects. All data is automatically imported from the output of the Broad Firehose.</p>
<p>This release includes data on 2531 samples from 11 tumor types.</p>