--
-- Copyright (c) 2015 - 2016 Memorial Sloan-Kettering Cancer Center.
--
-- This library is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF MERCHANTABILITY OR FITNESS
-- FOR A PARTICULAR PURPOSE. The software and documentation provided hereunder
-- is on an "as is" basis, and Memorial Sloan-Kettering Cancer Center has no
-- obligations to provide maintenance, support, updates, enhancements or
-- modifications. In no event shall Memorial Sloan-Kettering Cancer Center be
-- liable to any party for direct, indirect, special, incidental or
-- consequential damages, including lost profits, arising out of the use of this
-- software and its documentation, even if Memorial Sloan-Kettering Cancer
-- Center has been advised of the possibility of such damage.
--
-- This file is part of cBioPortal.
--
-- cBioPortal is free software: you can redistribute it and/or modify
-- it under the terms of the GNU Affero General Public License as
-- published by the Free Software Foundation, either version 3 of the
-- License.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU Affero General Public License for more details.
--
-- You should have received a copy of the GNU Affero General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
-- ----------------------------------------------------------------------------
-- A manually extracted subset of data for a small number of genes and samples from the BRCA
-- data set. This is intended to be used during unit testing, to validate the portal APIs.
-- In theory, it should be enough to run up a portal.
--
-- Prepared by Stuart Watt -- 13th May 2015

CREATE TABLE "type_of_cancer" (
  "TYPE_OF_CANCER_ID" VARCHAR(63) NOT NULL,
  "NAME" VARCHAR(255) NOT NULL,
  "CLINICAL_TRIAL_KEYWORDS" VARCHAR(1024) NOT NULL,
  "DEDICATED_COLOR" VARCHAR(31) NOT NULL,
  "SHORT_NAME" VARCHAR(127),
  "PARENT" VARCHAR(63),
  PRIMARY KEY  ("TYPE_OF_CANCER_ID")
);

CREATE TABLE "gene" (
  "ENTREZ_GENE_ID" INTEGER NOT NULL,
  "HUGO_GENE_SYMBOL" VARCHAR(255) NOT NULL,
  "GENETIC_ENTITY_ID" INTEGER, -- just a quick fix for GeneMapper.xml to work. TODO this part of the file should not exist...use cgds.sql instead. TODO 2 - add genetic_entity insert statements when making tests that involve this table
  "TYPE" VARCHAR(50),
  "CYTOBAND" VARCHAR(50),
  "LENGTH" INTEGER,
  PRIMARY KEY ("ENTREZ_GENE_ID")
);

CREATE TABLE "gene_alias" (
  "ENTREZ_GENE_ID" INTEGER DEFAULT NULL,
  "GENE_ALIAS" VARCHAR(255) DEFAULT NULL
);

CREATE INDEX "HUGO_GENE_SYMBOL" ON "gene"("HUGO_GENE_SYMBOL");

CREATE TABLE "cancer_study" (
  "CANCER_STUDY_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "CANCER_STUDY_IDENTIFIER" VARCHAR(255),
  "TYPE_OF_CANCER_ID" VARCHAR(25) NOT NULL,
  "NAME" VARCHAR(255) NOT NULL,
  "SHORT_NAME" VARCHAR(64) NOT NULL,
  "DESCRIPTION" VARCHAR(1024) NOT NULL,
  "PUBLIC" BOOLEAN NOT NULL,
  "PMID" VARCHAR(20) DEFAULT NULL,
  "CITATION" VARCHAR(200) DEFAULT NULL,
  "GROUPS" VARCHAR(200) DEFAULT NULL,
  "STATUS" INTEGER DEFAULT NULL,
  "IMPORT_DATE" DATETIME DEFAULT NULL,
  UNIQUE ("CANCER_STUDY_IDENTIFIER"),
  FOREIGN KEY ("TYPE_OF_CANCER_ID") REFERENCES "type_of_cancer" ("TYPE_OF_CANCER_ID")
);

CREATE TABLE "cosmic_mutation" (
  "COSMIC_MUTATION_ID" varchar(30) NOT NULL,
  "CHR" varchar(5),
  "START_POSITION" bigint(20),
  "REFERENCE_ALLELE" varchar(255),
  "TUMOR_SEQ_ALLELE" varchar(255),
  "STRAND" varchar(2),
  "CODON_CHANGE" varchar(255),
  "ENTREZ_GENE_ID" int(255) NOT NULL,
  "PROTEIN_CHANGE" varchar(255) NOT NULL,
  "COUNT" int(11) NOT NULL,
  "KEYWORD" varchar(50) DEFAULT NULL,
  KEY ("KEYWORD"),
  PRIMARY KEY ("COSMIC_MUTATION_ID"),
  FOREIGN KEY ("ENTREZ_GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID")
);

CREATE TABLE "genetic_profile" (
  "GENETIC_PROFILE_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "STABLE_ID" VARCHAR(255) NOT NULL,
  "CANCER_STUDY_ID" INTEGER NOT NULL,
  "GENETIC_ALTERATION_TYPE" VARCHAR(255) NOT NULL,
  "DATATYPE" VARCHAR(255) NOT NULL,
  "NAME" VARCHAR(255) NOT NULL,
  "DESCRIPTION" CLOB,
  "SHOW_PROFILE_IN_ANALYSIS_TAB" INTEGER NOT NULL,
  UNIQUE ("STABLE_ID"),
  FOREIGN KEY ("CANCER_STUDY_ID") REFERENCES "cancer_study" ("CANCER_STUDY_ID") ON DELETE CASCADE
);

CREATE TABLE "genetic_profile_samples" (
  "ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "GENETIC_PROFILE_ID" INTEGER NOT NULL,
  "ORDERED_SAMPLE_LIST" CLOB NOT NULL,
  UNIQUE ("GENETIC_PROFILE_ID"),
  FOREIGN KEY ("GENETIC_PROFILE_ID") REFERENCES "genetic_profile" ("GENETIC_PROFILE_ID") ON DELETE CASCADE
);

CREATE TABLE "patient" (
  "INTERNAL_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "STABLE_ID" VARCHAR(50) NOT NULL,
  "CANCER_STUDY_ID" INTEGER NOT NULL,
  FOREIGN KEY ("CANCER_STUDY_ID") REFERENCES "cancer_study" ("CANCER_STUDY_ID") ON DELETE CASCADE
);

CREATE TABLE "sample" (
  "INTERNAL_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "STABLE_ID" VARCHAR(50) NOT NULL,
  "SAMPLE_TYPE" VARCHAR(255) NOT NULL,
  "PATIENT_ID" INTEGER NOT NULL,
  "TYPE_OF_CANCER_ID" VARCHAR(25) NOT NULL,
  FOREIGN KEY ("PATIENT_ID") REFERENCES "patient" ("INTERNAL_ID") ON DELETE CASCADE,
  FOREIGN KEY ("TYPE_OF_CANCER_ID") REFERENCES "type_of_cancer" ("TYPE_OF_CANCER_ID")
);

CREATE TABLE "mutation_event" (
  "MUTATION_EVENT_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "ENTREZ_GENE_ID" INTEGER NOT NULL,
  "CHR" VARCHAR(5),
  "START_POSITION" BIGINT,
  "END_POSITION" BIGINT,
  "REFERENCE_ALLELE" VARCHAR(255),
  "TUMOR_SEQ_ALLELE" VARCHAR(255),
  "PROTEIN_CHANGE" VARCHAR(255),
  "MUTATION_TYPE" VARCHAR(255),
  "FUNCTIONAL_IMPACT_SCORE" VARCHAR(50),
  "FIS_VALUE" FLOAT,
  "LINK_XVAR" VARCHAR(500),
  "LINK_PDB" VARCHAR(500),
  "LINK_MSA" VARCHAR(500),
  "NCBI_BUILD" VARCHAR(10),
  "STRAND" VARCHAR(2),
  "VARIANT_TYPE" VARCHAR(15),
  "DB_SNP_RS" VARCHAR(25),
  "DB_SNP_VAL_STATUS" VARCHAR(255),
  "ONCOTATOR_DBSNP_RS" VARCHAR(255),
  "ONCOTATOR_REFSEQ_MRNA_ID" VARCHAR(64),
  "ONCOTATOR_CODON_CHANGE" VARCHAR(255),
  "ONCOTATOR_UNIPROT_ENTRY_NAME" VARCHAR(64),
  "ONCOTATOR_UNIPROT_ACCESSION" VARCHAR(64),
  "ONCOTATOR_PROTEIN_POS_START" INTEGER,
  "ONCOTATOR_PROTEIN_POS_END" INTEGER,
  "CANONICAL_TRANSCRIPT" INTEGER,
  "KEYWORD" VARCHAR(50) DEFAULT NULL,
  UNIQUE ("CHR", "START_POSITION", "END_POSITION", "TUMOR_SEQ_ALLELE", "ENTREZ_GENE_ID", "PROTEIN_CHANGE", "MUTATION_TYPE"),
  FOREIGN KEY ("ENTREZ_GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID")
);

CREATE INDEX "KEYWORD" ON "mutation_event"("KEYWORD");

CREATE TABLE "mutation" (
  "ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "MUTATION_EVENT_ID" INTEGER NOT NULL,
  "GENETIC_PROFILE_ID" INTEGER NOT NULL,
  "SAMPLE_ID" INTEGER NOT NULL,
  "ENTREZ_GENE_ID" INTEGER NOT NULL,
  "CENTER" VARCHAR(100),
  "SEQUENCER" VARCHAR(255),
  "MUTATION_STATUS" VARCHAR(25),
  "VALIDATION_STATUS" VARCHAR(25),
  "TUMOR_SEQ_ALLELE1" VARCHAR(255),
  "TUMOR_SEQ_ALLELE2" VARCHAR(255),
  "MATCHED_NORM_SAMPLE_BARCODE" VARCHAR(255),
  "MATCH_NORM_SEQ_ALLELE1" VARCHAR(255),
  "MATCH_NORM_SEQ_ALLELE2" VARCHAR(255),
  "TUMOR_VALIDATION_ALLELE1" VARCHAR(255),
  "TUMOR_VALIDATION_ALLELE2" VARCHAR(255),
  "MATCH_NORM_VALIDATION_ALLELE1" VARCHAR(255),
  "MATCH_NORM_VALIDATION_ALLELE2" VARCHAR(255),
  "VERIFICATION_STATUS" VARCHAR(10),
  "SEQUENCING_PHASE" VARCHAR(100),
  "SEQUENCE_SOURCE" VARCHAR(255) NOT NULL,
  "VALIDATION_METHOD" VARCHAR(255),
  "SCORE" VARCHAR(100),
  "BAM_FILE" VARCHAR(255),
  "TUMOR_ALT_COUNT" INTEGER,
  "TUMOR_REF_COUNT" INTEGER,
  "NORMAL_ALT_COUNT" INTEGER,
  "NORMAL_REF_COUNT" INTEGER,
  "AMINO_ACID_CHANGE" VARCHAR(255),
  FOREIGN KEY ("MUTATION_EVENT_ID") REFERENCES "mutation_event" ("MUTATION_EVENT_ID"),
  FOREIGN KEY ("ENTREZ_GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID"),
  FOREIGN KEY ("GENETIC_PROFILE_ID") REFERENCES "genetic_profile" ("GENETIC_PROFILE_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE
);

CREATE INDEX "GENETIC_PROFILE_ID_GENE" ON "mutation"("GENETIC_PROFILE_ID","ENTREZ_GENE_ID");
CREATE INDEX "GENETIC_PROFILE_ID_SAMPLE" ON "mutation"("GENETIC_PROFILE_ID","SAMPLE_ID");
CREATE INDEX "ENTREZ_GENE_ID" ON "mutation"("ENTREZ_GENE_ID");
CREATE INDEX "SAMPLE_ID" ON "mutation"("SAMPLE_ID");

CREATE TABLE "mutation_count" (
  "ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "GENETIC_PROFILE_ID" INTEGER NOT NULL,
  "SAMPLE_ID" INTEGER NOT NULL,
  "MUTATION_COUNT" INTEGER NOT NULL,
  FOREIGN KEY ("GENETIC_PROFILE_ID") REFERENCES "genetic_profile" ("GENETIC_PROFILE_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE
);

CREATE INDEX "MUTATION_COUNT_GENETIC_PROFILE_ID" ON "mutation_count"("GENETIC_PROFILE_ID","SAMPLE_ID");

CREATE TABLE "gene_panel" (
    "INTERNAL_ID" int(11) NOT NULL,
    "STABLE_ID" varchar(255) NOT NULL,
    "DESCRIPTION" mediumtext,
    PRIMARY KEY ("INTERNAL_ID"),
    UNIQUE ("STABLE_ID")
);

CREATE TABLE "gene_panel_list" (
    "INTERNAL_ID" int(11) NOT NULL,
    "GENE_ID" INTEGER NOT NULL,
    PRIMARY KEY ("INTERNAL_ID", "GENE_ID"),
    FOREIGN KEY ("INTERNAL_ID") REFERENCES "gene_panel" ("INTERNAL_ID") ON DELETE CASCADE,
    FOREIGN KEY ("GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID") ON DELETE CASCADE
);

CREATE TABLE "sample_profile" (
  "ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "SAMPLE_ID" INTEGER NOT NULL,
  "GENETIC_PROFILE_ID" INTEGER NOT NULL,
  "PANEL_ID" INTEGER DEFAULT NULL,
  FOREIGN KEY ("GENETIC_PROFILE_ID") REFERENCES "genetic_profile" ("GENETIC_PROFILE_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE,
  FOREIGN KEY ("PANEL_ID") REFERENCES "gene_panel" ("INTERNAL_ID") ON DELETE RESTRICT);

CREATE TABLE "sample_list" (
  "LIST_ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "STABLE_ID" VARCHAR(255) NOT NULL,
  "CATEGORY" VARCHAR(255) NOT NULL,
  "CANCER_STUDY_ID" INTEGER NOT NULL,
  "NAME" VARCHAR(255) NOT NULL,
  "DESCRIPTION" CLOB,
  UNIQUE ("STABLE_ID"),
  FOREIGN KEY ("CANCER_STUDY_ID") REFERENCES "cancer_study" ("CANCER_STUDY_ID") ON DELETE CASCADE
);

CREATE INDEX "STABLE_ID" ON "sample_list"("STABLE_ID");

CREATE TABLE "sample_list_list" (
  "ID" INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
  "LIST_ID" INTEGER NOT NULL,
  "SAMPLE_ID" INTEGER NOT NULL,
  FOREIGN KEY ("LIST_ID") REFERENCES "sample_list" ("LIST_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE
);

CREATE INDEX "SAMPLE_LIST_LIST" ON "sample_list_list"("LIST_ID","SAMPLE_ID");

CREATE TABLE `copy_number_seg` (
  `SEG_ID` int(255) NOT NULL AUTO_INCREMENT,
  `CANCER_STUDY_ID` int(11) NOT NULL,
  `SAMPLE_ID` int(11) NOT NULL,
  `CHR` varchar(5) NOT NULL,
  `START` int(11) NOT NULL,
  `END` int(11) NOT NULL,
  `NUM_PROBES` int(11) NOT NULL,
  `SEGMENT_MEAN` double NOT NULL,
  PRIMARY KEY (`SEG_ID`),
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE,
  FOREIGN KEY ("CANCER_STUDY_ID") REFERENCES "cancer_study" ("CANCER_STUDY_ID") ON DELETE CASCADE
);

CREATE TABLE `clinical_patient` (
  `INTERNAL_ID` int(11) NOT NULL,
  `ATTR_ID` varchar(255) NOT NULL,
  `ATTR_VALUE` varchar(255) NOT NULL,
  PRIMARY KEY (`INTERNAL_ID`, `ATTR_ID`),
  FOREIGN KEY (`INTERNAL_ID`) REFERENCES `patient` (`INTERNAL_ID`) ON DELETE CASCADE
);

CREATE TABLE `clinical_sample` (
  `INTERNAL_ID` int(11) NOT NULL,
  `ATTR_ID` varchar(255) NOT NULL,
  `ATTR_VALUE` varchar(255) NOT NULL,
  PRIMARY KEY (`INTERNAL_ID`,`ATTR_ID`),
  FOREIGN KEY (`INTERNAL_ID`) REFERENCES `sample` (`INTERNAL_ID`) ON DELETE CASCADE
);

CREATE TABLE `clinical_attribute_meta` (
  `ATTR_ID` varchar(255) NOT NULL,
  `DISPLAY_NAME` varchar(255) NOT NULL,
  `DESCRIPTION` varchar(2048) NOT NULL,
  `DATATYPE` varchar(255) NOT NULL,
  `PATIENT_ATTRIBUTE` BOOLEAN NOT NULL,
  `PRIORITY` varchar(255) NOT NULL,
  `CANCER_STUDY_ID` int(11) NOT NULL,
  PRIMARY KEY (`ATTR_ID`,`CANCER_STUDY_ID`),
  FOREIGN KEY (`CANCER_STUDY_ID`) REFERENCES `cancer_study` (`CANCER_STUDY_ID`) ON DELETE CASCADE
);

CREATE TABLE "structural_variant" (
  "SAMPLE_ID" INTEGER(11) NOT NULL,
  "INTERNAL_ID" INTEGER(11) NOT NULL auto_increment,
  "BREAKPOINT_TYPE" VARCHAR(25),
  "ANNOTATION" VARCHAR(255),
  "COMMENTS" VARCHAR(2048),
  "CONFIDENCE_CLASS" VARCHAR(25),
  "CONNECTION_TYPE" VARCHAR(25),
  "EVENT_INFO" VARCHAR(255),
  "MAPQ" INTEGER(11),
  "NORMAL_READ_COUNT" INTEGER(11),
  "NORMAL_VARIANT_COUNT" INTEGER(11),
  "PAIRED_END_READ_SUPPORT" VARCHAR(255),
  "SITE1_CHROM" VARCHAR(25),
  "SITE1_DESC" VARCHAR(255),
  "SITE1_ENTREZ_GENE_ID" INTEGER(11),
  "SITE1_POS" INTEGER(11),
  "SITE2_CHROM" VARCHAR(25),
  "SITE2_DESC" VARCHAR(255),
  "SITE2_ENTREZ_GENE_ID" INTEGER(11),
  "SITE2_POS" INTEGER(11),
  "SPLIT_READ_SUPPORT" VARCHAR(255),
  "SV_CLASS_NAME" VARCHAR(25),
  "SV_DESC" VARCHAR(255),
  "SV_LENGTH" INTEGER(11),
  "TUMOR_READ_COUNT" INTEGER(11),
  "TUMOR_VARIANT_COUNT" INTEGER(11),
  "VARIANT_STATUS_NAME" VARCHAR(255),
  "GENETIC_PROFILE_ID" INTEGER(11) NOT NULL,
  PRIMARY KEY ("INTERNAL_ID"),
  FOREIGN KEY ("SAMPLE_ID") REFERENCES "sample" ("INTERNAL_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SITE1_ENTREZ_GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID") ON DELETE CASCADE,
  FOREIGN KEY ("SITE2_ENTREZ_GENE_ID") REFERENCES "gene" ("ENTREZ_GENE_ID") ON DELETE CASCADE,
  FOREIGN KEY ("GENETIC_PROFILE_ID") REFERENCES "genetic_profile" ("GENETIC_PROFILE_ID") ON DELETE CASCADE
);

INSERT INTO "type_of_cancer" ("TYPE_OF_CANCER_ID","NAME","CLINICAL_TRIAL_KEYWORDS","DEDICATED_COLOR","SHORT_NAME","PARENT")
  VALUES ('brca','Breast Invasive Carcinoma','breast,breast invasive','HotPink','Breast','tissue');
INSERT INTO "type_of_cancer" ("TYPE_OF_CANCER_ID","NAME","CLINICAL_TRIAL_KEYWORDS","DEDICATED_COLOR","SHORT_NAME","PARENT")
  VALUES ('acc','Adrenocortical Carcinoma','adrenocortical carcinoma','Purple','ACC','adrenal_gland');


INSERT INTO "cancer_study" ("CANCER_STUDY_ID", "CANCER_STUDY_IDENTIFIER", "TYPE_OF_CANCER_ID", "NAME", "SHORT_NAME", "DESCRIPTION", "PUBLIC", "PMID", "CITATION", "GROUPS", "STATUS", "IMPORT_DATE")
  VALUES (1,'study_tcga_pub','brca','Breast Invasive Carcinoma (TCGA, Nature 2012)','BRCA (TCGA)','<a href=\"http://cancergenome.nih.gov/\">The Cancer Genome Atlas (TCGA)</a> Breast Invasive Carcinoma project. 825 cases.<br><i>Nature 2012.</i> <a href=\"http://tcga-data.nci.nih.gov/tcga/\">Raw data via the TCGA Data Portal</a>.',1,'23000897','TCGA, Nature 2012','SU2C-PI3K;PUBLIC;GDAC', 0, '2011-12-18 13:17:17+00:00');
INSERT INTO "cancer_study" ("CANCER_STUDY_ID", "CANCER_STUDY_IDENTIFIER", "TYPE_OF_CANCER_ID", "NAME", "SHORT_NAME", "DESCRIPTION", "PUBLIC", "PMID", "CITATION", "GROUPS", "STATUS", "IMPORT_DATE")
  VALUES (2,'acc_tcga','acc','Adrenocortical Carcinoma (TCGA, Provisional)','ACC (TCGA)','TCGA Adrenocortical Carcinoma; raw data at the <A HREF="https://tcga-data.nci.nih.gov/">NCI</A>.',1,'23000897','TCGA, Nature 2012','SU2C-PI3K;PUBLIC;GDAC', 0, '2013-10-12 11:11:15+00:00');

INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (207,'AKT1','protein-coding','14q32.32',10838);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (208,'AKT2','protein-coding','19q13.1-q13.2',15035);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (10000,'AKT3','protein-coding','1q44',7499);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (369,'ARAF','protein-coding','Xp11.4-p11.2',3204);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (472,'ATM','protein-coding','11q22-q23',22317);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (673,'BRAF','protein-coding','7q34',4564);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (672,'BRCA1','protein-coding','17q21',8426);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (675,'BRCA2','protein-coding','13q12.3',11269);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (3265,'HRAS','protein-coding','11p15.5',1854);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (3845,'KRAS','protein-coding','12p12.1',7302);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (4893,'NRAS','protein-coding','1p13.2',4449);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (79501,'OR4F5','protein-coding','1p36.33',918);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (148398,'SAMD11','protein-coding','1p36.33',3205);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (26155,'NOC2L','protein-coding','1p36.33',5539);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (2064,'ERBB2','protein-coding','17q12',10321);
INSERT INTO "gene" ("ENTREZ_GENE_ID","HUGO_GENE_SYMBOL","TYPE","CYTOBAND","LENGTH") VALUES (2886,'GRB7','protein-coding','17q12',3597);


-- cosmic_mutation
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (3677745, 'D45A', 79501, 1, 'OR4F5 D45 missense');
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (426644, 'G145C', 79501, 1, 'OR4F5 G145 missense');
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (460103, 'P23P', 148398, 1, 'SAMD11 P23 silent');
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (4010395, 'S146S', 26155, 1, 'NOC2L S146 silent');
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (1290240, 'M1T', 26155, 1, 'NOC2L truncating');
INSERT INTO "cosmic_mutation" ("COSMIC_MUTATION_ID", "PROTEIN_CHANGE", "ENTREZ_GENE_ID", "COUNT", "KEYWORD") VALUES (4010425, 'Q197*', 26155, 1, 'NOC2L truncating');


INSERT INTO "gene_alias" ("ENTREZ_GENE_ID", "GENE_ALIAS") VALUES (207, 'AKT alias');
INSERT INTO "gene_alias" ("ENTREZ_GENE_ID", "GENE_ALIAS") VALUES (207, 'AKT alias2');
INSERT INTO "gene_alias" ("ENTREZ_GENE_ID", "GENE_ALIAS") VALUES (675, 'BRCA1 alias');

INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (2,'study_tcga_pub_gistic',1,'COPY_NUMBER_ALTERATION','DISCRETE','Putative copy-number alterations from GISTIC','Putative copy-number from GISTIC 2.0. Values: -2 = homozygous deletion; -1 = hemizygous deletion; 0 = neutral / no change; 1 = gain; 2 = high level amplification.',1);
INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (3,'study_tcga_pub_mrna',1,'MRNA_EXPRESSION','Z-SCORE','mRNA expression (microarray)','Expression levels (Agilent microarray).',0);
INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (4,'study_tcga_pub_log2CNA',1,'COPY_NUMBER_ALTERATION','LOG-VALUE','Log2 copy-number values','Log2 copy-number values for each gene (from Affymetrix SNP6).',0);
INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (5,'study_tcga_pub_methylation_hm27',1,'METHYLATION','CONTINUOUS','Methylation (HM27)','Methylation beta-values (HM27 platform). For genes with multiple methylation probes, the probe least correlated with expression is selected.',0);
INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (6,'study_tcga_pub_mutations',1,'MUTATION_EXTENDED','MAF','Mutations','Mutation data from whole exome sequencing.',1);
INSERT INTO "genetic_profile" ("GENETIC_PROFILE_ID", "STABLE_ID", "CANCER_STUDY_ID", "GENETIC_ALTERATION_TYPE", "DATATYPE", "NAME", "DESCRIPTION", "SHOW_PROFILE_IN_ANALYSIS_TAB") VALUES (7,'study_tcga_pub_sv',1,'STRUCTURAL_VARIANT','SV','Structural Variants','Structural Variants detected by Illumina HiSeq sequencing.',1);

-- genetic_profile_samples
INSERT INTO "genetic_profile_samples" ("GENETIC_PROFILE_ID", "ORDERED_SAMPLE_LIST") VALUES (2,'1,2,3,4,5,6,7,8,9,10,11,12,13,14,');
INSERT INTO "genetic_profile_samples" ("GENETIC_PROFILE_ID", "ORDERED_SAMPLE_LIST") VALUES (3,'2,3,6,8,9,10,12,13,');
INSERT INTO "genetic_profile_samples" ("GENETIC_PROFILE_ID", "ORDERED_SAMPLE_LIST") VALUES (4,'1,2,3,4,5,6,7,8,9,10,11,12,13,14,');
INSERT INTO "genetic_profile_samples" ("GENETIC_PROFILE_ID", "ORDERED_SAMPLE_LIST") VALUES (5,'2,');

-- patient
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (1,'TCGA-A1-A0SB',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (2,'TCGA-A1-A0SD',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (3,'TCGA-A1-A0SE',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (4,'TCGA-A1-A0SF',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (5,'TCGA-A1-A0SG',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (6,'TCGA-A1-A0SH',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (7,'TCGA-A1-A0SI',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (8,'TCGA-A1-A0SJ',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (9,'TCGA-A1-A0SK',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (10,'TCGA-A1-A0SM',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (11,'TCGA-A1-A0SN',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (12,'TCGA-A1-A0SO',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (13,'TCGA-A1-A0SP',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (14,'TCGA-A1-A0SQ',1);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (15,'TCGA-A1-B0SO',2);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (16,'TCGA-A1-B0SP',2);
INSERT INTO "patient" ("INTERNAL_ID", "STABLE_ID", "CANCER_STUDY_ID") VALUES (17,'TCGA-A1-B0SQ',2);

-- sample
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (1,'TCGA-A1-A0SB-01','Primary Solid Tumor',1,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (2,'TCGA-A1-A0SD-01','Primary Solid Tumor',2,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (3,'TCGA-A1-A0SE-01','Primary Solid Tumor',3,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (4,'TCGA-A1-A0SF-01','Primary Solid Tumor',4,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (5,'TCGA-A1-A0SG-01','Primary Solid Tumor',5,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (6,'TCGA-A1-A0SH-01','Primary Solid Tumor',6,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (7,'TCGA-A1-A0SI-01','Primary Solid Tumor',7,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (8,'TCGA-A1-A0SJ-01','Primary Solid Tumor',8,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (9,'TCGA-A1-A0SK-01','Primary Solid Tumor',9,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (10,'TCGA-A1-A0SM-01','Primary Solid Tumor',10,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (11,'TCGA-A1-A0SN-01','Primary Solid Tumor',11,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (12,'TCGA-A1-A0SO-01','Primary Solid Tumor',12,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (13,'TCGA-A1-A0SP-01','Primary Solid Tumor',13,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (14,'TCGA-A1-A0SQ-01','Primary Solid Tumor',14,'brca');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (15,'TCGA-A1-B0SO-01','Primary Solid Tumor',15,'acc');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (16,'TCGA-A1-B0SP-01','Primary Solid Tumor',16,'acc');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (17,'TCGA-A1-B0SQ-01','Primary Solid Tumor',17,'acc');
INSERT INTO "sample" ("INTERNAL_ID","STABLE_ID","SAMPLE_TYPE","PATIENT_ID","TYPE_OF_CANCER_ID") VALUES (18,'TCGA-A1-A0SB-02','Primary Solid Tumor',1,'brca');

-- mutation_event
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (2038,672,'17',41244748,41244748,'G','A','Q934*','Nonsense_Mutation','NA',0,'getma.org/?cm=var&var=hg19,17,41244748,G,A&fts=all','NA','NA','37','+','SNP','rs80357223','unknown','rs80357223','NM_007294','c.(2800-2802)CAG>TAG','BRCA1_HUMAN','P38398',934,934,1,'BRCA1 truncating');
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (22604,672,'17',41258504,41258504,'A','C','C61G','Missense_Mutation','H',4.355,'getma.org/?cm=var&var=hg19,17,41258504,A,C&fts=all','getma.org/pdb.php?prot=BRCA1_HUMAN&from=24&to=64&var=C61G','getma.org/?cm=msa&ty=f&p=BRCA1_HUMAN&rb=24&re=64&var=C61G','37','+','SNP','rs28897672','byCluster','rs28897672','NM_007294','c.(181-183)TGT>GGT','BRCA1_HUMAN','P38398',61,61,1,'BRCA1 C61 missense');
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (2039,672,'17',41276033,41276033,'C','T','C27_splice','Splice_Site','NA',1.4013e-45,'NA','NA','NA','37','+','SNP','rs80358010','byCluster','rs80358010','NM_007294','c.e2+1','NA','NA',-1,-1,1,'BRCA1 truncating');
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (2040,207,'17',41244748,41244748,'G','A','Q934*','Nonsense_Mutation','NA',0,'getma.org/?cm=var&var=hg19,17,41244748,G,A&fts=all','NA','NA','37','+','SNP','rs80357223','unknown','rs80357223','NM_007294','c.(2800-2802)CAG>TAG','BRCA1_HUMAN','P38398',934,934,1,'BRCA1 truncating');
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (2041,207,'17',41258504,41258504,'A','C','C61G','Missense_Mutation','H',4.355,'getma.org/?cm=var&var=hg19,17,41258504,A,C&fts=all','getma.org/pdb.php?prot=BRCA1_HUMAN&from=24&to=64&var=C61G','getma.org/?cm=msa&ty=f&p=BRCA1_HUMAN&rb=24&re=64&var=C61G','37','+','SNP','rs28897672','byCluster','rs28897672','NM_007294','c.(181-183)TGT>GGT','BRCA1_HUMAN','P38398',61,61,1,'BRCA1 C61 missense');
INSERT INTO "mutation_event" ("MUTATION_EVENT_ID","ENTREZ_GENE_ID","CHR","START_POSITION","END_POSITION","REFERENCE_ALLELE","TUMOR_SEQ_ALLELE","PROTEIN_CHANGE","MUTATION_TYPE","FUNCTIONAL_IMPACT_SCORE","FIS_VALUE","LINK_XVAR","LINK_PDB","LINK_MSA","NCBI_BUILD","STRAND","VARIANT_TYPE","DB_SNP_RS","DB_SNP_VAL_STATUS","ONCOTATOR_DBSNP_RS","ONCOTATOR_REFSEQ_MRNA_ID","ONCOTATOR_CODON_CHANGE","ONCOTATOR_UNIPROT_ENTRY_NAME","ONCOTATOR_UNIPROT_ACCESSION","ONCOTATOR_PROTEIN_POS_START","ONCOTATOR_PROTEIN_POS_END","CANONICAL_TRANSCRIPT","KEYWORD") VALUES (2042,208,'17',41276033,41276033,'C','T','C27_splice','Splice_Site','NA',1.4013e-45,'NA','NA','NA','37','+','SNP','rs80358010','byCluster','rs80358010','NM_007294','c.e2+1','NA','NA',-1,-1,1,'BRCA1 truncating');

-- mutation
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2038,6,6,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','G','A','TCGA-A1-A0SH-10A-03D-A099-09','G','A','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',1,0,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (22604,6,6,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','A','C','TCGA-A1-A0SH-10A-03D-A099-09','A','C','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2039,6,12,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','T','T','TCGA-A1-A0SO-10A-03D-A099-09','T','T','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2038,6,7,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','G','A','TCGA-A1-A0SH-10A-03D-A099-09','G','A','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2038,6,7,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','A','C','TCGA-A1-A0SH-10A-03D-A099-09','A','C','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2039,6,13,672,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','T','T','TCGA-A1-A0SO-10A-03D-A099-09','T','T','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2040,6,1,207,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','G','A','TCGA-A1-A0SH-10A-03D-A099-09','G','A','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2041,6,2,207,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','A','C','TCGA-A1-A0SH-10A-03D-A099-09','A','C','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',0,-1,-1,-1);
INSERT INTO "mutation" ("MUTATION_EVENT_ID","GENETIC_PROFILE_ID","SAMPLE_ID","ENTREZ_GENE_ID","CENTER","SEQUENCER","MUTATION_STATUS","VALIDATION_STATUS","TUMOR_SEQ_ALLELE1","TUMOR_SEQ_ALLELE2","MATCHED_NORM_SAMPLE_BARCODE","MATCH_NORM_SEQ_ALLELE1","MATCH_NORM_SEQ_ALLELE2","TUMOR_VALIDATION_ALLELE1","TUMOR_VALIDATION_ALLELE2","MATCH_NORM_VALIDATION_ALLELE1","MATCH_NORM_VALIDATION_ALLELE2","VERIFICATION_STATUS","SEQUENCING_PHASE","SEQUENCE_SOURCE","VALIDATION_METHOD","SCORE","BAM_FILE","TUMOR_ALT_COUNT","TUMOR_REF_COUNT","NORMAL_ALT_COUNT","NORMAL_REF_COUNT") VALUES (2042,6,3,208,'genome.wustl.edu','IlluminaGAIIx','Germline','Unknown','T','T','TCGA-A1-A0SO-10A-03D-A099-09','T','T','NA','NA','NA','NA','Unknown','Phase_IV','Capture','NA','1','dbGAP',-1,-1,-1,-1);

-- mutation_count
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,1,1);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,2,1);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,3,1);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,6,2);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,7,2);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,12,1);
INSERT INTO "mutation_count" ("GENETIC_PROFILE_ID","SAMPLE_ID","MUTATION_COUNT") VALUES (6,13,1);

INSERT INTO "gene_panel" ("INTERNAL_ID", "STABLE_ID", "DESCRIPTION") VALUES(1, 'TESTPANEL1', 'A test panel consisting of a few genes');
INSERT INTO "gene_panel" ("INTERNAL_ID", "STABLE_ID", "DESCRIPTION") VALUES(2, 'TESTPANEL2', 'Another test panel consisting of a few genes');

INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(1, 207);
INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(1, 369);
INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(1, 672);

INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(2, 207);
INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(2, 208);
INSERT INTO "gene_panel_list" ("INTERNAL_ID", "GENE_ID") VALUES(2, 4893);
-- sample_profile
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (1,2,1);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (1,3,1);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (1,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (2,2,2);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (2,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (2,4,2);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (2,5,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (2,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (3,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (3,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (3,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (3,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (4,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (4,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (5,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (5,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (6,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (6,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (6,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (6,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (7,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (7,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (8,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (8,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (8,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (8,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (9,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (9,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (9,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (9,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (10,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (10,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (10,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (10,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (11,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (11,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (12,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (12,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (12,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (12,6,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (13,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (13,3,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (13,4,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (14,2,NULL);
INSERT INTO "sample_profile" ("SAMPLE_ID","GENETIC_PROFILE_ID","PANEL_ID") VALUES (14,4,NULL);

-- sample_list
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (1,'study_tcga_pub_all','other',1,'All Tumors','All tumor samples');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (2,'study_tcga_pub_acgh','other',1,'Tumors aCGH','All tumors with aCGH data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (3,'study_tcga_pub_cnaseq','other',1,'Tumors with sequencing and aCGH data','All tumor samples that have CNA and sequencing data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (4,'study_tcga_pub_complete','other',1,'Complete samples (mutations, copy-number, expression)','Samples with complete data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (5,'study_tcga_pub_log2CNA','other',1,'Tumors log2 copy-number','All tumors with log2 copy-number data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (6,'study_tcga_pub_methylation_hm27','all_cases_with_mutation_data',1,'Tumors with methylation data','All samples with methylation (HM27) data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (7,'study_tcga_pub_mrna','other',1,'Tumors with mRNA data (Agilent microarray)','All samples with mRNA expression data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (8,'study_tcga_pub_sequenced','other',1,'Sequenced Tumors','All sequenced samples');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (9,'study_tcga_pub_cna','other',1,'Tumor Samples with CNA data','All tumors with CNA data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (10,'study_tcga_pub_rna_seq_v2_mrna','other',1,'Tumor Samples with mRNA data (RNA Seq V2)','All samples with mRNA expression data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (11,'study_tcga_pub_microrna','other',1,'Tumors with microRNA data (microRNA-Seq)','All samples with microRNA data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (12,'study_tcga_pub_rppa','other',1,'Tumor Samples with RPPA data','Tumors with reverse phase protein array (RPPA) data for about 200 antibodies');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (13,'study_tcga_pub_3way_complete','other',1,'All Complete Tumors','All tumor samples that have mRNA, CNA and sequencing data');
INSERT INTO "sample_list" ("LIST_ID", "STABLE_ID", "CATEGORY", "CANCER_STUDY_ID", "NAME", "DESCRIPTION") VALUES (14,'acc_tcga_all','other',2,'All Tumors','All tumor samples');

-- sample_list_list
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,1);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,4);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,5);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,7);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,11);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,13);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (1,14);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,1);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,4);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,5);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,7);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,11);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,13);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (2,14);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (3,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (4,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,1);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,4);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,5);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,7);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,11);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,13);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (5,14);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (6,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (7,13);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (8,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (9,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (10,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,2);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,3);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,6);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,8);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,9);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,10);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (13,12);
INSERT INTO "sample_list_list" ("LIST_ID","SAMPLE_ID") VALUES (14,15);

INSERT INTO "copy_number_seg" ("SEG_ID","CANCER_STUDY_ID","SAMPLE_ID","CHR","START","END","NUM_PROBES","SEGMENT_MEAN") VALUES (50236594, 1, 1, '1', 224556, 180057677, 291, 0.0519);
INSERT INTO "copy_number_seg" ("SEG_ID","CANCER_STUDY_ID","SAMPLE_ID","CHR","START","END","NUM_PROBES","SEGMENT_MEAN") VALUES (50236593, 1, 2,	'2', 1402650, 190262486, 207, 0.0265);
INSERT INTO "copy_number_seg" ("SEG_ID","CANCER_STUDY_ID","SAMPLE_ID","CHR","START","END","NUM_PROBES","SEGMENT_MEAN") VALUES (50236592, 1, 3,	'3', 1449872, 194238390, 341, 0.0347);

INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'RETROSPECTIVE_COLLECTION', 'NO');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'FORM_COMPLETION_DATE', '2013-12-5');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'OTHER_PATIENT_ID', '286CF147-B7F7-4A05-8E41-7FBD3717AD71');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (2, 'PROSPECTIVE_COLLECTION', 'YES');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'DFS_MONTHS', '5.72');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'DFS_STATUS', 'Recurred/Progressed');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'OS_MONTHS', 'Recurred/Progressed');
INSERT INTO "clinical_patient" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'OS_STATUS', 'LIVING');

INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'OTHER_SAMPLE_ID', '5C631CE8-F96A-4C35-A459-556FC4AB21E1');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'DAYS_TO_COLLECTION', '276');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (1, 'IS_FFPE', 'NO');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (2, 'OCT_EMBEDDED', 'false');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (2, 'PATHOLOGY_REPORT_FILE_NAME', 'TCGA-GC-A3BM.F3408556-9259-4700-B9A0-F41E516B420C.pdf');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (2, 'SAMPLE_TYPE', 'Primary Tumor');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'OTHER_SAMPLE_ID', '91E7F41C-17B3-4724-96EF-D3C207B964E1');
INSERT INTO "clinical_sample" ("INTERNAL_ID", "ATTR_ID", "ATTR_VALUE") VALUES (15, 'DAYS_TO_COLLECTION', '111');

INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('RETROSPECTIVE_COLLECTION', 'Tissue Retrospective Collection Indicator', 'Text indicator for the time frame of tissue procurement, indicating that the tissue was obtained and stored prior to the initiation of the project.', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('PROSPECTIVE_COLLECTION', 'Tissue Prospective Collection Indicator', 'Text indicator for the time frame of tissue procurement, indicating that the tissue was procured in parallel to the project.', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('FORM_COMPLETION_DATE', 'Form completion date', 'Form completion date', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OTHER_PATIENT_ID', 'Other Patient ID', 'Legacy DMP patient identifier (DMPnnnn)', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DFS_MONTHS', 'Disease Free (Months)', 'Disease free (months) since initial treatment.', 'NUMBER', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DFS_STATUS', 'Disease Free Status', 'Disease free status since initial treatment.', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OS_MONTHS', 'Overall Survival (Months)', 'Overall survival in months since initial diagonosis.', 'NUMBER', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OS_STATUS', 'Overall Survival Status', 'Overall patient survival status.', 'STRING', 1, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OTHER_SAMPLE_ID', 'Other Sample ID', 'Legacy DMP sample identifier (DMPnnnn)', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DAYS_TO_COLLECTION', 'Days to Sample Collection.', 'Days to sample collection.', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('IS_FFPE', 'Is FFPE', 'If the sample is from FFPE', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OCT_EMBEDDED', 'Oct embedded', 'Oct embedded', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('PATHOLOGY_REPORT_FILE_NAME', 'Pathology Report File Name', 'Pathology Report File Name', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('SAMPLE_TYPE', 'Sample Type', 'The type of sample (i.e., normal, primary, met, recurrence).', 'STRING', 0, '1', 1);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('RETROSPECTIVE_COLLECTION', 'Tissue Retrospective Collection Indicator', 'Text indicator for the time frame of tissue procurement, indicating that the tissue was obtained and stored prior to the initiation of the project.', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('PROSPECTIVE_COLLECTION', 'Tissue Prospective Collection Indicator', 'Text indicator for the time frame of tissue procurement, indicating that the tissue was procured in parallel to the project.', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('FORM_COMPLETION_DATE', 'Form completion date', 'Form completion date', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OTHER_PATIENT_ID', 'Other Patient ID', 'Legacy DMP patient identifier (DMPnnnn)', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DFS_MONTHS', 'Disease Free (Months)', 'Disease free (months) since initial treatment.', 'NUMBER', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DFS_STATUS', 'Disease Free Status', 'Disease free status since initial treatment.', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OS_MONTHS', 'Overall Survival (Months)', 'Overall survival in months since initial diagonosis.', 'NUMBER', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OS_STATUS', 'Overall Survival Status', 'Overall patient survival status.', 'STRING', 1, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OTHER_SAMPLE_ID', 'Other Sample ID', 'Legacy DMP sample identifier (DMPnnnn)', 'STRING', 0, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('DAYS_TO_COLLECTION', 'Days to Sample Collection.', 'Days to sample collection.', 'STRING', 0, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('IS_FFPE', 'Is FFPE', 'If the sample is from FFPE', 'STRING', 0, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('OCT_EMBEDDED', 'Oct embedded', 'Oct embedded', 'STRING', 0, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('PATHOLOGY_REPORT_FILE_NAME', 'Pathology Report File Name', 'Pathology Report File Name', 'STRING', 0, '1', 2);
INSERT INTO "clinical_attribute_meta" ("ATTR_ID", "DISPLAY_NAME", "DESCRIPTION", "DATATYPE", "PATIENT_ATTRIBUTE", "PRIORITY", "CANCER_STUDY_ID") VALUES ('SAMPLE_TYPE', 'Sample Type', 'The type of sample (i.e., normal, primary, met, recurrence).', 'STRING', 0, '1', 2);

-- structural_variant
INSERT INTO "structural_variant" ("SAMPLE_ID","BREAKPOINT_TYPE","ANNOTATION","COMMENTS","CONFIDENCE_CLASS","CONNECTION_TYPE","EVENT_INFO","MAPQ","NORMAL_READ_COUNT","NORMAL_VARIANT_COUNT","PAIRED_END_READ_SUPPORT","SITE1_CHROM","SITE1_DESC","SITE1_ENTREZ_GENE_ID","SITE1_POS","SITE2_CHROM","SITE2_DESC","SITE2_ENTREZ_GENE_ID","SITE2_POS","SPLIT_READ_SUPPORT","SV_CLASS_NAME","SV_DESC","SV_LENGTH","TUMOR_READ_COUNT","TUMOR_VARIANT_COUNT","VARIANT_STATUS_NAME","GENETIC_PROFILE_ID") VALUES (1,'PRECISE','n/a','n/a','AUTO_OK','3to5','Transcript fusion (ERBB2-GRB7)',0,7181,0,11,'17','Intron of ERBB2(+): 51bp after exon 26',2064,37883851,'17','5-UTR of GRB7(+): 1Kb before coding start',2886,37897379,31,'DELETION','n/a',13528,4389,6,'NEW_VARIANT',7);
INSERT INTO "structural_variant" ("SAMPLE_ID","BREAKPOINT_TYPE","ANNOTATION","COMMENTS","CONFIDENCE_CLASS","CONNECTION_TYPE","EVENT_INFO","MAPQ","NORMAL_READ_COUNT","NORMAL_VARIANT_COUNT","PAIRED_END_READ_SUPPORT","SITE1_CHROM","SITE1_DESC","SITE1_ENTREZ_GENE_ID","SITE1_POS","SITE2_CHROM","SITE2_DESC","SITE2_ENTREZ_GENE_ID","SITE2_POS","SPLIT_READ_SUPPORT","SV_CLASS_NAME","SV_DESC","SV_LENGTH","TUMOR_READ_COUNT","TUMOR_VARIANT_COUNT","VARIANT_STATUS_NAME","GENETIC_PROFILE_ID") VALUES (2,'PRECISE','n/a','n/a','AUTO_OK','3to5','Protein fusion: mid-exon (ERBB2-GRB7)',0,7062,0,97,'17','Exon 25 of ERBB2(+)',2064,37883138,'17','Intron of GRB7(+): 56bp before exon 10',2886,37901416,71,'DELETION','n/a',18278,9849,60,'NEW_VARIANT',7);
INSERT INTO "structural_variant" ("SAMPLE_ID","BREAKPOINT_TYPE","ANNOTATION","COMMENTS","CONFIDENCE_CLASS","CONNECTION_TYPE","EVENT_INFO","MAPQ","NORMAL_READ_COUNT","NORMAL_VARIANT_COUNT","PAIRED_END_READ_SUPPORT","SITE1_CHROM","SITE1_DESC","SITE1_ENTREZ_GENE_ID","SITE1_POS","SITE2_CHROM","SITE2_DESC","SITE2_ENTREZ_GENE_ID","SITE2_POS","SPLIT_READ_SUPPORT","SV_CLASS_NAME","SV_DESC","SV_LENGTH","TUMOR_READ_COUNT","TUMOR_VARIANT_COUNT","VARIANT_STATUS_NAME","GENETIC_PROFILE_ID") VALUES (3,'PRECISE','n/a','ERBB2 (NM_004448) rearrangement','MANUAL_OK','3to5','Deletion of 1 exon: in frame',0,7212,0,31,'17','Intron of ERBB2(+): 46bp after exon 15',2064,37873779,'17','Intron of ERBB2(+): 501bp before exon 17',2064,37879041,10,'DELETION','n/a',5262,3101,7,'NEW_VARIANT',7);
INSERT INTO "structural_variant" ("SAMPLE_ID","BREAKPOINT_TYPE","ANNOTATION","COMMENTS","CONFIDENCE_CLASS","CONNECTION_TYPE","EVENT_INFO","MAPQ","NORMAL_READ_COUNT","NORMAL_VARIANT_COUNT","PAIRED_END_READ_SUPPORT","SITE1_CHROM","SITE1_DESC","SITE1_ENTREZ_GENE_ID","SITE1_POS","SITE2_CHROM","SITE2_DESC","SITE2_ENTREZ_GENE_ID","SITE2_POS","SPLIT_READ_SUPPORT","SV_CLASS_NAME","SV_DESC","SV_LENGTH","TUMOR_READ_COUNT","TUMOR_VARIANT_COUNT","VARIANT_STATUS_NAME","GENETIC_PROFILE_ID") VALUES (4,'PRECISE','n/a','n/a','MANUAL_OK','3to5','Deletion of 1 exon',0,5328,0,52,'17','Intron of ERBB2(+): 196bp after exon 15',2064,37873929,'17','Intron of ERBB2(+): 442bp after exon 16',2064,37876529,33,'DELETION','n/a',2600,19320,13,'NEW_VARIANT',7);
