<!DOCTYPE html>
<html lang="en" class="h-full">
<head>
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-HEJSRG35RS"></script>
    <script data-cfasync="false">window.dataLayer = window.dataLayer || [];function gtag(){dataLayer.push(arguments);}gtag('js', new Date());gtag('config', 'G-HEJSRG35RS');</script>

    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="turbo-cache-control" content="no-cache" data-turbo-track="reload" data-track-token="3.9.0.810668354192">

    <!-- See retype.com -->
    <meta name="generator" content="Retype 3.9.0">

    <!-- Primary Meta Tags -->
    <title>News</title>
    <meta name="title" content="News">
    <meta name="description" content="Added data consisting of 4,571 samples from 10 studies:">

    <!-- Canonical -->
    <link rel="canonical" href="https://docs.cbioportal.org/news/">

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="website">
    <meta property="og:url" content="https://docs.cbioportal.org/news/">
    <meta property="og:title" content="News">
    <meta property="og:description" content="Added data consisting of 4,571 samples from 10 studies:">
    <meta property="og:image" content="https://github.com/user-attachments/assets/c147794f-390a-4dec-a252-8fd324ea72e9">

    <!-- Twitter -->
    <meta property="twitter:card" content="summary_large_image">
    <meta property="twitter:url" content="https://docs.cbioportal.org/news/">
    <meta property="twitter:title" content="News">
    <meta property="twitter:description" content="Added data consisting of 4,571 samples from 10 studies:">
    <meta property="twitter:image" content="https://github.com/user-attachments/assets/c147794f-390a-4dec-a252-8fd324ea72e9">

    <script data-cfasync="false">(function(){var cl=document.documentElement.classList,ls=localStorage.getItem("retype_scheme"),hd=cl.contains("dark"),hl=cl.contains("light"),wm=window.matchMedia&&window.matchMedia("(prefers-color-scheme: dark)").matches;if(ls==="dark"||(!ls&&wm&&!hd&&!hl)){cl.remove("light");cl.add("dark")}else if(ls==="light"||(!ls&&!wm&&!hd&&!hl)){cl.remove("dark");cl.add("light")}})();</script>

    <link href="../images/cbio-logo.png" rel="icon">
    <link href="../resources/css/retype.css?v=3.9.0.810668354192" rel="stylesheet">

    <script data-cfasync="false" src="../resources/js/config.js?v=3.9.0.810668354192" data-turbo-eval="false" defer></script>
    <script data-cfasync="false" src="../resources/js/retype.js?v=3.9.0" data-turbo-eval="false" defer></script>
    <script id="lunr-js" data-cfasync="false" src="../resources/js/lunr.js?v=3.9.0.810668354192" data-turbo-eval="false" defer></script>

    <link rel="stylesheet" href="/static/custom.css"></link>
</head>
<body>
    <div id="docs-app" class="relative text-base antialiased text-gray-700 bg-white font-body dark:bg-dark-850 dark:text-dark-300">
        <div class="absolute bottom-0 left-0 bg-gray-100 dark:bg-dark-800" style="top: 5rem; right: 50%"></div>
    
        <header id="docs-site-header" class="sticky top-0 z-30 flex w-full h-16 bg-white border-b border-gray-200 md:h-20 dark:bg-dark-850 dark:border-dark-650">
            <div class="container relative flex items-center justify-between pr-6 grow md:justify-start">
                <!-- Mobile menu button skeleton -->
                <button v-cloak class="skeleton docs-mobile-menu-button flex items-center justify-center shrink-0 overflow-hidden dark:text-white focus:outline-none rounded-full w-10 h-10 ml-3.5 md:hidden"><svg xmlns="http://www.w3.org/2000/svg" class="mb-px shrink-0" width="24" height="24" viewBox="0 0 24 24" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor"><path d="M2 4h20v2H2zM2 11h20v2H2zM2 18h20v2H2z"></path></g></svg></button>
                <div v-cloak id="docs-sidebar-toggle"></div>
        
                <!-- Logo -->
                <div class="flex items-center justify-between h-full py-2 md:w-75">
                    <div class="flex items-center px-2 md:px-6">
                        <a id="docs-site-logo" href="../" class="flex items-center leading-snug text-xl">
                            <span class="w-10 mr-2 grow-0 shrink-0 overflow-hidden">
                                <img class="max-h-10 dark:hidden md:inline-block" src="../images/cbio-logo.png">
                                <img class="max-h-10 hidden dark:inline-block" src="../images/cbio-logo.png">
                            </span>
                            <span class="dark:text-white font-bold line-clamp-1 md:line-clamp-2">cBioPortal Docs</span>
                        </a>
                    </div>
        
                    <span class="hidden h-8 border-r md:inline-block dark:border-dark-650"></span>
                </div>
        
                <div class="flex justify-between md:grow">
                    <!-- Top Nav -->
                    <nav class="hidden md:flex">
                        <ul class="flex flex-col mb-4 md:pl-16 md:mb-0 md:flex-row md:items-center">
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.cbioportal.org">
                                    <img class="mb-px mr-1" width="18" alt="" src="../images/cbio-logo.png">
                                    <span>cBioPortal</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M12 1C5.923 1 1 5.923 1 12c0 4.867 3.149 8.979 7.521 10.436.55.096.756-.233.756-.522 0-.262-.013-1.128-.013-2.049-2.764.509-3.479-.674-3.699-1.292-.124-.317-.66-1.293-1.127-1.554-.385-.207-.936-.715-.014-.729.866-.014 1.485.797 1.691 1.128.99 1.663 2.571 1.196 3.204.907.096-.715.385-1.196.701-1.471-2.448-.275-5.005-1.224-5.005-5.432 0-1.196.426-2.186 1.128-2.956-.111-.275-.496-1.402.11-2.915 0 0 .921-.288 3.024 1.128a10.193 10.193 0 0 1 2.75-.371c.936 0 1.871.123 2.75.371 2.104-1.43 3.025-1.128 3.025-1.128.605 1.513.221 2.64.111 2.915.701.77 1.127 1.747 1.127 2.956 0 4.222-2.571 5.157-5.019 5.432.399.344.743 1.004.743 2.035 0 1.471-.014 2.654-.014 3.025 0 .289.206.632.756.522C19.851 20.979 23 16.854 23 12c0-6.077-4.922-11-11-11Z"/>
                                        </g>
                                    </svg>
                                    <span>Github</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/issues/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                        </g>
                                    </svg>
                                    <span>Issues</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://groups.google.com/g/cbioportal">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M1.75 1h12.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 14.25 14H8.061l-2.574 2.573A1.458 1.458 0 0 1 3 15.543V14H1.75A1.75 1.75 0 0 1 0 12.25v-9.5C0 1.784.784 1 1.75 1ZM1.5 2.75v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25H1.75a.25.25 0 0 0-.25.25Z"/><path d="M22.5 8.75a.25.25 0 0 0-.25-.25h-3.5a.75.75 0 0 1 0-1.5h3.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 22.25 20H21v1.543a1.457 1.457 0 0 1-2.487 1.03L15.939 20H10.75A1.75 1.75 0 0 1 9 18.25v-1.465a.75.75 0 0 1 1.5 0v1.465c0 .138.112.25.25.25h5.5a.75.75 0 0 1 .53.22l2.72 2.72v-2.19a.75.75 0 0 1 .75-.75h2a.25.25 0 0 0 .25-.25v-9.5Z"/>
                                        </g>
                                    </svg>
                                    <span>Google Group</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="../donate/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M14 20.408c-.492.308-.903.546-1.192.709-.153.086-.308.17-.463.252h-.002a.75.75 0 0 1-.686 0 16.709 16.709 0 0 1-.465-.252 31.147 31.147 0 0 1-4.803-3.34C3.8 15.572 1 12.331 1 8.513 1 5.052 3.829 2.5 6.736 2.5 9.03 2.5 10.881 3.726 12 5.605 13.12 3.726 14.97 2.5 17.264 2.5 20.17 2.5 23 5.052 23 8.514c0 3.818-2.801 7.06-5.389 9.262A31.146 31.146 0 0 1 14 20.408Z"/>
                                        </g>
                                    </svg>
                                    <span>Donate</span>
                                </a>
                            </li>
        
                        </ul>
                    </nav>
        
                    <!-- Header Right Skeleton -->
                    <div v-cloak class="flex justify-end grow skeleton">
        
                        <!-- Search input mock -->
                        <div class="relative hidden w-40 lg:block lg:max-w-sm lg:ml-auto">
                            <div class="absolute flex items-center justify-center h-full pl-3 dark:text-dark-300">
                                <svg xmlns="http://www.w3.org/2000/svg" class="icon-base" width="16" height="16" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 1px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                            </div>
                            <input class="w-full h-10 placeholder-gray-400 transition-colors duration-200 ease-in bg-gray-200 border border-transparent rounded md:text-sm hover:bg-white hover:border-gray-300 focus:outline-none focus:bg-white focus:border-gray-500 dark:bg-dark-600 dark:border-dark-600 dark:placeholder-dark-400" style="padding: 0.625rem 0.75rem 0.625rem 2rem" type="text" placeholder="Search">
                        </div>
        
                        <!-- Mobile search button -->
                        <div class="flex items-center justify-center w-10 h-10 lg:hidden">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="20" height="20" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                        </div>
        
                        <!-- Dark mode switch placeholder -->
                        <div class="w-10 h-10 lg:ml-2"></div>
        
                        <!-- History button -->
                        <div class="flex items-center justify-center w-10 h-10" style="margin-right: -0.625rem;">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="22" height="22" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><g ><path d="M12.01 6.01c-.55 0-1 .45-1 1V12a1 1 0 00.4.8l3 2.22a.985.985 0 001.39-.2.996.996 0 00-.21-1.4l-2.6-1.92V7.01c.02-.55-.43-1-.98-1z"></path><path d="M12.01 1.91c-5.33 0-9.69 4.16-10.05 9.4l-.29-.26a.997.997 0 10-1.34 1.48l1.97 1.79c.19.17.43.26.67.26s.48-.09.67-.26l1.97-1.79a.997.997 0 10-1.34-1.48l-.31.28c.34-4.14 3.82-7.41 8.05-7.41 4.46 0 8.08 3.63 8.08 8.09s-3.63 8.08-8.08 8.08c-2.18 0-4.22-.85-5.75-2.4a.996.996 0 10-1.42 1.4 10.02 10.02 0 007.17 2.99c5.56 0 10.08-4.52 10.08-10.08.01-5.56-4.52-10.09-10.08-10.09z"></path></g></g></svg>
                        </div>
                    </div>
        
                    <div v-cloak class="flex justify-end grow">
                        <div id="docs-mobile-search-button"></div>
                        <doc-search-desktop></doc-search-desktop>
        
                        <doc-theme-switch class="lg:ml-2"></doc-theme-switch>
                        <doc-history></doc-history>
                    </div>
                </div>
            </div>
        </header>
    
        <div class="container relative flex bg-white">
            <!-- Sidebar Skeleton -->
            <div v-cloak class="fixed flex flex-col shrink-0 duration-300 ease-in-out bg-white border-gray-200 sidebar top-20 w-75 border-r h-screen md:sticky transition-transform skeleton dark:bg-dark-800 dark:border-dark-650">
            
                <div class="flex items-center h-16 px-6">
                    <input class="w-full h-8 px-3 py-2 transition-colors duration-200 ease-linear bg-white border border-gray-200 rounded shadow-none text-sm focus:outline-none focus:border-gray-600 dark:bg-dark-600 dark:border-dark-600" type="text" placeholder="Filter">
                </div>
            
                <div class="pl-6 mt-1 mb-4">
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                </div>
            
                <div class="shrink-0 mt-auto bg-transparent dark:border-dark-650">
                    <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                        <span class="text-xs whitespace-nowrap">Powered by</span>
                        <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                    </a>
                </div>
            </div>
            
            <!-- Sidebar component -->
            <doc-sidebar v-cloak>
                <template #sidebar-footer>
                    <div class="shrink-0 mt-auto border-t md:bg-transparent md:border-none dark:border-dark-650">
            
                        <div class="py-3 px-6 md:hidden border-b dark:border-dark-650">
                            <nav>
                                <ul class="flex flex-wrap justify-center items-center">
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.cbioportal.org">
                                            <img class="mb-px mr-1" width="18" alt="" src="../images/cbio-logo.png">
                                            <span>cBioPortal</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M12 1C5.923 1 1 5.923 1 12c0 4.867 3.149 8.979 7.521 10.436.55.096.756-.233.756-.522 0-.262-.013-1.128-.013-2.049-2.764.509-3.479-.674-3.699-1.292-.124-.317-.66-1.293-1.127-1.554-.385-.207-.936-.715-.014-.729.866-.014 1.485.797 1.691 1.128.99 1.663 2.571 1.196 3.204.907.096-.715.385-1.196.701-1.471-2.448-.275-5.005-1.224-5.005-5.432 0-1.196.426-2.186 1.128-2.956-.111-.275-.496-1.402.11-2.915 0 0 .921-.288 3.024 1.128a10.193 10.193 0 0 1 2.75-.371c.936 0 1.871.123 2.75.371 2.104-1.43 3.025-1.128 3.025-1.128.605 1.513.221 2.64.111 2.915.701.77 1.127 1.747 1.127 2.956 0 4.222-2.571 5.157-5.019 5.432.399.344.743 1.004.743 2.035 0 1.471-.014 2.654-.014 3.025 0 .289.206.632.756.522C19.851 20.979 23 16.854 23 12c0-6.077-4.922-11-11-11Z"/>
                                                </g>
                                            </svg>
                                            <span>Github</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/issues/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                                </g>
                                            </svg>
                                            <span>Issues</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://groups.google.com/g/cbioportal">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M1.75 1h12.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 14.25 14H8.061l-2.574 2.573A1.458 1.458 0 0 1 3 15.543V14H1.75A1.75 1.75 0 0 1 0 12.25v-9.5C0 1.784.784 1 1.75 1ZM1.5 2.75v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25H1.75a.25.25 0 0 0-.25.25Z"/><path d="M22.5 8.75a.25.25 0 0 0-.25-.25h-3.5a.75.75 0 0 1 0-1.5h3.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 22.25 20H21v1.543a1.457 1.457 0 0 1-2.487 1.03L15.939 20H10.75A1.75 1.75 0 0 1 9 18.25v-1.465a.75.75 0 0 1 1.5 0v1.465c0 .138.112.25.25.25h5.5a.75.75 0 0 1 .53.22l2.72 2.72v-2.19a.75.75 0 0 1 .75-.75h2a.25.25 0 0 0 .25-.25v-9.5Z"/>
                                                </g>
                                            </svg>
                                            <span>Google Group</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="../donate/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M14 20.408c-.492.308-.903.546-1.192.709-.153.086-.308.17-.463.252h-.002a.75.75 0 0 1-.686 0 16.709 16.709 0 0 1-.465-.252 31.147 31.147 0 0 1-4.803-3.34C3.8 15.572 1 12.331 1 8.513 1 5.052 3.829 2.5 6.736 2.5 9.03 2.5 10.881 3.726 12 5.605 13.12 3.726 14.97 2.5 17.264 2.5 20.17 2.5 23 5.052 23 8.514c0 3.818-2.801 7.06-5.389 9.262A31.146 31.146 0 0 1 14 20.408Z"/>
                                                </g>
                                            </svg>
                                            <span>Donate</span>
                                        </a>
                                    </li>
            
                                </ul>
                            </nav>
                        </div>
            
                        <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                            <span class="text-xs whitespace-nowrap">Powered by</span>
                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                        </a>
                    </div>
                </template>
            </doc-sidebar>
    
            <div class="grow min-w-0 dark:bg-dark-850">
                <!-- Render "toolbar" template here on api pages --><!-- Render page content -->
                <div class="flex">
                    <div class="min-w-0 p-4 grow md:px-16">
                        <main class="relative pb-12 lg:pt-2">
                            <div class="docs-markdown" id="docs-content">
                                <!-- Rendered if sidebar right is enabled -->
                                <div id="docs-sidebar-right-toggle"></div>
                                <!-- Page content  -->
<doc-anchor-target id="news" class="break-words">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#news">#</doc-anchor-trigger>
        <span>News</span>
    </h1>
</doc-anchor-target>
<doc-anchor-target id="may-06-2025">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-06-2025">#</doc-anchor-trigger>
        <span>May 06, 2025</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 4,571 samples from 10 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=pdac_msk_2024">Pancreatic Adenocarcinoma (MSK, Nat Med 2024)</a> <em>2336 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=csf_msk_2024">Cerebrospinal Fluid Circulating Tumor DNA (MSK, Acta Neuropathol Commun 2024)</a> <em>1007 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ovary_geomx_gray_foundation_2024">Ovarian Cancer (Gray Foundation, Cancer Discov 2024)</a> <em>567 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=normal_skin_melanocytes_2024">Normal Melanocytes (UCSF, Nature 2020)</a> <em>153 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=normal_skin_keratinocytes_2024">Normal Keratinocytes from human skin (UCSF, BioRxiv 2024)</a> <em>136 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=braf_msk_archer_2024">BRAF Fusions - ARCHER Clinical Sequencing Cohort (MSK, Clin Cancer Res 2024)</a> <em>107 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=braf_msk_impact_2024">BRAF Fusions - IMPACT Clinical Sequencing Cohort (MSK, Clin Cancer Res 2024)</a> <em>105 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=sarcoma_ucla_2024">Sarcoma (UCLA, Cell 2024)</a> <em>63 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=chl_sccc_2023">Classical Hodgkins Lymphoma (SCCC, Blood Cancer Discov 2023)</a> <em>61 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=normal_skin_fibroblast_2024">Normal Fibroblasts from human skin (UCSF, BioRxiv 2024)</a> <em>18 samples</em></li>
</ul>
</li>
<li><p><strong>Gene Tables Update:</strong> Updated tables of genes (main and alias), based on <a href="https://www.genenames.org/download/archive/monthly/tsv/">Oct 1, 2024 HGNC release</a>. See seedDB release notes <a href="https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database-schema-2131">here</a> for details.</p>
</li>
</ul>
<doc-anchor-target id="april-17-2025">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-17-2025">#</doc-anchor-trigger>
        <span>April 17, 2025</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Released a <a href="https://about.cbioportal.org/">new about page</a>, which includes a roadmap and an overview of 2024 accomplishments.</li>
</ul>
<doc-anchor-target id="january-28-2025">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-28-2025">#</doc-anchor-trigger>
        <span>January 28, 2025</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement:</strong> You can now re-order the stacked bar charts in the Plots Tab alphabetically, by the number of samples or by the data category in the plot. This enhancement allows you to create more ready-to-publish figures using cBioPortal. As usual, the plots created in the Plots Tab can be downloaded in different formats. <a href="https://www.cbioportal.org/study/plots?id=msk_impact_2017&amp;plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_TYPE%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&amp;plots_vert_selection=%7B%22selectedGeneOption%22%3A7157%2C%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22SPECIMEN_TYPE%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&amp;plots_coloring_selection=%7B%7D">Example: specimen type for each cancer type in MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://github.com/user-attachments/assets/c147794f-390a-4dec-a252-8fd324ea72e9" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="december-18-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-18-2024">#</doc-anchor-trigger>
        <span>December 18, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 34,904 samples from 9 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_chord_2024">MSK-CHORD (MSK, Nature 2024)</a> <em>25040 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_ctdna_vte_2024">MSK ctDNA Sequencing Cohort (MSK, Nature Med 2024)</a> <em>5567 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prostate_msk_2024">Prostate Cancer (MSK, Clin Cancer Res 2024)</a> <em>2260 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hcc_msk_2024">Hepatocellular Carcinoma (MSK, Clin Cancer Res 2024)</a> <em>1370 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pancreas_msk_2024">Pancreatic Cancer (MSK, Cancer Cell 2024)</a> <em>395 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=panet_msk_2018">Metastatic Pancreatic Neuroendocrine Tumor (MSK, JCO Precis Oncol 2018)</a> <em>96 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pcnsl_msk_2024">Recurrent CNS Lymphoma - Ibrutinib Long Term Follow Up (MSK, Cancer Cell 2024)</a> <em>86 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_dfci_2020">Metastatic Breast Cancer (DFCI, Cancer Discovery 2020)</a> <em>59 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=asclc_msk_2024">Atypical Small Cell Lung Cancer (MSK, Cancer Discovery 2024)</a> <em>31 samples</em></li>
</ul>
</li>
<li><p><strong>Data Improvement</strong></p>
<ul>
<li>TBL scores addition: Tumor Break Load (TBL) scores is added to PCAWG, CCLE, and all 32 TCGA Pan-Cancer studies. The data is sourced from <a href="https://www.biorxiv.org/content/10.1101/2024.12.08.626771v1.full">Lakbir et al., bioRxiv 2024</a> and <a href="https://pubmed.ncbi.nlm.nih.gov/36334560/">Lakbir et al., Eur J Cancer 2022</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="december-11-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-11-2024">#</doc-anchor-trigger>
        <span>December 11, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 5,567 samples from 6 TARGET studies from the <a href="https://gdc.cancer.gov">Genomic Data Commons (GDC)</a> as part of the <a href="https://datacommons.cancer.gov">Cancer Research Data Commons (NCI-CRDC)</a> initiative.
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=aml_target_gdc">Acute Myeloid Leukemia (TARGET, GDC)</a> <em>2482 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bll_target_gdc">B-Lymphoblastic Leukemia/Lymphoma - Phase II (TARGET, GDC)</a> <em>1004 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nbl_target_gdc">Neuroblastoma (TARGET, GDC)</a> <em>840 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=wt_target_gdc">Wilms&#x27; Tumor (TARGET, GDC)</a> <em>655 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=alal_target_gdc">Acute Leukemias of Ambiguous Lineage - Phase III (TARGET, GDC)</a> <em>351 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=os_target_gdc">Osteosarcoma (TARGET, GDC)</a> <em>235 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="november-18-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-18-2024">#</doc-anchor-trigger>
        <span>November 18, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature:</strong> New data source <a href="https://alphamissense.hegelab.org/">AlphaMissense</a> is now integrated in Mutations tab. View the AlphaMissense data in the <a href="https://bit.ly/48yOi7A">Mutation Table</a> and the <a href="https://www.cbioportal.org/mutation_mapper">Mutation Mapper Tool</a>.</p>
<p><img src="https://github.com/user-attachments/assets/5c43b1b7-83e0-4bf2-b8ed-e43eda0f02c8" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="october-30-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-30-2024">#</doc-anchor-trigger>
        <span>October 30, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New integration</strong>: Structural Variants (SV)s of the <a href="https://www.cbioportal.org/study/summary?id=pancan_pcawg_2020">PCAWG dataset</a> can now be viewed in Chromoscope on the <em>Patient View</em> (Funding by <a href="https://www.the-ici-fund.org/recipients/details/chromoscope-interactive-visualization-for-clinical-interpretation-of-whole-cancer-genomes">the Fund for Innovation in Cancer Informatics (ICI)</a>). <a href="https://www.cbioportal.org/patient/openResource_CHROMOSCOPE?studyId=pancan_pcawg_2020&amp;caseId=DO2706">Example: SVs in breast cancer case DO2706 with BRCA1 mutation and HRD signature</a>
<img src="https://github.com/user-attachments/assets/9ddc68ff-840b-4470-8f26-824f1faf959f" alt="chromoscope image" /></li>
</ul>
<doc-anchor-target id="october-8-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-8-2024">#</doc-anchor-trigger>
        <span>October 8, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Mutation Assessor V4 data is available in the portal.
View the new V4 Mutation Assessor data in the <a href="https://bit.ly/48yOi7A">Mutation Table</a> and the <a href="https://www.cbioportal.org/mutation_mapper">Mutation Mapper Tool</a>.
<img src="https://github.com/user-attachments/assets/ff53ee41-59d3-4c69-a494-3a7ca9a7c2d0" alt="image" /></li>
</ul>
<doc-anchor-target id="september-27-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-27-2024">#</doc-anchor-trigger>
        <span>September 27, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 2,409 samples from 10 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=pancan_mappyacts_2022">Pediatric European MAPPYACTS Trial (Gustave Roussy, Cancer Discov 2022)</a> <em>786 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hcc_clca_2024">Hepatocellular Carcinoma (CLCA, Nature 2024)</a> <em>494 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gist_msk_2023">Gastrointestinal Stromal Tumors (MSK, Clin Cancer Res 2023)</a> <em>469 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mbn_msk_2024">Diffuse Large B-Cell Lymphoma (MSK, Haematologica 2024)</a> <em>396 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ptad_msk_2024">Pituitary Adenoma (MSK, Acta Neuropathologica 2024)</a> <em>104 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_fuscc_2020">Triple-Negative Breast Cancer (FUSCC, Cell Research 2020)</a> <em>69 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_msk_2024">Endometrial and Ovarian Cancer (MSK, Nature Medicine 2024)</a> <em>33 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cscc_ranson_2022">Cutaneous Squamous Cell Carcinoma (UOW, Front Oncol 2022)</a> <em>25 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_msk_tcga_2021">RAD51B Associated Mixed Cancers (MSK, NPJ Breast Cancer 2021)</a> <em>19 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=plmeso_msk_2024">Pleural Mesothelioma (MSK, Clin Cancer Res 2024)</a> <em>14 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="august-28-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-28-2024">#</doc-anchor-trigger>
        <span>August 28, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New feature</strong> (<em>Windows users only</em>): The Group Comparison results can now be visualized in 3D with AVM, an interactive 3D visualization software tool with data shaping functions and integrated pathways. Read more on <a href="https://bit.ly/avm-cbioportal">AVM for cBioPortal</a>.
<ul>
<li><strong>Note for private instances</strong>: To enable this feature in private cBioPortal instances, please see <a href="https://docs.cbioportal.org/deployment/customization/application.properties-reference#add-custom-buttons-to-data-tables">this instruction</a>.
<img src="https://github.com/user-attachments/assets/4150368b-c38c-41f8-a1da-af8c3d79ebf1" alt="image" />
<img src="https://github.com/user-attachments/assets/bba1221d-e07a-460a-ac6e-774d737cdc30" alt="image" /></li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-29-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-29-2024">#</doc-anchor-trigger>
        <span>July 29, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 12,863 samples from 33 TCGA and 10 CPTAC studies from the <a href="https://gdc.cancer.gov">Genomic Data Commons (GDC)</a> as part of the <a href="https://datacommons.cancer.gov">Cancer Research Data Commons (NCI-CRDC)</a> initiative. More information can be found on our <a href="https://docs.cbioportal.org/user-guide/faq/#tcga">FAQ</a>.
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=aml_tcga_gdc">Acute Myeloid Leukemia (TCGA, GDC)</a> <em>200 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=acc_tcga_gdc">Adrenocortical Carcinoma (TCGA, GDC)</a> <em>92 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=blca_tcga_gdc">Bladder Urothelial Carcinoma (TCGA, GDC)</a> <em>413 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cesc_tcga_gdc">Cervical Squamous Cell Carcinoma (TCGA, GDC)</a> <em>309 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=chol_tcga_gdc">Cholangiocarcinoma (TCGA, GDC)</a> <em>51 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=chrcc_tcga_gdc">Chromophobe Renal Cell Carcinoma (TCGA, GDC)</a> <em>66 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coad_tcga_gdc">Colon Adenocarcinoma (TCGA, GDC)</a> <em>463 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=skcm_tcga_gdc">Cutaneous Melanoma (TCGA, GDC)</a> <em>473 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=difg_tcga_gdc">Diffuse Glioma (TCGA, GDC)</a> <em>530 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=dlbclnos_tcga_gdc">Diffuse Large B-Cell Lymphoma, NOS (TCGA, GDC)</a> <em>48 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_tcga_gdc">Endometrial Carcinoma (TCGA, GDC)</a> <em>548 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=esca_tcga_gdc">Esophageal Adenocarcinoma (TCGA, GDC)</a> <em>186 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbm_tcga_gdc">Glioblastoma Multiforme (TCGA, GDC)</a> <em>610 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hnsc_tcga_gdc">Head and Neck Squamous Cell Carcinoma (TCGA, GDC)</a> <em>530 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hcc_tcga_gdc">Hepatocellular Carcinoma (TCGA, GDC)</a> <em>379 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hgsoc_tcga_gdc">High-Grade Serous Ovarian Cancer (TCGA, GDC)</a> <em>602 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_tcga_gdc">Invasive Breast Carcinoma (TCGA, GDC)</a> <em>1103 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_tcga_gdc">Lung Adenocarcinoma (TCGA, GDC)</a> <em>571 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lusc_tcga_gdc">Lung Squamous Cell Carcinoma (TCGA, GDC)</a> <em>503 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mnet_tcga_gdc">Miscellaneous Neuroepithelial Tumor (TCGA, GDC)</a> <em>184 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nsgct_tcga_gdc">Non-Seminomatous Germ Cell Tumor (TCGA, GDC)</a> <em>156 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=paad_tcga_gdc">Pancreatic Adenocarcinoma (TCGA, GDC)</a> <em>186 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prcc_tcga_gdc">Papillary Renal Cell Carcinoma (TCGA, GDC)</a> <em>292 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=thpa_tcga_gdc">Papillary Thyroid Cancer (TCGA, GDC)</a> <em>515 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=plmeso_tcga_gdc">Pleural Mesothelioma (TCGA, GDC)</a> <em>87 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_tcga_gdc">Prostate Adenocarcinoma (TCGA, GDC)</a> <em>501 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=read_tcga_gdc">Rectal Adenocarcinoma (TCGA, GDC)</a> <em>171 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ccrcc_tcga_gdc">Renal Clear Cell Carcinoma (TCGA, GDC)</a> <em>537 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=soft_tissue_tcga_gdc">Soft Tissue Cancer (TCGA, GDC)</a> <em>264 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=stad_tcga_gdc">Stomach Adenocarcinoma (TCGA, GDC)</a> <em>443 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=thym_tcga_gdc">Thymoma (TCGA, GDC)</a> <em>124 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucs_tcga_gdc">Uterine Carcinosarcoma/Uterine Malignant Mixed Mullerian Tumor (TCGA, GDC)</a> <em>57 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=um_tcga_gdc">Uveal Melanoma (TCGA, GDC)</a> <em>80 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=breast_cptac_gdc">Breast Cancer (CPTAC, GDC)</a> <em>154 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brain_cptac_gdc">CNS/Brain Cancer (CPTAC, GDC)</a> <em>74 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coad_cptac_gdc">Colon Adenocarcinoma (CPTAC, GDC)</a> <em>109 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ohnca_cptac_gdc">Head and Neck Carcinoma, Other (CPTAC, GDC)</a> <em>150 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_cptac_gdc">Lung Adenocarcinoma (CPTAC, GDC)</a> <em>180 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lusc_cptac_gdc">Lung Squamous Cell Carcinoma (CPTAC, GDC)</a> <em>90 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ovary_cptac_gdc">Ovarian Cancer (CPTAC, GDC)</a> <em>113 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pancreas_cptac_gdc">Pancreatic Cancer (CPTAC, GDC)</a> <em>139 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rcc_cptac_gdc">Renal Cell Carcinoma (CPTAC, GDC)</a> <em>288 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=uec_cptac_gdc">Uterine Endometrioid Carcinoma (CPTAC, GDC)</a> <em>292 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-10-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-10-2024">#</doc-anchor-trigger>
        <span>July 10, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature:</strong> Explore mutational signatures on the patient page. The Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020) study has mutational signatures data available for over 2,500 cases. The new Mutational Signatures tab on the patient page graphically displays the types of mutations found in each case along with COSMIC reference signatures. <a href="https://www.cbioportal.org/patient/mutationalSignatures?studyId=pancan_pcawg_2020&amp;caseId=DO8898">Example: POLE signature in case DO8898 in Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020)</a>
<figure class="content-center">
    <img src="https://github.com/user-attachments/assets/2f92c6d7-de26-42c4-ba04-24cb5df94160" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</li>
</ul>
<doc-anchor-target id="may-29-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-29-2024">#</doc-anchor-trigger>
        <span>May 29, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 9,632 samples from 10 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=sarcoma_msk_2022">Soft Tissue and Bone Sarcoma (MSK, Nat Commun 2022)</a> <em>7494 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cll_broad_2022">Chronic Lymphocytic Leukemia (Broad, Nature Genetics 2022)</a> <em>1154 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mbn_sfu_2023">Mature B-Cell Neoplasms (Simon Fraser University, Blood 2023)</a> <em>297 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=thyroid_gatci_2024">Anaplastic Thyroid Cancers (GATCI, Cell Reports 2024)</a> <em>190 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coadread_cass_2020">Colorectal Cancer (CAS Shanghai, Cancer Cell 2020)</a> <em>146 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_msk_mdanderson_2023">Prostate Cancer MDA PCa PDX (MD Anderson, Clin Cancer Res 2024)</a> <em>88 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_spectrum_tme_2022">Ovarian Cancer - MSK SPECTRUM (MSK, Nature 2022)</a> <em>82 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=difg_msk_2023">IDH-mutated Diffuse Glioma (MSK, Clin Cancer Res 2024)</a> <em>73 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=crc_hta11_htan_2021">Pre-cancer Colorectal Polyps (HTAN Vanderbilt, Cell 2021)</a> <em>61 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_organoids_msk_2022">Prostate Cancer (MSK, Science 2022)</a> <em>47 samples</em></li>
</ul>
</li>
<li><p><strong>Data Improvement</strong></p>
<ul>
<li>Pan-can studies Ancestry addition: Genetic Ancestry data is added to all 32 TCGA Pan-Can studies,in generic assay format. Data source: <a href="https://gdc.cancer.gov/about-data/publications/CCG-AIM-2020">GDC</a>. <a href="https://www.cbioportal.org/comparison/alterations?comparisonId=664275e383e9543d61907074&amp;patientEnrichments=true&amp;selectedEnrichmentEventTypes=%5B%22missense%22%2C%22missense_mutation%22%2C%22missense_variant%22%2C%22indel%22%2C%22nonframeshift%22%2C%22inframe%22%2C%22inframe_del%22%2C%22inframe_deletion%22%2C%22in_frame_del%22%2C%22in_frame_deletion%22%2C%22nonframeshift_deletion%22%2C%22inframe_ins%22%2C%22inframe_insertion%22%2C%22in_frame_ins%22%2C%22in_frame_insertion%22%2C%22nonframeshift_insertion%22%2C%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22silent%22%2C%22synonymous_variant%22%2C%22targeted_region%22%2C%22other%22%2C%22AMP%22%2C%22HOMDEL%22%2C%22structural_variant%22%5D">Example: Ancestry-associated somatic genetic alterations</a>
<img width="1006" alt="Screen Shot 2024-05-20 at 10 52 11 AM" src="https://github.com/cBioPortal/datahub/assets/61431648/628b7eed-9d72-4605-b114-2e438dbd442a"></li>
</ul>
</li>
</ul>
<doc-anchor-target id="may-7-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-7-2024">#</doc-anchor-trigger>
        <span>May 7, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Gene-specific charts in Study View now support mutation data. Two chart types are available: sample mutational status and mutation types.</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/10508276/d94c596a-1103-4aba-995a-ed1709236eee" alt="image" />
<img src="https://github.com/cBioPortal/cbioportal/assets/10508276/5fd2fdd0-07a5-434a-ad42-3f72b7c67bc5" alt="image" /></p>
</li>
<li><p><strong>New Feature</strong>: Categorical data loaded into the generic assay format is now available in group comparison. <a href="https://www.cbioportal.org/comparison/generic_assay_armlevel_cna?comparisonId=616ff06a2e3200072db98b99">Example: Arm-level CNA compared between astrocytoma vs oligodendroglioma in Brain Lower Grade Glioma (TCGA, PanCancer Atlas)</a></p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/10508276/9f8c99a2-eef8-4b5a-ae33-4a2b295e0514" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: OncoPrint now allows you to remove a gene track.</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/10508276/23cd422b-6a4e-4eff-9c04-ac88220591d2" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="apr-2-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-2-2024">#</doc-anchor-trigger>
        <span>Apr 2, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Introducing the cBioPortal Newsletter!</strong> Stay updated with the latest developments, insights, and community highlights of cBioPortal. Subscribe via <a href="https://www.linkedin.com/pulse/april-update-hello-cbioportal-community-cbioportal-22vle/">LinkedIn</a> or <a href="https://groups.google.com/g/cbioportal-news">google groups</a>. We&#x27;ll be sharing valuable updates every few months.</li>
</ul>
<doc-anchor-target id="march-27-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-27-2024">#</doc-anchor-trigger>
        <span>March 27, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: The Plots tab is now available in the Study View. After exploring a cohort and applying any filters of interest in the Study View Summary page, you can now click over to the new Plots tab to explore the cohort by plotting any two attributes against each other. <a href="https://www.cbioportal.org/study/plots?id=ucec_tcga_pub&amp;plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22SUBTYPE%22%7D&amp;plots_vert_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22MUTATION_COUNT%22%2C%22logScale%22%3A%22false%22%7D&amp;plots_coloring_selection=%7B%7D">Example: Mutation count vs subtype in Uterine Corpus Endometrial Carcinoma (TCGA, Nature 2013)</a></p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/10508276/cf74a4ff-57c6-4d71-86cb-4f7440fc3db6" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: The custom selection feature in Study View, which filters to a user-defined list of samples/patients, no longer requires the inclusion of the study ID with each sample ID. If viewing multiple studies in study view and filtering to a sample ID which exists in more than one study, all samples will be displayed by default or the study ID can be included for additional specificity.</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/10508276/2143593d-5fd3-4b1d-b8a4-73a68e027e27" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="mar-5-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#mar-5-2024">#</doc-anchor-trigger>
        <span>Mar 5, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Local Installations Feature</strong>: Make representation of custom driver annotation configurable:
<img src="https://github.com/cBioPortal/cbioportal/assets/1334004/f67fd82f-1789-4427-9ca1-f17fc70c1a09" alt="image" /></li>
</ul>
<doc-anchor-target id="feb-7-2024">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#feb-7-2024">#</doc-anchor-trigger>
        <span>Feb 7, 2024</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New feature</strong>: The <em>Datasets Page</em> now lists what studies have samples profiled for Structural Variants:</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/1334004/0e4e73e4-d9d8-417d-912c-710163480905" alt="image" /></p>
</li>
<li><p><strong>New Major Release</strong>: <strong>v6.0.0</strong> includes a major repackaging of the backend software to improve the development experience. The backend has been upgraded to use Java Spring Boot v3.1.4. The CORE and MAF modules have been moved to new repositories. All other modules have been compacted into a single source repository. We have also updated the JVM to 21 and many libraries have been updated to address security and performance issues. See more information in the <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v6.0.0">v6.0.0 release notes</a>.</p>
</li>
<li><p><strong>Local Installations Feature</strong>: When mutational signature data is loaded, show COSMIC reference signatures on the <em>Patient View</em>:</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/1334004/10f73baa-d28c-4bb7-ad62-4751bd7f61cd" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="dec-29-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#dec-29-2023">#</doc-anchor-trigger>
        <span>Dec 29, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 5,120 samples from 13 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_ancestry_cds_msk_2023">Endometrial Cancer (MSK, Cancer Discovery 2023)</a> <em>1882 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=egc_msk_2023">Esophagogastric Cancer (MSK, J Natl Cancer Inst 2023)</a> <em>902 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=difg_glass">Diffuse Glioma (GLASS Consortium, 2022)</a> <em>693 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bladder_msk_2023">Bladder Cancer (MSK, Clin Cancer Res 2023)</a> <em>526 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bm_nsclc_mskcc_2023">Non-Small Cell Lung Cancer Brain Metastasis (MSK, Nat Commun 2023)</a> <em>322 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=makeanimpact_ccr_2023">MSK Make-an-IMPACT Rare Cancers (MSK, Clin Cancer Res 2023)</a> <em>184 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cervix_msk_2023">Cervical Cancer (MSK, Clin Cancer Res 2023)</a> <em>177 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mtnn_msk_2022">Mature T and NK Neoplasms (MSK, Blood Adv 2023)</a> <em>132 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hcc_jcopo_msk_2023">Hepatocellular Carcinoma (MSK, JCO Precis Oncol 2023)</a> <em>90 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=sarcoma_msk_2023">Sarcoma (MSK, J Pathol 2023)</a> <em>82 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=egc_trap_ccr_msk_2023">Esophagogastric Cancer (MSK, Clin Cancer Res 2023)</a> <em>64 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=soft_tissue_msk_2023">Rhabdomyosarcomas (MSK, NPJ Precis Oncol 2023)</a> <em>42 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rms_msk_2023">Pediatric Rhabdomyosarcomas (MSK, JCO Precis Oncol 2023)</a> <em>24 samples</em></li>
</ul>
</li>
<li><p><strong>Gene Tables Update:</strong> Updated tables of genes (main and alias), based on <a href="http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/">Oct 1, 2023 HGNC release</a>. See seedDB release notes <a href="https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database-schema-2131">here</a> for details.</p>
</li>
</ul>
<doc-anchor-target id="dec-5-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#dec-5-2023">#</doc-anchor-trigger>
        <span>Dec 5, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New feature</strong>: Filter by data types on the homepage:</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/1334004/cf6d5486-bfd7-4f69-add0-442dbb4d354a" alt="image" /></p>
</li>
<li><p><strong>New feature</strong>: Change colors of tracks in Oncoprint:</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/301b0efb-3063-44db-bc4c-86973e7c566a" width="700" />
</li>
<li><p><strong>New feature</strong>: Show alteration frequencies per group in OncoPrint. <a href="https://bit.ly/3PwKsDJ">Example: RTK-RAS alterations in Smokers vs Never Smokers Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort</a></p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/f2a9576e-c06d-447a-8eac-f0292bd1f935" width="700" />
</li>
</ul>
<doc-anchor-target id="oct-17-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oct-17-2023">#</doc-anchor-trigger>
        <span>Oct 17, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New feature</strong>: Survival charts with landmark events and hazard ratios. <a href="https://bit.ly/48QktxS">Example: TP53 in Lung Cancer Cases in MSK-IMPACT Clinical Sequencing Cohort</a></p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/dc97c3ae-ebd7-4cd3-bb76-165517c9d2ba" width="700" />
</li>
</ul>
<doc-anchor-target id="oct-3-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oct-3-2023">#</doc-anchor-trigger>
        <span>Oct 3, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New feature</strong>: New Structural Variants Tab on <em>Results View</em>. <a href="https://bit.ly/49FDhRk">Example: TMPRSS2 Structural Variants in MSK-IMPACT Clinical Sequencing Cohort</a></p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/141e2817-3300-4af2-a845-005403d75e74" width="700" />
</li>
</ul>
<doc-anchor-target id="sep-5-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#sep-5-2023">#</doc-anchor-trigger>
        <span>Sep 5, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New feature</strong>:  Add gene-specific CNA charts to show all levels of copy number alterations (including gain and hetloss) on <em>Study View</em>:</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/873401ad-cbcf-4b07-9b80-8487af5abb18" width="700" />
</li>
</ul>
<doc-anchor-target id="aug-30-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-30-2023">#</doc-anchor-trigger>
        <span>Aug 30, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Exclude a patient from your selection directly from the <em>Patient View</em>:</p>
<p><img src="https://github.com/cBioPortal/cbioportal/assets/1334004/501fe998-8408-48e3-a183-e27ab3ff3f28" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="aug-21-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-21-2023">#</doc-anchor-trigger>
        <span>Aug 21, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 4,488 samples from 7 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_mskcc_2023_met_organotropism">Lung Adenocarcinoma Met Organotropism (MSK, Cancer Cell 2023)</a> <em>2653 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=aml_ohsu_2022">Acute Myeloid Leukemia (OHSU, Cancer Cell 2022)</a> <em>942 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coad_silu_2022">Colon Cancer (Sidra-LUMC AC-ICAM, Nat Med 2023)</a> <em>348 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nbl_msk_2023">Pediatric Neuroblastoma (MSK, Nat Genet 2023)</a> <em>223 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bowel_colitis_msk_2022">Colorectal Adenocarcinoma (MSK, Nat Commun 2022)</a> <em>180 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bladder_columbia_msk_2018">Bladder Cancer (Columbia University/MSK, Cell 2018)</a> <em>130 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=stmyec_wcm_2022">Myoepithelial Carcinomas of Soft Tissue (WCM, CSH Molecular Case Studies 2022)</a> <em>12 samples</em></li>
</ul>
</li>
<li><p><strong>Gene Tables Update:</strong> Updated tables of genes (main and alias), based on <a href="http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/">Apr 1, 2023 HGNC release</a>. See seedDB release notes <a href="https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database">here</a> for details.</p>
</li>
</ul>
<doc-anchor-target id="aug-1-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-1-2023">#</doc-anchor-trigger>
        <span>Aug 1, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: One-sided Fisher&#x27;s exact tests were changed to be two-sided. The affected pages are:</p>
<ul>
<li>Results View Page - Mutual Exclusivity Tab</li>
<li>Results View Page - Comparison Tab - Genomic Alterations Tab</li>
<li>Comparison Page - Genomic Alterations Tab</li>
<li>Comparison Page - Mutations Tab</li>
</ul>
<p>Please note that the Mutations tab on the Comparison page is a recent feature and was introduced with the two-sided Fisher&#x27;s exact test already implemented.</p>
<p>Several users pointed out that using a one-sided test was incorrect for these comparisons. Please see discussions <a href="https://github.com/cBioPortal/cbioportal/issues/9943">here</a> for more information.</p>
</li>
</ul>
<doc-anchor-target id="jul-18-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jul-18-2023">#</doc-anchor-trigger>
        <span>Jul 18, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add mutations table and two-sided exact p-value to comparison. <a href="https://www.cbioportal.org/comparison/mutations?sessionId=5cf89323e4b0ab413787436c&amp;groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D&amp;selectedGene=AR">Example: AR mutations in Primary vs Metastatic Prostate Cancer samples in MSK-IMPACT Clinical Sequencing Cohort</a></p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/da0cd1ef-b9b4-49f0-a9d3-42cff0204dc2" width="700" />
</li>
</ul>
<doc-anchor-target id="may-2-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-2-2023">#</doc-anchor-trigger>
        <span>May 2, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: The mutations tab now shows variant annotations from the <a href="https://www.cancerrevue.org/">repository of Variant with Unexpected Effects (reVUE)</a>.</p>
<img src="https://user-images.githubusercontent.com/16869603/234271379-6a0876ea-8477-426e-a0e4-5e0b0b66aef9.gif" width="700" />
</li>
</ul>
<doc-anchor-target id="apr-25-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-25-2023">#</doc-anchor-trigger>
        <span>Apr 25, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Local Installations Feature</strong>: When custom driver annotations for structural variants are loaded, one can now filter by them in both the Oncoprint and the <em>Study View</em>.</li>
</ul>
<doc-anchor-target id="apr-11-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-11-2023">#</doc-anchor-trigger>
        <span>Apr 11, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Disable autocommit and manually commit filters in study view. Manually commit filters can improve cBioPortal performance when query large dataset.</p>
<img src="https://user-images.githubusercontent.com/3323006/221857599-1eb6e9a5-55c7-4fb5-bc14-283699be0990.png" width="700" />
<img src="https://user-images.githubusercontent.com/3323006/221858244-bf359850-79b5-4f9c-8008-6b5271721854.png" width="700" />
</li>
</ul>
<doc-anchor-target id="apr-5-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-5-2023">#</doc-anchor-trigger>
        <span>Apr 5, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 2,472 samples from 5 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=paired_bladder_2022">Bladder Cancer (MSK, Cell Reports 2022)</a> <em>1659 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gist_msk_2022">Gastrointestinal Stromal Tumor (MSK, NPJ Precis Oncol 2023)</a> <em>499 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=appendiceal_msk_2022">Appendiceal Cancer (MSK, J Clin Oncol 2022)</a> <em>273 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coadread_mskresistance_2022">Colorectal Cancer (MSK, Cancer Discovery 2022)</a> <em>22 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nst_nfosi_ntap">Nerve Sheath Tumors (Johns Hopkins, Sci Data 2020)</a> <em>19 samples</em> [<strong>First GRCh38 Study</strong>]</li>
</ul>
</li>
<li><p><strong>Data Improvement</strong></p>
<ul>
<li>Added TERT promoter mutation status to <a href="https://www.cbioportal.org/study/summary?id=skcm_tcga_pub_2015">Melanomas (TCGA, Cell 2015)</a>, <a href="https://www.cbioportal.org/study/summary?id=thca_tcga_pub">Papillary Thyroid Carcinoma (TCGA, Cell 2014)</a> TCGA studies.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="apr-4-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-4-2023">#</doc-anchor-trigger>
        <span>Apr 4, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Allow numeric data type for custom data charts.</p>
<img src="https://user-images.githubusercontent.com/33220871/222459360-9c251f7f-1005-442f-811d-dd1ee1103161.png" width="700" />
<img src="https://user-images.githubusercontent.com/33220871/222460439-ac7007c8-059a-4ecc-816c-90752adff444.png" width="700" />
<p>This also allows to have numerical custom data after we query based on genes (custom data 2 in the image):</p>
<img src="https://user-images.githubusercontent.com/33220871/223729410-03a094f4-7671-453f-b8b8-f7bc10b61c7c.png" width="700" />
<img src="https://user-images.githubusercontent.com/33220871/223729522-fd9e19e2-6218-40d4-b9fb-f253c4faa858.png" width="700" />
</li>
</ul>
<doc-anchor-target id="jan-10-2023">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jan-10-2023">#</doc-anchor-trigger>
        <span>Jan 10, 2023</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: New Pathways tab on the Group Comparison view. <a href="https://www.cbioportal.org/comparison/pathways?comparisonId=61791fa92e3200072db9b64e&amp;groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D">Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort</a></p>
<img src="https://user-images.githubusercontent.com/15748980/235750129-661682f7-6593-4b7e-8873-ebba70eeae32.png" width="700" />
</li>
</ul>
<doc-anchor-target id="dec-13-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#dec-13-2022">#</doc-anchor-trigger>
        <span>Dec 13, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: New Mutations tab on the Group Comparison view. <a href="https://www.cbioportal.org/comparison/mutations?comparisonId=61791fa92e3200072db9b64e&amp;groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D">Example: Primary vs Metastasis samples in MSK-IMPACT Clinical Sequencing Cohort</a></p>
<img src="https://user-images.githubusercontent.com/15748980/235748831-78652ccd-b3ab-496f-9626-74ceb8c7230d.png" width="700" />
</li>
</ul>
<doc-anchor-target id="oct-12-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oct-12-2022">#</doc-anchor-trigger>
        <span>Oct 12, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 1,459 samples from 10 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=hcc_meric_2021">Hepatocellular Carcinoma (MERiC/Basel, Nat Commun. 2022)</a> <em>122 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prostate_pcbm_swiss_2019">Prostate Cancer Brain Metastases (Bern, Nat Commun. 2022)</a> <em>168 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_impact_subset_2022">Pan-Cancer MSK-IMPACT MET Validation Cohort (MSK 2022)</a> <em>69 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_ccr_cfdna_msk_2022">Endometrial Carcinoma cfDNA (MSK, Clin Cancer Res 2022)</a> <em>44 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_ccr_msk_2022">Endometrial Carcinoma MSI (MSK, Clin Cancer Res 2022)</a> <em>181 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbc_mskcc_2022">Gallbladder Cancer (MSK, Clin Cancer Res, 2022)</a> <em>244 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mng_utoronto_2021">Meningioma (University of Toronto, Nature 2021)</a> <em>121 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_selpercatinib_2020">Mixed Tumors: Selpercatinib RET Trial (MSK, Nat Commun. 2022)</a> <em>188 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lgsoc_mapk_msk_2022">Low-Grade Serous Ovarian Cancer (MSK, Clin Cancer Res 2022)</a> <em>119 samples</em></li>
<li><a href="http://www.cbioportal.org/study/summary?id=blca_bcan_hcrn_2022">Urothelial Carcinoma (BCAN/HCRN 2022)</a> <em>203 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="sep-6-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#sep-6-2022">#</doc-anchor-trigger>
        <span>Sep 6, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Oncoprint can now save clinical tracks after login:</p>
<img src="https://user-images.githubusercontent.com/3323006/188828765-ded10947-7343-49dd-b605-50b58b67a388.png" width="700" />
</li>
</ul>
<doc-anchor-target id="aug-11-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-11-2022">#</doc-anchor-trigger>
        <span>Aug 11, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Major Release</strong>: <strong>v5.0.0</strong> release drops support for fusions in the mutation data format. Going forward fusions can only be imported in the Structural Variant (SV) format. This is mainly a refactoring effort to simplify the codebase and pave the way for the development of novel structural variant visualizations in the future. For cBioPortal instance maintainer, please reference our <a href="https://docs.cbioportal.org/migration-guide/#v4---v5">Migration Guide</a> for instruction.</li>
</ul>
<doc-anchor-target id="jul-26-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jul-26-2022">#</doc-anchor-trigger>
        <span>Jul 26, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 6,631 samples from 7 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=biliary_tract_summit_2022">Metastatic Biliary Tract Cancers (SUMMIT - Neratinib Basket Trial, 2022)</a> <em>36 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rectal_msk_2022">Rectal Cancer (MSK, Nature Medicine 2022)</a> <em>801 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lung_msk_mind_2020">Lung Adenocarcinoma (MSK Mind,Nature Cancer 2022)</a> <em>247 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mds_iwg_2022">Myelodysplastic Syndromes (MDS IWG, IPSSM, NEJM Evidence 2022)</a> <em>3,323 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=egc_msk_tp53_ccr_2022">Esophagogastric Cancer (MSK, Clin Cancer Res 2022)</a> <em>237 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pog570_bcgsc_2020">Pan-cancer Analysis of Advanced and Metastatic Tumors (BCGSC, Nature Cancer 2020)</a> <em>570 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_pik3r1_msk_2021">Prostate Adenocarcinoma (MSK, Clin Cancer Res. 2022)</a> <em>1,417 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="jun-7-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jun-7-2022">#</doc-anchor-trigger>
        <span>Jun 7, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add heatmap to plot options on <em>Comparison Page</em>. <a href="https://www.cbioportal.org/comparison/clinical?sessionId=5cf89323e4b0ab413787436c">Example: Primary vs Metastatic Prostate Cancer in MSK-IMPACT (2017)</a></p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/1a048863-7db6-43c6-8250-c31039a2d67f" width="700" />
</li>
</ul>
<doc-anchor-target id="may-31-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-31-2022">#</doc-anchor-trigger>
        <span>May 31, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Added Quartiles, Median split and Generate bins options for bar charts on the study view page, where Generate bins allows user to define bin size and min value</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/290e6d45-1c10-4a51-a823-24db27063486" width="700" />
</li>
</ul>
<doc-anchor-target id="may-12-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-12-2022">#</doc-anchor-trigger>
        <span>May 12, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Show cohort alteration frequencies in pathways from <a href="https://www.ndexbio.org/iquery/">NDEx</a> on the <em>Results View</em>. <a href="https://bit.ly/3sE4UqD">Example: Glioblastoma signaling pathways in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/168684594-fafd1fd6-b2a9-45d7-ad3a-d5611f3f469c.png" width="700" />
</li>
</ul>
<doc-anchor-target id="may-24-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-24-2022">#</doc-anchor-trigger>
        <span>May 24, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: Add Help buttons on various pages and tabs, including the homepage</li>
</ul>
<doc-anchor-target id="may-10-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-10-2022">#</doc-anchor-trigger>
        <span>May 10, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Use IGV for the genomic overview on the <em>Patient View</em>. <a href="https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&amp;caseId=TCGA-BK-A0CC">Example: Endometrial cancer patient in TCGA</a></p>
<img src="https://github.com/cBioPortal/cbioportal/assets/1334004/6fa6ab4b-4c09-4bd5-a7c2-3be2f418d83a" width="700" /> 
</li>
</ul>
<doc-anchor-target id="may-5-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-5-2022">#</doc-anchor-trigger>
        <span>May 5, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: View mutations and copy number changes in the <a href="https://igv.org/">Integrative Genomics Viewer (IGV)</a> on the <em>Patient View</em>. <a href="https://www.cbioportal.org/patient?studyId=ucec_tcga_pub&amp;caseId=TCGA-BK-A0CC">Example: Endometrial cancer patient in TCGA</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/168593020-253a4bfe-3922-4bf9-8dcc-89e298ba87c7.png" width="700" />
</li>
<li><p><strong>New Feature</strong>: Add charts that plot categorical vs continuous data on the <em>Study View</em>. <a href="https://www.cbioportal.org/study/summary?id=msk_impact_2017">Example: MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/168595352-6416db7e-b6b5-4ef8-95e7-eaf67c841d94.png" width="700" />
</li>
<li><p><strong>New Feature</strong>: Several single cell data integrations are now available for the <a href="https://www.cbioportal.org/study/summary?id=gbm_cptac_2021">CPTAC glioblastoma study</a>, allowing one to:</p>
<ul>
<li>Compare genomic alterations and cell type fractions in oncoprints on the <em>Results View</em> (<a href="https://bit.ly/37QiC2x">Example</a>)</li>
<li>Explore the single cell data further in <a href="http://vitessce.io/">Vitessce</a> on the <em>Patient View</em> (<a href="https://www.cbioportal.org/patient/customTab1?studyId=gbm_cptac_2021&amp;caseId=C3N-02783">Example</a>)</li>
<li>Create cohorts and groups based on cell type fractions on the <em>Study View</em> (<a href="https://tinyurl.com/singlecellbioportal">Example</a>)</li>
<li>Compare differences in cell type fractions between groups on the <em>Comparison Page</em> (<a href="https://www.cbioportal.org/comparison/generic_assay_single_cell_type_fractions?comparisonId=6282d44d0934121b56df47fb&amp;unselectedGroups=%5B%220.05-0.06%22%2C%220.03-0.05%22%5D">Example</a>)</li>
</ul>
  <img src="https://user-images.githubusercontent.com/1334004/168699333-a50ef433-5efe-45cc-a477-a3a99ed419dc.png" width="700" />
</li>
</ul>
<doc-anchor-target id="apr-20-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#apr-20-2022">#</doc-anchor-trigger>
        <span>Apr 20, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 2,557 samples from 5 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_hta9_htan_2022">Breast Cancer (HTAN OHSU, 2022)</a> <em>5 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=crc_dd_2022">Colorectal Cancer (MSK, 2022)</a> <em>47 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_kunga_msk_2022">Pediatric Pancan Tumors (MSK, 2022)</a> <em>135 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=sarcoma_mskcc_2022">Sarcoma (MSK, 2022)</a> <em>2,138 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lung_nci_2022">Lung Cancer in Never Smokers (NCI, Nature Genetics 2021)</a> <em>232 samples</em></li>
</ul>
</li>
<li><p><strong>Gene Tables Update</strong> Updated tables of genes (main and alias), based on <a href="http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/archive/monthly/tsv/">Jan 1, 2022 HGNC release</a>. See seedDB release note <a href="https://github.com/cBioPortal/datahub/tree/master/seedDB#latest-seed-database">here</a> for details.</p>
</li>
<li><p><strong>Data Improvement</strong></p>
<ul>
<li><p>Pan-can studies timeline addition: <code v-pre>TREATMENT</code>, <code v-pre>OTHER MALIGNANCY FORM</code>, <code v-pre>SAMPLE ACQUISITION</code>, <code v-pre>STATUS</code> are added to all 32 TCGA Pan-Can studies. Details for data source and transformation process can be found <a href="https://github.com/cBioPortal/datahub/pull/1597">here</a> or in the <a href="https://github.com/cBioPortal/datahub/blob/master/public/acc_tcga_pan_can_atlas_2018/README.md">README.md</a> files included in each study folder on datahub. <a href="https://www.cbioportal.org/patient?studyId=brca_tcga_pan_can_atlas_2018&amp;caseId=TCGA-A2-A04P">Example: patient view of TCGA-A2-A04P in Breast Invasive Carcinoma Tumor Type</a><img width="1204" alt="Screen Shot 2022-04-25 at 11 40 51 AM" src="https://user-images.githubusercontent.com/5973438/165153293-3d400e03-48d5-4919-ae32-7f16fa7d2f4f.png"></p>
</li>
<li><p>Pan-can studies methylation addition: methylation profile (27k and 450k merged) are added to all 32 TCGA Pan-Can studies, in generic assay format. Data source: <a href="https://gdc.cancer.gov/node/977">GDC</a>. Example: search by gene or probe from dropdown, to add a chart in study view, a track in Oncoprint (single study query only), or plots in plots tab. <img width="860" alt="Screen Shot 2022-03-22 at 9 52 20 PM" src="https://user-images.githubusercontent.com/5973438/159626976-72601dc9-e5be-4b50-bae9-dedb5be6e7fc.png"><img width="340" alt="Screen Shot 2022-03-22 at 9 52 27 PM" src="https://user-images.githubusercontent.com/5973438/159626985-2423c556-2180-4a00-b9b8-69b4d2e361ba.png"><img width="430" alt="Screen Shot 2022-03-22 at 9 53 38 PM" src="https://user-images.githubusercontent.com/5973438/159626987-e6b51bef-814e-4a76-826a-211fcdd5e9a1.png"></p>
</li>
<li><p>Single cell (type fraction and phases) data (in generic assay format) is added to <a href="https://www.cbioportal.org/study/summary?id=gbm_cptac_2021">Glioblastoma (CPTAC, Cell 2021)</a></p>
</li>
</ul>
</li>
<li><p><strong>New Feature</strong> For the new HTAN OHSU study there is now also an integration with <a href="https://labsyspharm.org/">Harvard LSP</a>&#x27;s <a href="https://www.cycif.org/software/minerva">Minerva</a> for exploring multiplex imaging:</p>
<img src="https://github.com/cBioPortal/cbioportal/assets/1334004/aa0f0c99-4b9c-481d-9f7f-891ea03fdd34" width="700" />
</li>
</ul>
<doc-anchor-target id="mar-1-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#mar-1-2022">#</doc-anchor-trigger>
        <span>Mar 1, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Documentation</strong>:
<ul>
<li>Add a <a href="https://docs.cbioportal.org/user-guide/">User Guide</a></li>
<li>Add <a href="https://docs.cbioportal.org/user-guide/overview/#how-to-videos">Study View and Query How-to videos</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="feb-8-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#feb-8-2022">#</doc-anchor-trigger>
        <span>Feb 8, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Create X vs Y violin plots in <em>Study View</em> using any categorical and numerical clinical data:</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/94d7dd21-4415-405d-8e9b-46e1de3dc095" width="700" />
</li>
</ul>
<doc-anchor-target id="jan-19-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jan-19-2022">#</doc-anchor-trigger>
        <span>Jan 19, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Numerical filters on the <em>Study View</em> are now editable:</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/00926fec-f32b-48c7-a85f-fced7c29070f" width="700" />
</li>
<li><p><strong>New Feature</strong>: In the annotation column choose between showing a single icon OncoKB icon or multiple (one for therapeutic, diagnostic and prognostic):</p>
  <img src="https://github.com/cBioPortal/cbioportal/assets/1334004/87c5f7d6-10d7-4631-8b37-fa8cef4cabe2" width="700" />
</li>
</ul>
<doc-anchor-target id="jan-10-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jan-10-2022">#</doc-anchor-trigger>
        <span>Jan 10, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Documentation for Local cBioPortal Installations</strong>:
<ul>
<li>Explain how to use <a href="https://docs.cbioportal.org/file-formats/#create-the-cbioportal-mutation-data-file-with-genome-nexus-with-a-minimal-maf-file">Genome Nexus to annotate MAF files</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="jan-4-2022">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jan-4-2022">#</doc-anchor-trigger>
        <span>Jan 4, 2022</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 27,447 samples from 10 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_cptac_2020">Endometrial Carcinoma (CPTAC, Cell 2020)</a> <em>95 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=paad_cptac_2021">Pancreatic Ductal Adenocarcinoma (CPTAC, Cell 2021)</a> <em>140 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lusc_cptac_2021">Lung Squamous Cell Carcinoma (CPTAC, Cell 2021)</a> <em>108 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_cptac_2020">Lung Adenocarcinoma (CPTAC, Cell 2020)</a> <em>110 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbm_cptac_2021">Glioblastoma (CPTAC, Cell 2021)</a> <em>99 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_cptac_2020">Breast Cancer (CPTAC, Cell 2020)</a> <em>122 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brain_cptac_2020">Pediatric Brain Cancer (CPTAC/CHOP, Cell 2020)</a> <em>218 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mpcproject_broad_2021">Metastatic Prostate Cancer (Provisional, June 2021)</a> <em>123 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_met_2021">MSK MetTropism (MSK, Cell 2021)</a> <em>25,775 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_ch_ped_2021">Cancer Therapy and Clonal Hematopoiesis (MSK, 2021)</a> <em>657 samples</em></li>
</ul>
</li>
<li><p><strong>Added TMB (nonsynonymous) scores</strong> for all studies. <a href="https://www.cbioportal.org/study/clinicalData?id=gbm_cptac_2021">Example: new TMB field for study gbm_cptac_2021</a><br />
(Details for the calculation can be found <a href="https://github.com/cBioPortal/datahub-study-curation-tools/blob/master/tmb/calculate_tmb/Readme.md">HERE</a>)<br />
<img width="361" alt="Screen Shot 2022-01-04 at 11 39 20 PM" src="https://user-images.githubusercontent.com/5973438/148133673-a23506e9-7957-4581-99f0-820ca9548672.png"></p>
</li>
</ul>
<doc-anchor-target id="nov-12-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#nov-12-2021">#</doc-anchor-trigger>
        <span>Nov 12, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 3,680 samples from 6 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_mapk_hp_msk_2021">Breast Cancer MAPK (MSKCC, Nat Commun 2021)</a> <em>145 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=crc_nigerian_2020">Colorectal Cancer (MSK, 2020)</a> <em>64 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_pareja_msk_2020">Breast Cancer (MSK, Clinical Cancer Res 2020)</a> <em>60 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hgsoc_msk_2021">High-Grade Serous Ovarian Cancer (MSK, 2021)</a> <em>45 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=difg_glass_2019">Diffuse Glioma (GLASS Consortium, Nature 2019)</a> <em>444 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pancan_pcawg_2020">Pan-cancer analysis of whole genomes (ICGC/TCGA, Nature 2020)</a> <em>2,922 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="nov-32021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#nov-32021">#</doc-anchor-trigger>
        <span>Nov 3,2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add Uniprot topology as a new annotation track on the Mutations Tab of the Results View. <a href="https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;profileFilter=mutations%2Cfusion%2Ccna&amp;case_set_id=msk_impact_2017_cnaseq&amp;gene_list=EGFR&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: EGFR in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/16869603/141861719-f91d787e-e8c3-4c27-8273-7642ec3fd5e1.png" width="700" />
</li>
</ul>
<doc-anchor-target id="oct-1-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oct-1-2021">#</doc-anchor-trigger>
        <span>Oct 1, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Arm level Copy Number events are now loaded into cBioPortal using the Categorial Generic Assay Data Type. They can be found in a tab under the Add Charts Button of the <em>Study View</em> <a href="https://www.cbioportal.org/study/summary?id=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018">Example: Arm Level Data in TCGA PanCancer Atlas</a></p>
   <img src="https://user-images.githubusercontent.com/1334004/139692484-8245d94f-1821-4476-974e-b6c1a6cd8d12.png" width="700" />
</li>
</ul>
<doc-anchor-target id="sep-22-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#sep-22-2021">#</doc-anchor-trigger>
        <span>Sep 22, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 14,844 samples from 7 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=crc_apc_impact_2020">Colorectal Cancer (MSK, Gastroenterology 2020)</a> <em>471 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=breast_ink4_msk_2021">Metastatic Breast Cancer (MSK, Cancer Discovery 2021)</a> <em>1,365 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_mskimpact_2021">Lung Adenocarcinoma (MSKCC, 2021)</a> <em>186 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_msk_stopsack_2021">Race Differences in Prostate Cancer (MSK, 2021) </a> <em>2,069 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mbl_dkfz_2017">Medulloblastoma (DKFZ, Nature 2017)</a> <em>491 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lung_pdx_msk_2021">Thoracic Cancer (MSK, 2021)</a> <em>68 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pan_origimed_2020">China Pan-cancer (OrigiMed, 2020)</a> <em>10,194 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="sep-21-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#sep-21-2021">#</doc-anchor-trigger>
        <span>Sep 21, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Dowloading the Lollipop plot on the Mutations Tab of the <em>Results View</em> will now also include the annotation tracks:</p>
  <img src="https://user-images.githubusercontent.com/1334004/139694195-51ad5580-efa5-4b25-9a10-2754726a246b.png" width="700" />
</li>
</ul>
<doc-anchor-target id="aug-17-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-17-2021">#</doc-anchor-trigger>
        <span>Aug 17, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: The Mutations Tab of the <em>Results View</em> can now show exon numbers as an annotation track <a href="https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;cancer_study_list=msk_impact_2017&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;profileFilter=mutations%2Cfusion%2Ccna&amp;case_set_id=msk_impact_2017_cnaseq&amp;gene_list=MET&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: MET Exon 14 Mutations in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139694614-62f62640-2fab-4117-ae09-db2281690f76.png" width="700" />
</li>
</ul>
<doc-anchor-target id="aug-10-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#aug-10-2021">#</doc-anchor-trigger>
        <span>Aug 10, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Use the filtering capabilities in the Mutations Tab of the <em>Results View</em> to create a custom cohort that one can open directly in the <em>Study View</em> <a href="https://bit.ly/3w6wu00">Example: CTNNB1 in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139696976-6e145c05-10d3-44c9-8499-146ad6e13f2f.png" width="700" />
</li>
</ul>
<doc-anchor-target id="jul-27-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jul-27-2021">#</doc-anchor-trigger>
        <span>Jul 27, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add a custom filter to any column of the Mutations Tab in the <em>Results View</em> <a href="https://bit.ly/3w6wu00">Example: CTNNB1 in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139697555-34507fa8-2b57-4aa8-84ee-c477efffa864.png" width="700" />
</li>
<li><p><strong>New Feature</strong>: Show detailed descriptions for each annotation source in the header of the the Mutations Table in both the <em>Results View</em> and the <em>Patient View</em> <a href="https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&amp;caseId=P04">Example link</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139699259-1be86599-c2e0-457c-b257-1b0d446facf4.png" width="350" />
</li>
</ul>
<doc-anchor-target id="jul-6-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#jul-6-2021">#</doc-anchor-trigger>
        <span>Jul 6, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add any clinical data as a column on the Mutations Tab in the <em>Results View</em> <a href="https://bit.ly/3bxFlyp">Example: EGFR in MSK-IMPACT (2017) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139700811-4b6ce432-4abf-4dc4-8371-634a14d2b603.png" width="700" />
</li>
</ul>
<doc-anchor-target id="june-23-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-23-2021">#</doc-anchor-trigger>
        <span>June 23, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 1,084 samples from 5 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ihch_msk_2021">Intrahepatic Cholangiocarcinoma (MSK, Hepatology 2021)</a> <em>412 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ihch_ismms_2015">Intrahepatic Cholangiocarcinoma (Mount Sinai 2015)</a> <em>8 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_msk_tcga_2021">RAD51B Associated Mixed Cancers (Mandelker 2021</a> <em>19 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ihch_mskcc_2020">Intrahepatic Cholangiocarcinoma (MSK, 2020)</a> <em>219 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_msk_npjpo_2021">Lung Adenocarcinoma (NPJ Precision Oncology, MSK 2021)</a> <em>426 samples</em></li>
</ul>
</li>
<li><strong>Added mass-spec proteome data</strong> from CPTAC to <a href="https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018">Breast Invasive Carcinoma (TCGA, PanCancer Atlas)</a>, <a href="https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018">Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)</a> and <a href="https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018">Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)</a>.</li>
<li><strong>Added mass-spec phosphoproteome site level expression</strong> from CPTAC to <a href="https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018">Breast Invasive Carcinoma (TCGA, PanCancer Atlas)</a> and <a href="https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018">Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)</a>.</li>
<li><strong>Updated gene tables</strong> Updated tables of genes (main and alias), based on <a href="https://www.genenames.org/">HGNC</a>. See details <a href="https://github.com/cBioPortal/datahub/tree/master/seedDB#seed-database-schema-2128">HERE</a> in section <code v-pre>Contents of seed database</code>. Sripts/resources/process used to construct new tables are described <a href="https://github.com/cBioPortal/datahub-study-curation-tools/tree/master/gene-table-update/build-input-for-importer">HERE</a>.</li>
</ul>
<doc-anchor-target id="june-1-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-1-2021">#</doc-anchor-trigger>
        <span>June 1, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: In certain studies where we have the data we show read counts for uncalled mutations on the <em>Patient View</em> <a href="https://www.cbioportal.org/patient?studyId=glioma_msk_2018&amp;caseId=IM-GBM-1">Example: A patient in the Glioma (MSK, 2019) cohort</a></p>
  <img src="https://user-images.githubusercontent.com/1334004/139701937-91a40335-e95a-45f4-9800-356737a17fab.png" width="700" />
</li>
</ul>
<doc-anchor-target id="may-10-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-10-2021">#</doc-anchor-trigger>
        <span>May 10, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Pick color for User Defined Groups <a href="https://www.cbioportal.org/study/summary?id=msk_impact_2017">Example: Color Bladder Cancer Group in MSK-IMPACT (2017) cohort</a>, implemented by <a href="https://www.thehyve.nl/services/cbioportal">The Hyve</a>.</p>
 <img src="https://user-images.githubusercontent.com/1334004/122818654-7b018280-d2a7-11eb-94c1-834b780618b9.png" width="700" />
</li>
</ul>
<doc-anchor-target id="may-4-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-4-2021">#</doc-anchor-trigger>
        <span>May 4, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add more categories of mutations to the Mutations Tab on the <em>Results View</em>, including Driver/VUS, Splice and Structural Variants <a href="https://www.cbioportal.org/results/mutations?cancer_study_list=msk_impact_2017&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;profileFilter=mutations%2Cfusion%2Ccna&amp;case_set_id=msk_impact_2017_cnaseq&amp;gene_list=TP53&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: TP53 alterations in the MSK-IMPACT (2017) cohort</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122817171-99ff1500-d2a5-11eb-8a58-374c9ea32013.png" width="700" />
</li>
</ul>
<doc-anchor-target id="april-21-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-21-2021">#</doc-anchor-trigger>
        <span>April 21, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 4074 samples from 9 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=mbc_msk_2021">Metaplastic Breast Cancer (MSK, 2021)</a> <em>19 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_mskcc_2020">Lung Adenocarcinoma (MSKCC, 2020)</a> <em>604 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cscc_ucsf_2021">Cutaneous Squamous Cell Carcinoma (UCSF, 2021)</a> <em>105 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_access_2021">MSK-IMPACT and MSK-ACCESS Mixed Cohort (MSK, 2021)</a> <em>1446 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mel_mskimpact_2020">Melanoma (MSKCC, 2018)</a> <em>720 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=chol_icgc_2017">Cholangiocarcinoma (ICGC, Cancer Discov 2017)</a> <em>489 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=egc_mskcc_2020">Esophageal/Stomach Cancer (MSK, 2020)</a> <em>487 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rbl_mskcc_2020">Retinoblastoma (MSK, Cancers 2021)</a> <em>83 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hccihch_pku_2019">Combined Hepatocellular and Intrahepatic Cholangiocarcinoma (Peking University, Cancer Cell 2019)</a> <em>121 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="april-20-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-20-2021">#</doc-anchor-trigger>
        <span>April 20, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add driver annotations to download tab on <em>Results View</em> <a href="https://www.cbioportal.org/results/download?cancer_study_list=coadread_tcga_pub&amp;cancer_study_id=coadread_tcga_pub&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&amp;Z_SCORE_THRESHOLD=2.0&amp;case_set_id=coadread_tcga_pub_nonhypermut&amp;gene_list=KRAS%20NRAS%20BRAF&amp;gene_set_choice=user-defined-list">Example: RAS/RAF alterations in colorectal cancer</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122816424-b2bafb00-d2a4-11eb-8a0f-5e878bb70ccd.png" width="700" />
</li>
</ul>
<doc-anchor-target id="march-30-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-30-2021">#</doc-anchor-trigger>
        <span>March 30, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Add 95% Confidence Interval for Survival Plots <a href="https://www.cbioportal.org/results/comparison?plots_horz_selection=%7B%7D&amp;plots_vert_selection=%7B%7D&amp;plots_coloring_selection=%7B%7D&amp;tab_index=tab_visualize&amp;Action=Submit&amp;session_id=60d0d9fce4b015b63e9f1116&amp;comparison_subtab=survival">Example: Altered vs Unaltered EGFR in Lung Cancer</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122810147-0295c400-d29d-11eb-9be7-fa14787d72b7.png" width="700" />
</li>
</ul>
<doc-anchor-target id="march-11-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-11-2021">#</doc-anchor-trigger>
        <span>March 11, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Combine different types of alterations in <em>Comparison View</em> <a href="https://www.cbioportal.org/comparison/alterations?comparisonId=5f73ac54e4b0242bd5d22ab0&amp;selectedEnrichmentEventTypes=%5B%22truncating%22%2C%22feature_truncation%22%2C%22nonsense_mutation%22%2C%22nonsense%22%2C%22stopgain_snv%22%2C%22stop_gained%22%2C%22frameshift%22%2C%22frameshift_variant%22%2C%22frame_shift_del%22%2C%22frameshift_deletion%22%2C%22frame_shift_ins%22%2C%22frameshift_insertion%22%2C%22translation_start_site%22%2C%22initiator_codon_variant%22%2C%22start_codon_snp%22%2C%22start_codon_del%22%2C%22de_novo_start_outofframe%22%2C%22nonstop_mutation%22%2C%22stop_lost%22%2C%22splice_site%22%2C%22splice%22%2C%22splicing%22%2C%22splice_site_snp%22%2C%22splice_site_del%22%2C%22splice_site_indel%22%2C%22splice_region_variant%22%2C%22splice_region%22%2C%22HOMDEL%22%5D&amp;groupOrder=%5B%22Primary%22%2C%22Metastasis%22%5D">Example: Deletions and Truncating events in primary vs metastases</a> or read more on <a href="https://www.thehyve.nl/articles/find-differentially-altered-genes-by-combining-alteration-types">The Hyve&#x27;s blog</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122808698-35d75380-d29b-11eb-8928-0fca5f6e3b34.png" width="700" />
</li>
<li><p><strong>Enhancement</strong>: Improve UI for OncoPrint, aggregating various data modalities in a single add track dropdown button <a href="https://www.cbioportal.org/results/oncoprint?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ccle_broad_2019_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ccle_broad_2019_cna&amp;cancer_study_list=ccle_broad_2019&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;profileFilter=0&amp;case_set_id=ccle_broad_2019_cnaseq&amp;gene_list=TP53&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: Add clinical, heatmap and treatment response data into the OncoPrint</a></p>
<img src="https://user-images.githubusercontent.com/15748980/107722872-f32eba80-6cad-11eb-89d0-b6c5f300ab41.png" width="700" />
</li>
</ul>
<doc-anchor-target id="february-16-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-16-2021">#</doc-anchor-trigger>
        <span>February 16, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Show only TCGA PanCancer Atlas Pathways in <em>Results</em> and <em>Patient View</em> to avoid showing many similar pathways <a href="https://www.cbioportal.org/results/pathways?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;cancer_study_list=msk_impact_2017&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;profileFilter=mutations%2Cfusion%2Ccna&amp;case_set_id=msk_impact_2017_cnaseq&amp;gene_list=APC%2520CTNNB1&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: Clinvar APC and CTNNB1 alterations in WNT pathway</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122806359-3a4e3d00-d298-11eb-89fc-9e107a4651b4.png" width="700" />
</li>
</ul>
<doc-anchor-target id="january-28-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-28-2021">#</doc-anchor-trigger>
        <span>January 28, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Show ClinVar Interpretation in Mutation tables <a href="https://www.cbioportal.org/results/mutations?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;cancer_study_list=msk_impact_2017&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;profileFilter=0&amp;case_set_id=msk_impact_2017_cnaseq&amp;gene_list=BRCA2&amp;geneset_list=%20&amp;tab_index=tab_visualize&amp;Action=Submit">Example: Clinvar Interpretations in BRCA2</a></p>
<img src="https://user-images.githubusercontent.com/1334004/122805797-8a78cf80-d297-11eb-9454-fed37ed1ddb5.png" width="700" />
</li>
</ul>
<doc-anchor-target id="january-12-2021">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-12-2021">#</doc-anchor-trigger>
        <span>January 12, 2021</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add your own custom data for a sample or patient to use on the study or comparison view  <a href="https://www.cbioportal.org/study/summary?id=msk_impact_2017">Example: Add custom data to three samples and do a comparison</a></p>
<img src="https://user-images.githubusercontent.com/1334004/104651553-ce7efd00-5685-11eb-8017-c43f24421e3f.png" width="700">
</li>
<li><p><strong>New Feature</strong>: Show the mutations of a patient inside a pathway schematic using PathwayMapper  <a href="https://www.cbioportal.org/patient/pathways?studyId=msk_impact_2017&amp;caseId=P-0000377">Example: Notch signaling pathway in a prostate cancer patient</a></p>
<img src="https://user-images.githubusercontent.com/1334004/104649378-9f1ac100-5682-11eb-91eb-4e0f024a4ded.png" width="700">
</li>
<li><p><strong>New Feature</strong>: Display and compare generic assays, such as microbiome and treatment response, on the study view <a href="https://www.cbioportal.org/study?id=6000b6aae4b015b63e9d4d44">Example: Prasinovirus microbiome signatures in TCGA</a></p>
<img src="https://user-images.githubusercontent.com/1334004/104651211-54e70f00-5685-11eb-94f1-f20930793c28.png" width="700">
</li>
<li><p><strong>New Feature</strong>: The Plots tab on <em>Results View</em> now allows you to group alterations by Driver and VUS <a href="https://bit.ly/3ssY9Gu">Example: POLE driver mutations vs VUSs against mutation counts in TCGA Colorectal Adenocarcinoma</a>
<img src="https://user-images.githubusercontent.com/1334004/104768867-56bfd980-573c-11eb-8e03-5ad86d01f25a.png" width="700"></p>
</li>
</ul>
<doc-anchor-target id="december-31-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-31-2020">#</doc-anchor-trigger>
        <span>December 31, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 430 samples from 5 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_jup_msk_2020">Juvenile Papillomatosis and Breast Cancer (MSK, 2020)</a> <em>5 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mixed_cfdna_msk_2020">Mixed cfDNA (MSKCC, 2020)</a> <em>229 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mel_dfci_2019">Metastatic Melanoma (DFCI, Nature Medicine 2019)</a> <em>144 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lung_smc_2016">Lung Cancer (SMC, Cancer Research 2016)</a> <em>22 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=utuc_igbmc_2021">Upper Tract Urothelial Carcinoma (IGBMC, Genome Biology 2021)</a> <em>30 samples</em></li>
</ul>
</li>
<li><strong>Added survival data</strong> to <a href="https://www.cbioportal.org/study/summary?id=brca_metabric">Breast Cancer (METABRIC, Nature 2012 &amp; Nat Commun 2016)</a></li>
</ul>
<doc-anchor-target id="november-3-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-3-2020">#</doc-anchor-trigger>
        <span>November 3, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: The <a href="https://www.cbioportal.org/installations">map</a> of local installations of cBioPortal is available now. Please consider registering your instance <a href="http://bit.ly/2XW3wNG">here</a>.
<img src="https://user-images.githubusercontent.com/840895/99544685-96f41d00-2982-11eb-9de2-457a4456c294.png" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: upgraded the <em>Genomic Evolution</em> tab in <em>Patient View</em> with timeline <a href="https://www.cbioportal.org/patient/genomicEvolution?caseId=P04&amp;studyId=lgg_ucsf_2014">Example</a>
<img src="https://user-images.githubusercontent.com/840895/99543852-a3c44100-2981-11eb-8f53-14a0b51b767c.png" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="october-20-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-20-2020">#</doc-anchor-trigger>
        <span>October 20, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: <em>Expression</em> tab has now been merged into the <em>Plots</em> tab
<img src="https://user-images.githubusercontent.com/840895/99565969-4ee0f480-299a-11eb-8371-8cfd09c6e4ba.png" alt="image" /></li>
</ul>
<doc-anchor-target id="october-16-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-16-2020">#</doc-anchor-trigger>
        <span>October 16, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of 25,078 samples from 5 studies:</p>
<ul>
<li><a href="https://cbioportal.org/study/summary?id=skcm_tcga_pub_2015">Melanomas (TCGA, Cell 2015)</a> <em>359 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rbl_cfdna_msk_2020">Retinoblastoma cfDNA (MSKCC 2020)</a> <em>14 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=angs_painter_2020">The Angiosarcoma Project (Provisional, July 2020)</a> <em>83 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=blca_msk_tcga_2020">Bladder Cancer (MSK/TCGA, 2020)</a> <em>476 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=msk_ch_2020">Cancer Therapy and Clonal Hematopoiesis (MSK, 2020)</a> <em>24,146 samples</em></li>
</ul>
</li>
<li><p><strong>Added MSI data</strong> (MSIsensor from <a href="https://www.sciencedirect.com/science/article/pii/S0092867418303131#bib66">Mariamidze et al. 2018</a> and MANTIS scores from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5972025/">Roychowdhury et al. 2017</a>) for all 32 TCGA PanCan Atlas Cohorts.
<img width="600" src="https://user-images.githubusercontent.com/5973438/95988728-52131e80-0df7-11eb-8e25-0617d4b9e8c4.png"></p>
</li>
<li><p><strong>Added new profile</strong> “RNA-Seq V2 expression Z-scores relative to normal samples” for 16 TCGA PanCan Atlas Cohorts.
The normals samples RNA-Seq V2 expression data were curated from <a href="https://gdc.cancer.gov/about-data/publications/pancanatlas">GDC</a>, and can be downloaded from our <a href="https://github.com/cBioPortal/datahub">Datahub</a> or <a href="https://www.cbioportal.org/datasets">Data Set</a> page. <a href="https://www.cbioportal.org/results/plots?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&amp;case_set_id=all&amp;data_priority=0&amp;gene_list=ERBB2&amp;geneset_list=%20&amp;plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22true%22%7D&amp;plots_horz_selection=%7B%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_STUDY%22%7D&amp;plots_vert_selection=%7B%22selectedGeneOption%22%3A2064%2C%22dataType%22%3A%22MRNA_EXPRESSION%22%2C%22selectedDataSourceOption%22%3A%22rna_seq_v2_mrna_median_all_sample_ref_normal_Zscores%22%2C%22logScale%22%3A%22true%22%7D&amp;profileFilter=0&amp;tab_index=tab_visualize">Example: ERBB2 expression z-scores relative to normal expression</a></p>
</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/96191246-b32f1500-0f11-11eb-8727-bcfdd2d724a3.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="october-13-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-13-2020">#</doc-anchor-trigger>
        <span>October 13, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: <em>Study View</em> now allows comparing samples with mutations or copy number alterations in different genes
<img src="https://user-images.githubusercontent.com/840895/99558433-ba729400-2991-11eb-9054-1d8751b4d84f.png" alt="image" /></p>
</li>
<li><p><strong>New Feature</strong>: When treatment timeline is available (e.g. in <a href="https://www.cbioportal.org/study/summary?id=lgg_ucsf_2014">this study</a>), <em>Study View</em> now allows the selection and comparison of patients treated with specific drugs, or samples sequenced pre or post specific drug treatments
<img src="https://user-images.githubusercontent.com/840895/99556958-248a3980-2990-11eb-97e3-d833518f8aef.png" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="september-30-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-30-2020">#</doc-anchor-trigger>
        <span>September 30, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: Microbiome signature data is available for comparison now. <a href="https://www.cbioportal.org/comparison/generic_assay_microbiome_signature?sessionId=5dea8bbde4b05f1dfb8ace51">Example: comparing colorectal subtypes for enriched microbiome signatures</a>
<img src="https://user-images.githubusercontent.com/840895/99566847-666cad00-299b-11eb-91d3-21b21f267733.png" alt="image" /></li>
</ul>
<doc-anchor-target id="september-22-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-22-2020">#</doc-anchor-trigger>
        <span>September 22, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: The timeline feature in <em>Patient View</em> has been refactored with an improved UI. <a href="https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&amp;caseId=P04">Example</a>
<img src="https://user-images.githubusercontent.com/840895/94962574-ae577380-04c4-11eb-9ff7-1a79e242dd73.png" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: Logrank p-values are now provided for all survival analysis (previously only availalbe when comparing two groups). <a href="https://www.cbioportal.org/comparison/survival?sessionId=5f77816de4b015b63e9c0f81">Example</a>
<img width="600" alt="" src="https://user-images.githubusercontent.com/840895/94962916-481f2080-04c5-11eb-811f-40d8fc4437cf.png"></p>
</li>
</ul>
<doc-anchor-target id="august-11-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-11-2020">#</doc-anchor-trigger>
        <span>August 11, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: microbiome data of TCGA samples from <a href="https://www.nature.com/articles/s41586-020-2095-1">Poore et al. 2020</a> are now available for analysis in the <em>OncoPrint</em> and <em>Plots</em> tabs. <a href="https://www.cbioportal.org/results/plots?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&amp;case_set_id=all&amp;data_priority=0&amp;gene_list=EGFR&amp;geneset_list=%20&amp;plots_coloring_selection=%7B%22selectedOption%22%3A%22-1_undefined%22%2C%22colorByMutationType%22%3A%22false%22%2C%22colorByCopyNumber%22%3A%22false%22%7D&amp;plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%2C%22dataType%22%3A%22clinical_attribute%22%2C%22selectedDataSourceOption%22%3A%22CANCER_TYPE%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&amp;plots_vert_selection=%7B%22selectedGeneOption%22%3A1956%2C%22selectedGenericAssayOption%22%3A%22Orthohepadnavirus%22%2C%22dataType%22%3A%22MICROBIOME_SIGNATURE%22%2C%22selectedDataSourceOption%22%3A%22microbiome_signature%22%2C%22mutationCountBy%22%3A%22MutationType%22%2C%22logScale%22%3A%22false%22%7D&amp;profileFilter=0&amp;tab_index=tab_visualize">Example: Orthohepadnavirus across TCGA cancers</a>
<img src="https://user-images.githubusercontent.com/840895/90060684-1783ed00-dcb3-11ea-8368-e96dd69c552c.png" alt="image" /></p>
</li>
<li><p><strong>New Feature</strong>: You can now compare <em>DNA Methylation</em> data between groups using the <em>Comparison</em> feature. <a href="https://www.cbioportal.org/results/comparison?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_id=ov_tcga_pub&amp;cancer_study_list=ov_tcga_pub&amp;case_set_id=ov_tcga_pub_3way_complete&amp;comparison_selectedGroups=%5B%22BRCA1%3A%20EXP%3E1%22%2C%22BRCA1%3A%20EXP%3C-1%22%5D&amp;comparison_subtab=dna_methylation&amp;data_priority=0&amp;gene_list=BRCA1%253AEXP%253E1%2520%250ABRCA1%253A%2520EXP%253C-1&amp;gene_set_choice=user-defined-list&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_MRNA_EXPRESSION=ov_tcga_pub_mrna_median_Zscores&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pub_mutations&amp;plots_coloring_selection=%7B%7D&amp;plots_horz_selection=%7B%7D&amp;plots_vert_selection=%7B%22selectedDataSourceOption%22%3A%22mrna_median_Zscores%22%7D&amp;profileFilter=0&amp;tab_index=tab_visualize">Example: Comparing DNA methylation levels between samples with high vs low BRCA1 expression</a>
<img src="https://user-images.githubusercontent.com/840895/89957811-632c8d00-dc06-11ea-87dc-bdbb4fa23a0b.png" alt="image" /></p>
</li>
<li><p><strong>Added data</strong> consisting of 513 samples from 3 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_smc_2018">Breast Cancer (SMC 2018)</a> <em>187 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gct_msk_2020">Germ Cell Tumors and Shared Leukemias (MSK     2020)</a> <em>21 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_oncosg_2020">Lung Adenocarcinoma (OncoSG, Nat Genet 2020)</a> <em>305 samples</em></li>
</ul>
</li>
<li><p><strong>Added RPPA data</strong> in addition to the microbiome data for <a href="https://www.cbioportal.org/study/summary?id=acc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018">31 TCGA Pancan studies (except LAML)</a>
<img width="400" alt="" src="https://user-images.githubusercontent.com/840895/90141147-ba3a7b00-dd48-11ea-964f-840dd591e535.png"></p>
</li>
</ul>
<doc-anchor-target id="july-21-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-21-2020">#</doc-anchor-trigger>
        <span>July 21, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: The <em>Mutations</em> tab now has the option to show mutation effects for different transcripts / isoforms. Note that some annotation features are only available for the canonical isoform. <a href="https://www.cbioportal.org/results/mutations?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=msk_impact_2017&amp;case_set_id=msk_impact_2017_cnaseq&amp;data_priority=0&amp;gene_list=FGFR2&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;mutations_transcript_id=ENST00000358487&amp;profileFilter=0&amp;tab_index=tab_visualize">example</a>
<img src="https://user-images.githubusercontent.com/840895/88306535-d95d6400-ccd8-11ea-9e64-0c6600f65e50.png" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: The <em>Plots</em> tab is now supported in multi-study queries. <a href="https://www.cbioportal.org/results/plots?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=laml_tcga_pan_can_atlas_2018%2Cacc_tcga_pan_can_atlas_2018%2Cblca_tcga_pan_can_atlas_2018%2Clgg_tcga_pan_can_atlas_2018%2Cbrca_tcga_pan_can_atlas_2018%2Ccesc_tcga_pan_can_atlas_2018%2Cchol_tcga_pan_can_atlas_2018%2Ccoadread_tcga_pan_can_atlas_2018%2Cdlbc_tcga_pan_can_atlas_2018%2Cesca_tcga_pan_can_atlas_2018%2Cgbm_tcga_pan_can_atlas_2018%2Chnsc_tcga_pan_can_atlas_2018%2Ckich_tcga_pan_can_atlas_2018%2Ckirc_tcga_pan_can_atlas_2018%2Ckirp_tcga_pan_can_atlas_2018%2Clihc_tcga_pan_can_atlas_2018%2Cluad_tcga_pan_can_atlas_2018%2Clusc_tcga_pan_can_atlas_2018%2Cmeso_tcga_pan_can_atlas_2018%2Cov_tcga_pan_can_atlas_2018%2Cpaad_tcga_pan_can_atlas_2018%2Cpcpg_tcga_pan_can_atlas_2018%2Cprad_tcga_pan_can_atlas_2018%2Csarc_tcga_pan_can_atlas_2018%2Cskcm_tcga_pan_can_atlas_2018%2Cstad_tcga_pan_can_atlas_2018%2Ctgct_tcga_pan_can_atlas_2018%2Cthym_tcga_pan_can_atlas_2018%2Cthca_tcga_pan_can_atlas_2018%2Cucs_tcga_pan_can_atlas_2018%2Cucec_tcga_pan_can_atlas_2018%2Cuvm_tcga_pan_can_atlas_2018&amp;case_set_id=all&amp;data_priority=0&amp;gene_list=TP53%2520PTEN%2520BRCA1%2520IGF2%2520EGFR&amp;geneset_list=%20&amp;plots_coloring_selection=%7B%22colorByCopyNumber%22%3A%22false%22%7D&amp;plots_horz_selection=%7B%22selectedGeneOption%22%3A1956%7D&amp;plots_vert_selection=%7B%22logScale%22%3A%22true%22%7D&amp;profileFilter=0&amp;tab_index=tab_visualize">example</a>
<img src="https://user-images.githubusercontent.com/840895/88309239-07907300-ccdc-11ea-995b-7ddd658ec46e.png" alt="image" /></p>
</li>
<li><p><strong>New Feature</strong>: You can now share custom groups in the <em>Study View</em> <a href="https://www.cbioportal.org/study/summary?id=msk_impact_2017#sharedGroups=5d092670e4b0ab4137876374,5c99454ee4b0ab4137873dc5">example</a></p>
<img width="333" alt="" src="https://user-images.githubusercontent.com/840895/88309724-b7fe7700-ccdc-11ea-969a-28ed8551ffd7.png">
</li>
</ul>
<doc-anchor-target id="june-11-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-11-2020">#</doc-anchor-trigger>
        <span>June 11, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 267 samples from 2 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=stad_oncosg_2018">Gastric Cancer (OncoSG, 2018)</a> <em>147 samples</em></li>
<li>120 ctDNA samples added to <a href="https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017">Non-Small Cell Lung Cancer (TRACERx, NEJM &amp; Nature 2017)</a> <em>447 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="june-9-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-9-2020">#</doc-anchor-trigger>
        <span>June 9, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: using <a href="https://www.cbioportal.org/oql#oql-modifiers">OQL</a> to query for mutations based on a protein position range. <a href="https://www.cbioportal.org/results/mutations?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=msk_impact_2017&amp;case_set_id=msk_impact_2017_cnaseq&amp;data_priority=0&amp;gene_list=TP53%253AMUT_(95-288*)&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;profileFilter=0&amp;tab_index=tab_visualize">example</a>
<img src="https://user-images.githubusercontent.com/840895/84427197-83ce6b80-abf2-11ea-9d18-3a4f4524e545.png" alt="image" /></p>
</li>
<li><p><strong>New Feature</strong>: you can now send the OncoPrint data to the <a href="https://www.cbioportal.org/oncoprinter">OncoPrinter tool</a> for customization.
<img src="https://user-images.githubusercontent.com/840895/84318326-3c85a380-ab3c-11ea-97c1-34343cb3e996.png" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: Mutational spectrum data can be downloaded from OncoPrint
<img src="https://user-images.githubusercontent.com/840895/84322695-68585780-ab43-11ea-9224-a965331e35fc.png" alt="image" /></p>
</li>
</ul>
<doc-anchor-target id="june-2-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-2-2020">#</doc-anchor-trigger>
        <span>June 2, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Pediatric cancer studies are now grouped and highlighted in the query page
<img src="https://user-images.githubusercontent.com/840895/84318659-c897cb00-ab3c-11ea-8209-93c940321a0f.png" alt="image" /></li>
</ul>
<doc-anchor-target id="may-6-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-6-2020">#</doc-anchor-trigger>
        <span>May 6, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 574 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=usarc_msk_2020">Uterine Sarcoma/Mesenchymal (MSK, Clin Cancer Res 2020)</a> <em>108 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_mcspc_mskcc_2020">Metastatic castration-sensitive prostate cancer (MSK, Clin Cancer Res 2020)</a> <em>424 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbm_columbia_2019">Glioblastoma (Columbia, Nat Med. 2019)</a> <em>42 samples</em></li>
</ul>
</li>
<li><strong>Updated one study:</strong>
<ul>
<li>Expression data was added to <a href="https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017">The Metastatic Breast Cancer Project (Provisional, February 2020)</a>.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="april-24-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-24-2020">#</doc-anchor-trigger>
        <span>April 24, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Add a new chart on the <em>Study View</em> for selecting samples based on pre-defined case lists:</p>
  <img width="333" alt="Screen Shot 2020-04-24 at 9 18 25 AM" src="https://user-images.githubusercontent.com/1334004/80216807-a1c51a00-860c-11ea-8ba5-3781bd603ee5.png">
</li>
</ul>
<doc-anchor-target id="april-10-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-10-2020">#</doc-anchor-trigger>
        <span>April 10, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Make cohorts on the <em>Study View</em> using continuous molecular profiles of one or more gene(s), such as mRNA expression, methylation, RPPA and continuous CNA. <a href="https://www.cbioportal.org/study/summary?id=brca_tcga#filterJson=%7B%22genomicDataFilters%22:%5B%7B%22hugoGeneSymbol%22:%22ERBB2%22,%22profileType%22:%22mrna_median_Zscores%22,%22values%22:%5B%7B%22start%22:1.5,%22end%22:2%7D,%7B%22start%22:2,%22end%22:2.5%7D,%7B%22start%22:2.5,%22end%22:3%7D,%7B%22start%22:3,%22end%22:3.5%7D,%7B%22start%22:3.5,%22end%22:4%7D,%7B%22start%22:4%7D%5D%7D%5D,%22genomicProfiles%22:%5B%5B%22mrna_median_Zscores%22%5D%5D,%22studyIds%22:%5B%22brca_tcga%22%5D%7D">example</a></p>
<img src="https://user-images.githubusercontent.com/1334004/79270846-da2a6280-7e6c-11ea-8cb9-b40e7d201ea1.png" width=500 />
<p>Combine this with the group comparison feature to compare e.g. all quartiles of expression:</p>
<img src="https://user-images.githubusercontent.com/1334004/79270618-79028f00-7e6c-11ea-8a30-9eaedee948ca.png" width=400 />
</li>
<li><p><strong>New Feature</strong>: Annotate mutations using the <em>Mutation Mapper Tool</em> on the GRCh38 reference genome:</p>
<p><a href="https://www.cbioportal.org/mutation_mapper"><img src="https://user-images.githubusercontent.com/1334004/79233555-07f4b480-7e37-11ea-9d0a-0cafff434fa5.png" alt="mutation_mapper_tool_grch38" /></a></p>
</li>
</ul>
<doc-anchor-target id="april-3-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-3-2020">#</doc-anchor-trigger>
        <span>April 3, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: Extended the <em>Comparison</em> tab to support the comparison of altered samples per gene or alteration. This <a href="https://www.cbioportal.org/results/comparison?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=msk_impact_2017&amp;case_set_id=msk_impact_2017_Non-Small_Cell_Lung_Cancer&amp;comparison_createdGroupsSessionId=5e8a2d4ae4b0ff7ef5fdd7d0&amp;comparison_selectedGroups=%5B%22EGFR%3A%20AMP%20%26%20MUT%22%2C%22EGFR%3A%20MUT%20only%22%2C%22EGFR%3A%20AMP%20only%22%5D&amp;comparison_subtab=survival&amp;data_priority=0&amp;gene_list=EGFR%253AAMP%250AEGFR%253AMUT_DRIVER&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=msk_impact_2017_cna&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=msk_impact_2017_mutations&amp;profileFilter=0&amp;tab_index=tab_visualize">example query</a> compares NSCLC patients with 1) both mutated and amplified EGFR, 2) mutated EGFR only, and 3) amplified EGFR only.</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/78508085-3ea13f80-7752-11ea-8e87-e486061def12.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="march-27-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-27-2020">#</doc-anchor-trigger>
        <span>March 27, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: User selections in the Plots tab are now saved in the URL. <a href="https://www.cbioportal.org/results/plots?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_id=coadread_tcga_pub&amp;cancer_study_list=ov_tcga_pan_can_atlas_2018&amp;case_set_id=ov_tcga_pan_can_atlas_2018_cnaseq&amp;data_priority=0&amp;gene_list=TP53&amp;gene_set_choice=user-defined-list&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ov_tcga_pan_can_atlas_2018_gistic&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ov_tcga_pan_can_atlas_2018_mutations&amp;plots_horz_selection=%7B%22dataType%22%3A%22MUTATION_EXTENDED%22%7D&amp;profileFilter=0&amp;tab_index=tab_visualize">example</a></p>
</li>
<li><p><strong>New Feature</strong>: Added table of data availability per profile in the <em>Study View</em>. <a href="https://www.cbioportal.org/study/summary?id=ucec_tcga_pub">example</a></p>
</li>
</ul>
<img src="https://user-images.githubusercontent.com/840895/78508182-e454ae80-7752-11ea-914b-bac17f725a9f.png" width="300">
<doc-anchor-target id="march-20-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-20-2020">#</doc-anchor-trigger>
        <span>March 20, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Extended <em>Survival Analysis</em> to support more outcome measures. <a href="https://www.cbioportal.org/results/comparison?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=luad_tcga_pan_can_atlas_2018&amp;case_set_id=luad_tcga_pan_can_atlas_2018_cnaseq&amp;comparison_selectedGroups=%5B%22Unaltered%20group%22%2C%22Altered%20group%22%5D&amp;comparison_subtab=survival&amp;data_priority=0&amp;gene_list=KRAS%250ATP53&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=luad_tcga_pan_can_atlas_2018_gistic&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=luad_tcga_pan_can_atlas_2018_mutations&amp;profileFilter=0&amp;tab_index=tab_visualize">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/78508508-5c23d880-7755-11ea-9e16-36bff32678ad.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="march-18-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-18-2020">#</doc-anchor-trigger>
        <span>March 18, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 1,393 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=breast_alpelisib_2020">Breast Cancer (Alpelisib plus AI, Nature Cancer 2020)</a> <em>141 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=glioma_mskcc_2019">Glioma (MSKCC, Clin Cancer Res 2019)</a> <em>1,004 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cfdna_msk_2019">Mixed cfDNA (MSK, Nature Medicine 2019)</a> <em>248 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="march-3-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-3-2020">#</doc-anchor-trigger>
        <span>March 3, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Added Pathways tab to the <em>Results View</em> page, which visualizes the alteration frequencies of genes in pathways of interest. The pathways are pulled from <a href="https://www.pathwaymapper.org">https://www.pathwaymapper.org</a> and shown in a read only view. One can edit these pathways in the PathwayMapper editor. For more information see the <a href="https://www.cbioportal.org/tutorials#pathways">tutorial</a>.</p>
<p><a href="https://www.cbioportal.org/results/pathways?Action=Submit&amp;Z_SCORE_THRESHOLD=1.0&amp;cancer_study_id=gbm_tcga_pub&amp;cancer_study_list=gbm_tcga_pub&amp;case_set_id=gbm_tcga_pub_sequenced&amp;gene_list=TP53%20MDM2%20MDM4&amp;gene_set_choice=user-defined_list&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=gbm_tcga_pub_cna_rae&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=gbm_tcga_pub_mutations"><img src="https://user-images.githubusercontent.com/1334004/76771001-133fbc00-6775-11ea-8b36-82edb1e7be09.png" alt="pathwaymapper_screenshot" /></a></p>
</li>
</ul>
<doc-anchor-target id="february-12-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-12-2020">#</doc-anchor-trigger>
        <span>February 12, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 1,605 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ntrk_msk_2019">Tumors with TRK fusions (MSK, 2019)</a> <em>106 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lymphoma_cellline_msk_2020">Lymphoma Cell Lines (MSKCC, 2020)</a> <em>34 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_cdk12_mskcc_2020">Prostate Adenocarcinoma (MSKCC, 2020)</a> <em>1,465 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="february-6-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-6-2020">#</doc-anchor-trigger>
        <span>February 6, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Extend the <a href="https://www.cbioportal.org/tutorials#group-comparison">recent group comparison feature</a> by allowing comparisons inside the <em>Results View</em> page. The new tab allows for quick comparison of altered vs unaltered cases by survival, clinical information, mutation, copy number events and mRNA expression:</p>
<p><a href="https://www.cbioportal.org/results/comparison?Z_SCORE_THRESHOLD=2.0&amp;cancer_study_id=coadread_tcga_pub&amp;cancer_study_list=coadread_tcga_pub&amp;case_set_id=coadread_tcga_pub_nonhypermut&amp;gene_list=KRAS%20NRAS%20BRAF&amp;gene_set_choice=user-defined-list&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=coadread_tcga_pub_gistic&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=coadread_tcga_pub_mutations"><img src="https://user-images.githubusercontent.com/1334004/74002155-2a041f00-493c-11ea-9867-14740202c368.gif" alt="group_results640px" /></a></p>
</li>
<li><p><strong>Performance enhancement</strong>: the <em>Study View</em>&#x27;s mutation table now loads faster for studies with multiple gene panels. For the <a href="https://genie.cbioportal.org">genie portal</a>, which has a study with many different gene panels this resulted in a speed-up from ~90-120 seconds to 5 seconds.</p>
</li>
<li><p>Read more about the v3.2.2 release <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.2">here</a></p>
</li>
</ul>
<doc-anchor-target id="january-30-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-30-2020">#</doc-anchor-trigger>
        <span>January 30, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Show HGVSg in mutations table and linkout to <a href="https://www.genomenexus.org">Genome Nexus</a>:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/73494837-4d026200-4383-11ea-8968-f9a8a6e00675.png" alt="hgvsg genome nexus" /></p>
</li>
<li><p><strong>Enhancement</strong>: Add a pencil button near gene list in results page which opens interface for quickly modifying the oql of the query:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/73494684-f72dba00-4382-11ea-8bd0-001fdf0cb681.png" alt="edit query pencil" /></p>
</li>
<li><p>See more updates <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v3.2.1">here</a></p>
</li>
</ul>
<doc-anchor-target id="january-29-2020">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-29-2020">#</doc-anchor-trigger>
        <span>January 29, 2020</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 197 samples from 2 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=utuc_pdx_msk_2019">Bladder/Urinary Tract Cancer (MSK, 2019)</a> <em>78 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=utuc_msk_2019">Upper Tract Urothelial Carcinoma (MSK, 2019)</a> <em>119 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="december-19-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-19-2019">#</doc-anchor-trigger>
        <span>December 19, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: We restored support for submitting large queries from external applications using HTTP POST requests.  Accepted parameters are the same
as appear in the url of a query submitted from the homepage.</p>
</li>
<li><p>See more updates <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.8">here</a></p>
</li>
</ul>
<doc-anchor-target id="december-12-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-12-2019">#</doc-anchor-trigger>
        <span>December 12, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: Several enhancements to the display of gene panels on the <em>Patient View</em> page, by <a href="https://www.thehyve.nl/">The Hyve</a>, described in more detail <a href="https://www.thehyve.nl/articles/gene-panels-cbioportal-patient-view">here</a></p>
<p><img src="https://api.thehyve.nl/uploads/gene-panels-patient-view-cbioportal.png" alt="image" /></p>
</li>
<li><p><strong>Enhancement</strong>: Add Count Bubbles to Oncoprint Toolbar</p>
<p><img src="https://user-images.githubusercontent.com/20069833/70339336-aa16e700-181c-11ea-94ac-c4acff272e4f.png" alt="Screenshot from 2019-12-06 11-36-21" /></p>
</li>
<li><p>See more updates <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.7">here</a></p>
</li>
</ul>
<doc-anchor-target id="november-29-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-29-2019">#</doc-anchor-trigger>
        <span>November 29, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Support group comparison for custom charts in <em>Study View</em> page</li>
<li><strong>Enhancement</strong>: Performance improvement of Co-Expression analysis.</li>
<li><strong>Enhancement</strong>: Kaplan-Meier plots now supports custom time range.</li>
<li>See more updates <a href="https://github.com/cBioPortal/cbioportal/releases/tag/v3.1.6">here</a></li>
</ul>
<img src="https://user-images.githubusercontent.com/840895/69887458-02963380-12b5-11ea-8151-4d531db0fcc5.png" width="500">
<doc-anchor-target id="november-22-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-22-2019">#</doc-anchor-trigger>
        <span>November 22, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: Support for Treatment response data in the Oncoprint and Plots tab, including new <em>Waterfall</em> plot type. Read more in <a href="https://www.thehyve.nl/articles/treatment-response-data-cbioportal">The Hyve&#x27;s blog post</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://api.thehyve.nl/uploads/Waterfall-plots-cbioportal.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="november-15-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-15-2019">#</doc-anchor-trigger>
        <span>November 15, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: heatmap tracks in OncoPrint now has separate headers and sub-menus. <a href="https://www.cbioportal.org/results/oncoprint?Action=Submit&amp;RPPA_SCORE_THRESHOLD=2.0&amp;Z_SCORE_THRESHOLD=2.0&amp;cancer_study_list=brca_tcga&amp;case_set_id=brca_tcga_protein_quantification&amp;clinicallist=PROFILED_IN_brca_tcga_mutations&amp;data_priority=0&amp;gene_list=ESR1%2520PGR%2520ERBB2&amp;geneset_list=%20&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=brca_tcga_gistic&amp;genetic_profile_ids_PROFILE_MRNA_EXPRESSION=brca_tcga_mrna_median_Zscores&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=brca_tcga_mutations&amp;genetic_profile_ids_PROFILE_PROTEIN_EXPRESSION=brca_tcga_rppa_Zscores&amp;heatmap_track_groups=brca_tcga_rna_seq_v2_mrna_median_Zscores%2CESR1%2CPGR%2CERBB2%3Bbrca_tcga_protein_quantification_zscores%2CESR1%2CPGR%2CERBB2&amp;tab_index=tab_visualize">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/69502724-7fee2c80-0ee0-11ea-81b7-2ada01fc99a2.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<ul>
<li><strong>Enhancement</strong>: global settings for query session</li>
</ul>
<img src="https://user-images.githubusercontent.com/840895/69502843-b6787700-0ee1-11ea-927b-049635a5a8ac.png" width="300">
<doc-anchor-target id="november-7-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-7-2019">#</doc-anchor-trigger>
        <span>November 7, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 212 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=skcm_dfci_2015">Metastatic Melanoma (DFCI, Science 2015)</a> <em>110 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=skcm_mskcc_2014">Melanoma (MSKCC, NEJM 2014)</a> <em>64 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mel_ucla_2016">Metastatic Melanoma (UCLA, Cell 2016)</a> <em>38 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-30-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-30-2019">#</doc-anchor-trigger>
        <span>October 30, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 178 samples from 2 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ihch_smmu_2014">Intrahepatic Cholangiocarcinoma (Shanghai, Nat Commun 2014)</a> <em>103 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2018">Non-Small Cell Lung Cancer (MSK, Cancer Cell 2018)</a> <em>75 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-23-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-23-2019">#</doc-anchor-trigger>
        <span>October 23, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Quick example links in Plots tab. <a href="https://www.cbioportal.org/results/plots?session_id=5ba11ce6498eb8b3d567e7ed">example</a></li>
</ul>
<doc-anchor-target id="october-14-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-14-2019">#</doc-anchor-trigger>
        <span>October 14, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: Fusion Genes table in <code v-pre>Study View</code>. <a href="https://www.cbioportal.org/study/summary?id=msk_impact_2017">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/68320157-6ef28e00-008d-11ea-8cb3-e4b54aa1d848.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="october-11-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-11-2019">#</doc-anchor-trigger>
        <span>October 11, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Enhancement</strong>: The Download interface on the homepage has been removed. Enhanced download functionality is now available after querying on the results page.</p>
<p><em>Home page</em>:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/66735303-48627e00-ee66-11e9-863f-5ec713ea444a.png" alt="homepage download tab removed" /></p>
<p><em>Results page</em>:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/66736057-7ea0fd00-ee68-11e9-942b-e123b9659126.png" alt="results page download tab" /></p>
<p>Note that as before one can always download the full raw data on the <a href="https://www.cbioportal.org/datasets">Data Sets page</a> or from <a href="https://github.com/cBioPortal/datahub/">Datahub</a>.</p>
</li>
</ul>
<doc-anchor-target id="october-9-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-9-2019">#</doc-anchor-trigger>
        <span>October 9, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Added data</strong> consisting of  2725 samples from 4 studies:</p>
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ccle_broad_2019">Cancer Cell Line Encyclopedia (Broad, 2019)</a> <em>1739 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cll_broad_2015#summary">Chronic Lymphocytic Leukemia (Broad, Nature 2015)</a> <em>537 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=rectal_msk_2019">Rectal Cancer (MSK,Nature Medicine 2019)</a> <em>339 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coad_cptac_2019">Colon Cancer (CPTAC-2 Prospective, Cell 2019)</a> <em>110 samples</em></li>
</ul>
</li>
<li><p><strong>Updated</strong>  <a href="https://www.cbioportal.org/study/summary?id=stes_tcga_pub#summary">Esophageal Carcinoma (TCGA, Nature 2017)</a> with addition of CNA data for Esophageal Squamous Cell Carcinoma cases <em>90 samples</em>.</p>
</li>
</ul>
<doc-anchor-target id="september-18-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-18-2019">#</doc-anchor-trigger>
        <span>September 18, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Feature</strong>: The list and order of charts of a study will be automatically saved now as a user preference on the <em>study view</em> page.</li>
</ul>
<doc-anchor-target id="september-6-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-6-2019">#</doc-anchor-trigger>
        <span>September 6, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 1216 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_mskcc_2019">Breast Cancer (MSKCC, 2019)</a> <em>70 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbm_mayo_pdx_sarkaria_2019">Brain Tumor PDXs (Mayo Clinic, 2019)</a><em>97 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=acc_2019">Adenoid Cystic Carcinoma Project (2019)</a> <em>1049 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="august-13-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-13-2019">#</doc-anchor-trigger>
        <span>August 13, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 295 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=pptc_2019">Pediatric Preclinical Testing Consortium (PPTC, 2019)</a> <em>261 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=nsclc_mskcc_2015">Non-small cell lung cancer (MSK, Science 2015)</a> <em>16 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_msk_2019">Prostate Cancer (MSK, 2019)</a> <em>18 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-26-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-26-2019">#</doc-anchor-trigger>
        <span>July 26, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 35 samples from 1 study:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=ccrcc_dfci_2019">Clear Cell Renal Cell Carcinoma (DFCI, Science 2019)</a> <em>35 samples</em></li>
</ul>
</li>
<li>Added Hypoxia data for:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=brca_tcga_pan_can_atlas_2018">Breast Invasive Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coadread_tcga_pan_can_atlas_2018">Colorectal Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=cesc_tcga_pan_can_atlas_2018">Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=gbm_tcga_pan_can_atlas_2018">Glioblastoma Multiforme (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=hnsc_tcga_pan_can_atlas_2018">Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=kirc_tcga_pan_can_atlas_2018">Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=kirp_tcga_pan_can_atlas_2018">Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lgg_tcga_pan_can_atlas_2018">Brain Lower Grade Glioma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lihc_tcga_pan_can_atlas_2018">Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=luad_tcga_pan_can_atlas_2018">Lung Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=lusc_tcga_pan_can_atlas_2018">Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ov_tcga_pan_can_atlas_2018">Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=paad_tcga_pan_can_atlas_2018">Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=pcpg_tcga_pan_can_atlas_2018">Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=prad_tcga_pan_can_atlas_2018">Prostate Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=skcm_tcga_pan_can_atlas_2018">Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=thca_tcga_pan_can_atlas_2018">Thyroid Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study/summary?id=ucec_tcga_pan_can_atlas_2018">Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-24-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-24-2019">#</doc-anchor-trigger>
        <span>July 24, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consisting of 151 samples from 1 study:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=mpn_cimr_2013">Myeloproliferative Neoplasms (CIMR, NEJM 2013)</a> <em>151 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-13-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-13-2019">#</doc-anchor-trigger>
        <span>July 13, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Public Release 6.1 of AACR Project GENIE:</strong>
<ul>
<li>The sixth data set, GENIE 6.0-public, was released in early July 2019.  A patch to GENIE 6.0-public, GENIE 6.1-pubic, was subsequently released on July 13, 2019.  The combined data set now includes nearly 70,000 de-identified genomic records collected from patients who were treated at each of the consortium&#x27;s participating institutions, making it among the largest fully public cancer genomic data sets released to date. The combined data set now includes data for nearly 80 major cancer types, including data from nearly 11,000 patients with lung cancer, greater than  9,700 patients with breast cancer, and nearly 7,000 patients with colorectal cancer.</li>
</ul>
</li>
<li>More detailed information can be found in the <a href="https://www.aacr.org/Research/Research/Documents/GENIE%20Data%20Guide.pdf">AACR GENIE Data Guide</a>.  In addition to accessing the data via the cBioPortal, users can download the data directly from <a href="https://synapse.org/genie">Sage Bionetworks</a>. Users will need to create an account for either site and agree to the <a href="https://www.aacr.org/Documents/Terms%20of%20Access.pdf">terms of access</a>.</li>
<li>For frequently asked questions, visit the <a href="https://www.aacr.org/Research/Research/Pages/aacr-project-genie.aspx">AACR FAQ page</a>.</li>
</ul>
<doc-anchor-target id="july-2-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-2-2019">#</doc-anchor-trigger>
        <span>July 2, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> consistng of 785 samples from 4 studies:
<ul>
<li><a href="https://www.cbioportal.org/study/summary?id=nsclc_tracerx_2017">Non-Small Cell Lung Cancer (TRACERx, NEJM 2017)</a> <em>327 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=mnm_washu_2016">Acute myeloid leukemia or myelodysplastic syndromes (WashU, 2016)</a> <em>136 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=bcc_unige_2016">Basal Cell Carcinoma (UNIGE, Nat Genet 2016)</a> <em>293 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=coad_caseccc_2015">Colon Adenocarcinoma (CaseCCC, PNAS 2015)</a> <em>29 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="june-19-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-19-2019">#</doc-anchor-trigger>
        <span>June 19, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Show <em>Genome Aggregation Database (gnomAD)</em> population frequencies in the mutations table - see <a href="https://bit.ly/2ISHgiu">example</a>:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/59794400-e07c9c00-92a6-11e9-97ea-a79bfc8f3885.gif" alt="gnomad feature news" /></p>
</li>
</ul>
<doc-anchor-target id="june-12-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-12-2019">#</doc-anchor-trigger>
        <span>June 12, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 1350 samples from 3 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=pcpg_tcga_pub">Pheochromocytoma and Paraganglioma (TCGA, Cell 2017)</a> <em>178 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=metastatic_solid_tumors_mich_2017">Metastatic Solid Cancers (UMich, Nature 2017)</a> <em>500 samples</em></li>
<li><a href="https://www.cbioportal.org/study/summary?id=aml_ohsu_2018">Acute Myeloid Leukemia (OHSU, Nature 2018)</a> <em>672 samples</em></li>
</ul>
</li>
<li>Added survival data for TCGA PanCan Atlas Cohorts (&gt;10,000 samples across 33 tumor types).</li>
<li>Added hypoxia data for <a href="https://www.cbioportal.org/study/summary?id=blca_tcga_pan_can_atlas_2018">Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)</a></li>
</ul>
<doc-anchor-target id="june-7-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-7-2019">#</doc-anchor-trigger>
        <span>June 7, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New Group Comparison Feature</strong>: Compare clinical and genomic features of user-defined groups of samples/patients. <a href="https://www.cbioportal.org/tutorials#group-comparison">View Tutorial</a></li>
</ul>
<p><figure class="content-center">
    <a href="https://www.cbioportal.org/comparison/mutations?sessionId=5cf92520e4b0ab4137874376">
        <img src="https://user-images.githubusercontent.com/840895/59052073-40f9eb00-885c-11e9-9ddb-6d036533e5f5.png" alt="group-comparison" />
    </a>
    <figcaption class="caption">group-comparison</figcaption>
</figure>
</p>
<doc-anchor-target id="may-8-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-8-2019">#</doc-anchor-trigger>
        <span>May 8, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Show <em>Post Translational Modification (PTM)</em> information from <a href="https://dbptm.mbc.nctu.edu.tw/">dbPTM</a> on the <em>Mutation Mapper</em> - see <a href="https://bit.ly/2VVrWdi">example</a>:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/57391288-34667200-718c-11e9-9fcc-b849542c74b3.gif" alt="ptm feature_news" /></p>
</li>
</ul>
<doc-anchor-target id="april-26-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-26-2019">#</doc-anchor-trigger>
        <span>April 26, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 568 samples from 4 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=acyc_jhu_2016">Adenoid Cystic Carcinoma (JHU, Cancer Prev Res 2016)</a> <em>25 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=histiocytosis_cobi_msk_2019">Histiocytosis Cobimetinib (MSK, Nature 2019)</a> <em>52 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=utuc_cornell_baylor_mdacc_2019">Upper Tract Urothelial Carcinoma (Cornell/Baylor/MDACC, Nat Comm 2019)</a> <em>47 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=prad_su2c_2019">Metastatic Prostate Adenocarcinoma (SU2C/PCF Dream Team, PNAS 2019)</a> <em>444 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="march-29-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-29-2019">#</doc-anchor-trigger>
        <span>March 29, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New Feature</strong>: Use the new quick search tab on the homepage to more easily navigate to a study, gene or patient:</p>
<p><img src="https://user-images.githubusercontent.com/1334004/55113078-8f4c7a00-50b4-11e9-9d95-e9a6e1dcda52.gif" alt="quick_search_news" /></p>
</li>
</ul>
<doc-anchor-target id="march-15-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-15-2019">#</doc-anchor-trigger>
        <span>March 15, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 338 samples from 4 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=acyc_mgh_2016">Adenoid Cystic Carcinoma (MGH, Nat Gen  2016)</a> <em>10 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=gbc_msk_2018">Gallbladder Cancer (MSK, Cancer 2018)</a> <em>103 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=prad_mpcproject_2018">The Metastatic Prostate Cancer Project (Provisional, December 2018)</a> <em>19 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=sarc_tcga_pub">Adult Soft Tissue Sarcomas (TCGA, Cell 2017)</a> <em>206 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="february-22-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-22-2019">#</doc-anchor-trigger>
        <span>February 22, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Enhancement</strong>: Exon number and HGVSc annotations are available in optional columns in the Mutations tab on the <em>Results</em> page and in the <em>Patient View</em>.</li>
<li><strong>New feature</strong>: option to a show regression line in the scatter plot in the Plots tab on the <em>Results</em> page</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/53193265-99271d00-35de-11e9-8ffc-352afc2ba906.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="february-19-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-19-2019">#</doc-anchor-trigger>
        <span>February 19, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New feature</strong>: <em>Copy-Number Segments tab</em> on the <em>Study View</em> page using <a href="https://github.com/igvteam/igv.js">igv.js v2</a> - see <a href="https://www.cbioportal.org/study?id=ucec_tcga_pub&amp;tab=cnSegments">example</a></li>
<li>Improved <em>Copy-Number Segments</em> tab on the <em>Results</em> page</li>
<li><strong>New feature</strong>: <em>OncoKB</em> and <em>Cancer Hotspots tracks</em> in the Mutations tab on the <em>Results</em> page</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/53054860-d1f6b300-3473-11e9-8178-1df10eed36fc.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="january-24-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-24-2019">#</doc-anchor-trigger>
        <span>January 24, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 2328 samples from 8 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=um_qimr_2016#summary">Uveal Melanoma (QIMR, Oncotarget 2016)</a> <em>28 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=vsc_cuk_2018">Squamous Cell Carcinoma of the Vulva (CUK, Exp Mol Med 2018)</a> <em>15 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=tmb_mskcc_2018#summary">TMB and Immunotherapy (MSKCC, Nat Genet 2019)</a> <em>1661 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=glioma_msk_2018">Glioma (MSK, 2018)</a> <em>91 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=blca_cornell_2016">Urothelial Carcinoma (Cornell/Trento, Nat Gen 2016)</a> <em>72 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=hcc_mskimpact_2018">Hepatocellular Carcinoma (MSK, Clin Cancer Res 2018)</a> <em>127 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=lung_msk_pdx">MSK Thoracic PDX (MSK, Provisional)</a> <em>139 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=chol_msk_2018">Cholangiocarcinoma (MSK, Clin Cancer Res 2018)</a> <em>195 samples</em></li>
</ul>
</li>
<li><strong>Updated data</strong> for <a href="https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017">The Metastatic Breast Cancer Project (Provisional, October 2018)</a> <em>237 samples</em></li>
</ul>
<doc-anchor-target id="january-10-2019">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-10-2019">#</doc-anchor-trigger>
        <span>January 10, 2019</span>
    </h2>
</doc-anchor-target>
<ul>
<li>cBioPortal now supports queries for driver mutations, fusions and copy number alterations as well as germline/somatic mutations using <a href="https://www.cbioportal.org/oql">Onco Query Language (OQL)</a> -- see <a href="https://www.cbioportal.org/results/oncoprint?session_id=5c23ea81e4b05228701f9d44">example</a></li>
<li>A new <a href="https://www.cbioportal.org/tutorials">tutorial</a> explores OQL and provides examples of how OQL can be a powerful tool to refine queries.</li>
</ul>
<doc-anchor-target id="december-17-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-17-2018">#</doc-anchor-trigger>
        <span>December 17, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The 10th phase of cBioPortal architectural upgrade is now complete: the <em>Study View</em> has been moved to the new architecture with numerous improvements. This marks the completion of the cBioPortal architectural refactoring! <span class="docs-emoji">&#x1F389;</span><span class="docs-emoji">&#x1F389;</span><span class="docs-emoji">&#x1F389;</span></li>
</ul>
<p><figure class="content-center">
    <a href="https://www.cbioportal.org/study?id=msk_impact_2017">
        <img src="https://user-images.githubusercontent.com/840895/50120599-11aadf80-0224-11e9-9f9f-adae0ed42c72.png" alt="image" />
    </a>
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="october-29-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-29-2018">#</doc-anchor-trigger>
        <span>October 29, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The ninth phase of the cBioPortal architectural upgrade is now complete: the <em>results page</em> is now a single-page application with better performance.</li>
<li>Supported plotting mutations by type in <em>Plots</em> tab</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/47598695-d996c700-d96d-11e8-9371-9e591cf25ea3.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="october-19-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-19-2018">#</doc-anchor-trigger>
        <span>October 19, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Support selection of transcript of interest in the <a href="https://www.cbioportal.org/mutation_mapper.jsp">MutationMapper tool</a> via <a href="https://www.genomenexus.org">Genome Nexus</a>.</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/1334004/47240017-f99a1980-d3b4-11e8-99a4-e8c73b25ecec.png" alt="mutation_mapper_dropdown" />
    <figcaption class="caption">mutation_mapper_dropdown</figcaption>
</figure>
</p>
<doc-anchor-target id="october-17-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-17-2018">#</doc-anchor-trigger>
        <span>October 17, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 3578 samples from 8 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=mrt_bcgsc_2016#summary">Rhabdoid Cancer (BCGSC, Cancer Cell 2016)</a> <em>40 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=dlbcl_duke_2017#summary">Diffuse Large B-Cell Lymphoma (Duke, Cell 2017)</a> <em>1001 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=dlbcl_dfci_2018#summary">Diffuse Large B cell Lymphoma (DFCI, Nat Med 2018)</a> <em>135 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=bfn_duke_nus_2015#summary">Breast Fibroepithelial Tumors (Duke-NUS, Nat Genet 2015)</a> <em>22 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=uccc_nih_2017#summary">Uterine Clear Cell Carcinoma (NIH, Cancer 2017)</a> <em>16 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=ucec_msk_2018#summary">Endometrial Cancer (MSK, 2018)</a> <em>197 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=breast_msk_2018#summary">Breast Cancer (MSK, Cancer Cell 2018)</a> <em>1918 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=mixed_allen_2018#summary">MSS Mixed Solid Tumors (Van Allen, 2018)</a> <em>249 samples</em></li>
</ul>
</li>
<li>Updated data for <a href="https://www.cbioportal.org/study?id=angs_project_painter_2018#summary">The Angiosarcoma Project (Provisional, September 2018)</a> <em>48 samples</em></li>
</ul>
<doc-anchor-target id="august-20-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-20-2018">#</doc-anchor-trigger>
        <span>August 20, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Now you can log in on the public cBioPortal with your Google account and save your virtual studies for quick analysis.</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/44370286-902a8700-a4a7-11e8-9c8f-dfda87fbd66b.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="august-7-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-7-2018">#</doc-anchor-trigger>
        <span>August 7, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The eighth phase of the cBioPortal architectural upgrade is now complete: The <em>Plots</em>, <em>Expression</em>, <em>Network</em>, and <em>Bookmarks</em> tabs, and therefore all analysis tabs in the <em>results page</em>, have been moved to the new architecture.</li>
<li>Updated the <a href="https://www.cbioportal.org/mutation_mapper.jsp">MutationMapper tool</a>, now connecting to <a href="https://genomenexus.org/">Genome Nexus</a> for annotating mutations on the fly.</li>
<li><em>Total Mutations</em> and <em>Fraction Genome Altered</em> are now available in <em>Plots</em> tab for visualization and analysis.</li>
<li>Enhanced clinical attribute selector for OncoPrint, now showing sample counts per attribute.</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/43609786-ec946d64-9672-11e8-8218-4e63bd5007e8.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="july-27-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-27-2018">#</doc-anchor-trigger>
        <span>July 27, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 2787 samples from 10 studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=mixed_pipseq_2017#summary">Mixed Tumors (PIP-Seq 2017)</a> <em>103 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=blca_nmibc_2017#summary">Nonmuscle Invasive Bladder Cancer (MSK Eur Urol 2017)</a> <em>105 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nbl_target_2018_pub">Pediatric Neuroblastoma (TARGET, 2018)</a> <em>1089 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=pediatric_dkfz_2017">Pediatric Pan-Cancer (DKFZ - German Cancer Consortium, 2017)</a> <em>961 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=skcm_broad_brafresist_2012">Skin Cutaneous Melanoma (Broad, Cancer Discov 2014)</a> <em>78 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=cscc_hgsc_bcm_2014#summary">Cutaneous Squamous Cell Carcinoma (MD Anderson, Clin Cancer Res 2014)</a> <em>39 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nhl_bcgsc_2013#summary">Diffuse Large B-cell Lymphoma (BCGSC, Blood 2013)</a> <em>53 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nhl_bcgsc_2011#summary">Non-Hodgkin Lymphoma (BCGSC, Nature 2011)</a> <em>14 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=cllsll_icgc_2011#summary">Chronic lymphocytic leukemia (ICGA, Nat 2011)</a> <em>105 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nbl_broad_2013#summary">Neuroblastoma (Broad Institute 2013)</a> <em>240 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="june-20-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-20-2018">#</doc-anchor-trigger>
        <span>June 20, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The seventh phase of the cBioPortal architectural upgrade is now complete: The <em>Enrichments</em> and <em>Co-Expression</em> tabs have been moved to the new architecture.</li>
<li>Supported merged gene tracks in OncoPrint and <a href="https://www.cbioportal.org/onco_query_lang_desc.jsp">Onco Query Language</a> -- see <a href="https://www.cbioportal.org/index.do?session_id=5b2bedca498eb8b3d566ab60">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/41738366-def14850-755f-11e8-8a7c-deb789dc03da.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="may-10-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-10-2018">#</doc-anchor-trigger>
        <span>May 10, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Enhanced OncoPrint to show germline mutations -- see <a href="https://www.cbioportal.org/index.do?session_id=5af5a304498eb8b3d56615b6">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/39926636-c3304d8c-54fd-11e8-9cb7-76430c8772d7.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="april-17-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-17-2018">#</doc-anchor-trigger>
        <span>April 17, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong>  of 3732 samples from 4 TARGET studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=all_phase2_target_2018_pub#summary">Pediatric Acute Lymphoid Leukemia - Phase II (TARGET, 2018)</a> <em>1978 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=aml_target_2018_pub#summary">Pediatric Acute Myeloid Leukemia (TARGET, 2018)</a> <em>1025 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=rt_target_2018_pub#summary">Pediatric Rhabdoid Tumor (TARGET, 2018)</a> <em>72 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=wt_target_2018_pub#summary">Pediatric Wilms&#x27; Tumor (TARGET, 2018)</a> <em>657 samples</em></li>
</ul>
</li>
<li><strong>Added data</strong>  of 3416 samples from 10 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=prad_p1000#summary">Prostate Adenocarcinoma (MSKCC/DFCI, Nature Genetics 2018)</a> <em>1013 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=prad_eururol_2017#summary">Prostate Adenocarcinoma (EurUrol, 2017)</a> <em>65 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nsclc_pd1_msk_2018#summary">Non-Small Cell Lung Cancer (MSK, JCO 2018)</a> <em>240 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=sclc_cancercell_gardner_2017#summary">Small-Cell Lung Cancer (Multi-Institute 2017)</a> <em>20 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=angs_project_painter_2018#summary">The Angiosarcoma Project (Provisional, February 2018)</a> <em>14 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=all_stjude_2016#summary">Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2016)</a> <em>73 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=brca_mbcproject_wagle_2017#summary">Updated Segment data and Allele Frequencies for The Metastatic Breast Cancer Project (Provisional, October 2017)</a> <em>103 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=crc_msk_2018#summary">Colorectal Cancer (MSK, Cancer Cell 2018)</a> <em>1134 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=egc_msk_2017#summary">Metastatic Esophagogastric Cancer (MSK,Cancer Discovery 2017)</a> <em>341 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=blca_tcga_pub_2017#summary">Bladder Cancer (TCGA, Cell 2017)</a> <em>413 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="april-5-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-5-2018">#</doc-anchor-trigger>
        <span>April 5, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> from the TCGA PanCanAtlas project with &gt;10,000 samples from 33 tumor types:
<ul>
<li><a href="https://www.cbioportal.org/study?id=acc_tcga_pan_can_atlas_2018">Adrenocortical Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=blca_tcga_pan_can_atlas_2018">Bladder Urothelial Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=brca_tcga_pan_can_atlas_2018">Breast Invasive Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=cesc_tcga_pan_can_atlas_2018">Cervical Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=coad_tcga_pan_can_atlas_2018">Colon Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=chol_tcga_pan_can_atlas_2018">Cholangiocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=dlbc_tcga_pan_can_atlas_2018">Diffuse Large B-Cell Lymphoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=esca_tcga_pan_can_atlas_2018">Esophageal Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=gbm_tcga_pan_can_atlas_2018">Glioblastoma Multiforme (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=hnsc_tcga_pan_can_atlas_2018">Head and Neck Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=kich_tcga_pan_can_atlas_2018">Kidney Chromophobe (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=kirc_tcga_pan_can_atlas_2018">Kidney Renal Clear Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=kirp_tcga_pan_can_atlas_2018">Kidney Renal Papillary Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=laml_tcga_pan_can_atlas_2018">Acute Myeloid Leukemia (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=lgg_tcga_pan_can_atlas_2018">Brain Lower Grade Glioma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=lihc_tcga_pan_can_atlas_2018">Liver Hepatocellular Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=luad_tcga_pan_can_atlas_2018">Lung Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=lusc_tcga_pan_can_atlas_2018">Lung Squamous Cell Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=meso_tcga_pan_can_atlas_2018">Mesothelioma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ov_tcga_pan_can_atlas_2018">Ovarian Serous Cystadenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=paad_tcga_pan_can_atlas_2018">Pancreatic Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=pcpg_tcga_pan_can_atlas_2018">Pheochromocytoma and Paraganglioma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=prad_tcga_pan_can_atlas_2018">Prostate Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=read_tcga_pan_can_atlas_2018">Rectum Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=sarc_tcga_pan_can_atlas_2018">Sarcoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=skcm_tcga_pan_can_atlas_2018">Skin Cutaneous Melanoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=stad_tcga_pan_can_atlas_2018">Stomach Adenocarcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=tgct_tcga_pan_can_atlas_2018">Testicular Germ Cell Tumors (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=thca_tcga_pan_can_atlas_2018">Thyroid Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=thym_tcga_pan_can_atlas_2018">Thymoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ucec_tcga_pan_can_atlas_2018">Uterine Corpus Endometrial Carcinoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ucs_tcga_pan_can_atlas_2018">Uterine Carcinosarcoma (TCGA, PanCancer Atlas)</a></li>
<li><a href="https://www.cbioportal.org/study?id=uvm_tcga_pan_can_atlas_2018">Uveal Melanoma (TCGA, PanCancer Atlas)</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="march-20-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-20-2018">#</doc-anchor-trigger>
        <span>March 20, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The sixth phase of the cBioPortal architectural upgrade is now complete: The <em>Download</em> tab has been moved to the new architecture.</li>
<li>Data can now be downloaded in <a href="https://www.thehyve.nl/articles/download-data-from-cbioportal-oncoprint">tabular format</a> from OncoPrint.</li>
<li>Added an option to download an SVG file on the <em>Cancer Type Summary</em> tab.</li>
</ul>
<doc-anchor-target id="january-15-2018">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-15-2018">#</doc-anchor-trigger>
        <span>January 15, 2018</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The fifth phase of the cBioPortal architectural upgrade is now complete: The OncoPrint and Survival tabs have been moved to the new architecture.</li>
</ul>
<doc-anchor-target id="november-20-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-20-2017">#</doc-anchor-trigger>
        <span>November 20, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li>You can now combine multiple studies and view them on the study summary page. <a href="https://www.cbioportal.org/study?id=liad_inserm_fr_2014%2Chcc_inserm_fr_2015%2Clihc_amc_prv%2Clihc_riken%2Clihc_tcga">Example: liver cancer studies</a></li>
<li>You can now bookmark or share your selected samples as virtual studies with the <em>share</em> icon on the study summary page. <a href="https://www.cbioportal.org/study?id=5a12fd57498eb8b3d5605cd4">Example: a virtual study of breast tumors</a></li>
<li>Cross-study query reimplemented: Now you can view an OncoPrint of multiple studies. <a href="https://www.cbioportal.org/index.do?session_id=5a135bee498eb8b3d5605d01">Example: querying NSCLC tumors from 5 studies</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/33045546-3c8a636c-ce1b-11e7-9f52-7060c89a8dfd.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="october-17-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-17-2017">#</doc-anchor-trigger>
        <span>October 17, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The fourth phase of the cBioPortal architectural upgrade is now complete: The <em>Mutual Exclusivity</em> and <em>Cancer Type Summary</em> tabs have been moved to the new architecture.</li>
<li>Updated protein structure alignment data in Mutations tab are now retrieved from Genome Nexus via the <a href="https://g2s.genomenexus.org/">G2S web service</a>.</li>
</ul>
<doc-anchor-target id="october-2-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-2-2017">#</doc-anchor-trigger>
        <span>October 2, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> of 1646 samples from 7 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=odg_msk_2017#summary">NGS in Anaplastic Oligodendroglioma and Anaplastic Oligoastrocytomas tumors (MSK, Neuro Oncol 2017)</a>   <em>22 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=lung_msk_2017">MSK-IMPACT Clinical Sequencing Cohort for Non-Small Cell Cancer (MSK, Cancer Discovery 2017)</a> <em>915 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=mel_tsam_liang_2017#summary">Paired-exome sequencing of acral melanoma (TGEN, Genome Res 2017)</a> <em>38 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=prad_mskcc_2017#summary">MSK-IMPACT Clinical Sequencing Cohort in Prostate Cancer (MSK, JCO Precision Oncology 2017)</a> <em>504 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=skcm_ucla_2016#summary">Whole-exome sequences (WES) of pretreatment melanoma tumors (UCLA, Cell 2016)</a> <em>39 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=skcm_vanderbilt_mskcc_2015#summary">Next generation sequencing (NGS) of pre-treatment metastatic melanoma samples (MSK, JCO Precision Oncology 2017)</a> <em>66 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=urcc_mskcc_2016#summary">Targeted gene sequencing in 62 high-grade primary Unclassified Renal Cell Carcinoma (MSK, Nature 2016)</a> <em>62 samples</em></li>
</ul>
</li>
<li><strong>Updated data</strong> for <a href="https://www.cbioportal.org/study?id=msk_impact_2017">MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)</a> with overall survival data.</li>
</ul>
<doc-anchor-target id="august-3-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-3-2017">#</doc-anchor-trigger>
        <span>August 3, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The third phase of the cBioPortal architectural upgrade is now complete: The <em>Mutations</em> tab now has a fresh look and faster performance -- see <a href="https://www.cbioportal.org/index.do?session_id=598386e5498e5df2e29376ab&amp;show_samples=false&amp;#mutation_details&amp;">example</a></li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/28942244-9ea43bee-7868-11e7-9b24-0aaf0f9c010d.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<ul>
<li>Variant interpretations from the <a href="https://civic.genome.wustl.edu">CIViC database</a> are now integrated into the annotation columns on the Mutations tab and in the patient view pages</li>
<li>New summary graph for all cancer studies and samples on the front page</li>
</ul>
<doc-anchor-target id="june-26-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-26-2017">#</doc-anchor-trigger>
        <span>June 26, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The second phase of the cBioPortal architectural upgrade is now complete: The query interface now has a fresh look and faster performance.</li>
</ul>
<p><figure class="content-center">
    <img src="https://user-images.githubusercontent.com/840895/27399281-3ba5357a-568a-11e7-9730-d230285d7805.png" alt="image" />
    <figcaption class="caption">image</figcaption>
</figure>
</p>
<doc-anchor-target id="may-12-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-12-2017">#</doc-anchor-trigger>
        <span>May 12, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> of 12,211 samples from 11 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=msk_impact_2017">MSK-IMPACT Clinical Sequencing Cohort (MSK, Nat Med 2017)</a>   <em>10,945 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=past_dkfz_heidelberg_2013">Whole-genome sequencing of pilocytic astrocytomasatic (DKFZ, Nat Genetics, 2013)</a> <em>96 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=hcc_inserm_fr_2015">Hepatocellular Carcinomas (INSERM, Nat Genet 2015)</a> <em>243 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=pact_jhu_2011">Cystic Tumor of the Pancreas (Johns Hopkins, PNAS 2011)</a> <em>32 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=panet_arcnet_2017">Whole-Genome Sequencing of Pancreatic Neuroendocrine Tumors (ARC- Net, Nature, 2017)</a> <em>98 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=mbl_sickkids_2016">Medulloblastoma (Sickkids, Nature 2016)</a> <em>46 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nsclc_unito_2016">Genetic Characterization of NSCLC young adult patients (University of Turin, Lung Cancer 2016)</a> <em>41 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=gct_msk_2016">Genomic Profile of Patients with Advanced Germ Cell Tumors (MSK, JCO 2016).</a> <em>180 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=ampca_bcm_2016">Ampullary Carcinoma (Baylor, Cell Reports 2016)</a> <em>160 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=brca_igr_2015">Mutational profiles of metastatic breast cancer (INSERM, 2016)</a> <em>216 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=prad_fhcrc">Prostate Adenocarcinoma (Fred Hutchinson CRC, Nat Med 2016)</a> <em>154 samples</em></li>
</ul>
</li>
</ul>
<doc-anchor-target id="may-5-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-5-2017">#</doc-anchor-trigger>
        <span>May 5, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li>First phase of cBioPortal architectural upgrade complete: Patient view now has fresh look and faster performance. <a href="https://bit.ly/2pNY961">example</a></li>
</ul>
<doc-anchor-target id="march-28-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-28-2017">#</doc-anchor-trigger>
        <span>March 28, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New features</strong>:</p>
<ul>
<li>Per-sample mutation spectra are now available in OncoPrints -- see <a href="https://www.cbioportal.org/index.do?cancer_study_list=ucec_tcga_pub&amp;cancer_study_id=ucec_tcga_pub&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ucec_tcga_pub_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ucec_tcga_pub_gistic&amp;Z_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;case_set_id=ucec_tcga_pub_manuscript&amp;case_ids=&amp;patient_case_select=sample&amp;gene_set_choice=user-defined-list&amp;gene_list=POLE%0D%0AERBB2%0D%0AKRAS%0D%0ACTNNB1&amp;clinical_param_selection=null&amp;tab_index=tab_visualize&amp;Action=Submit&amp;show_samples=false&amp;clinicallist=SUBTYPE,NO_CONTEXT_MUTATION_SIGNATURE,%23%20mutations&amp;">example</a></li>
</ul>
<p><img src="https://cloud.githubusercontent.com/assets/840895/24209131/d34e33b2-0efb-11e7-945a-fd2b3c66d195.png" alt="image" /></p>
<ul>
<li>mRNA heat map clustering is now supported in OncoPrints</li>
<li>MDACC Next-Generation Clustered Heat Maps are now available in the patient view</li>
<li>cBioPortal web site style change</li>
</ul>
</li>
</ul>
<doc-anchor-target id="feburary-2-2017">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#feburary-2-2017">#</doc-anchor-trigger>
        <span>Feburary 2, 2017</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>3D hotspot mutation annotations are now available from 3dhotspots.org</li>
</ul>
</li>
<li><strong>New data</strong>:
<ul>
<li>CPTAC proteomics data have been integrated for TCGA breast, ovarian, and colorectal provisional studies</li>
</ul>
</li>
</ul>
<doc-anchor-target id="december-23-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-23-2016">#</doc-anchor-trigger>
        <span>December 23, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New features</strong>:</p>
<ul>
<li>Heat map visualization of gene expression data in the OncoPrint</li>
</ul>
<p><img src="https://cloud.githubusercontent.com/assets/840895/21407479/81f8c008-c79e-11e6-9b82-3f4eb8f8e1ea.png" alt="OncoPrint Heatmap" /></p>
<ul>
<li>Heat map visualization of gene expression data in the Study View page connecting to MDACC&#x27;s TCGA Next-Generation Clustered Heat Map Compendium</li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-7-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-7-2016">#</doc-anchor-trigger>
        <span>October 7, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>New features</strong>:</p>
<ul>
<li>All data sets can now be downloaded as flat files from the new <a href="https://github.com/cBioPortal/datahub">Data Hub</a></li>
<li>Annotation of putative driver missense mutations in OncoPrints, based on <a href="https://www.oncokb.org">OncoKB</a>, mutation hotspots, and recurrence in cBioPortal and COSMIC</li>
</ul>
<p><img src="https://cloud.githubusercontent.com/assets/840895/19208804/094dd320-8ccd-11e6-8012-f30104b62ff4.png" alt="OncoPrint-OncoKB" /></p>
<ul>
<li>Copy number segments visualization directly in the browser in a new <em>CN Segments</em> tab via <a href="https://igv.org/">IGV.js</a></li>
</ul>
<p><img src="https://cloud.githubusercontent.com/assets/840895/19200747/ad03c824-8c98-11e6-8051-91fc40e1dc56.png" alt="image" /></p>
</li>
<li><p><strong>Improvements</strong>:</p>
<ul>
<li>Improved cancer study view page (bug fixes and increased performance)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-24-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-24-2016">#</doc-anchor-trigger>
        <span>July 24, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Added data</strong> of 4,375 samples from 21 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=acyc_mda_2015">Adenoid Cystic Carcinoma (MDA, Clin Cancer Res 2015)</a> <em>102 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=acyc_fmi_2014">Adenoid Cystic Carcinoma (FMI, Am J Surg Pathl. 2014)</a> <em>28 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=acyc_sanger_2013">Adenoid Cystic Carcinoma (Sanger/MDA, JCI 2013)</a> <em>24 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=acbc_mskcc_2015">Adenoid Cystic Carcinoma of the Breast (MSKCC, J Pathol. 2015)</a> <em>12 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=blca_plasmacytoid_mskcc_2016">Bladder Cancer, Plasmacytoid Variant (MSKCC, Nat Genet 2016)</a> <em>34 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=brca_metabric">Breast Cancer (METABRIC, Nat Commun 2016)</a> <em>1980 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=lcll_broad_2013">Chronic Lymphocytic Leukemia (Broad, Cell 2013)</a> <em>160 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=cll_iuopa_2015">Chronic Lymphocytic Leukemia (IUOPA, Nature 2015)</a> <em>506 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=coadread_dfci_2016">Colorectal Adenocarcinoma (DFCI, Cell Reports 2016)</a> <em>619 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=ctcl_columbia_2015">Cutaneous T Cell Lymphoma (Columbia U, Nat Genet 2015)</a> <em>42 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=dlbc_broad_2012">Diffuse Large B-Cell Lymphoma (Broad, PNAS 2012)</a> <em>58 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=liad_inserm_fr_2014">Hepatocellular Adenoma (Inserm, Cancer Cell 2014)</a> <em>46 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=all_stjude_2013">Hypodiploid Acute Lymphoid Leukemia (St Jude, Nat Genet 2013)</a> <em>44 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=panet_shanghai_2013">Insulinoma (Shanghai, Nat Commun 2013)</a> <em>10 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=plmeso_nyu_2015">Malignant Pleural Mesothelioma (NYU, Cancer Res 2015)</a> <em>22 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=mcl_idibips_2013">Mantle Cell Lymphoma (IDIBIPS, PNAS 2013)</a> <em>29 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=mds_tokyo_2011">Myelodysplasia (Tokyo, Nature 2011)</a> <em>29 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=nbl_ucologne_2015">Neuroblastoma (Broad, Nat Genet 2013)</a> <em>56 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=hnsc_mdanderson_2013">Oral Squamous Cell Carcinoma (MD Anderson, Cancer Discov 2013)</a> <em>40 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=paad_qcmg_uq_2016">Pancreatic Adenocarcinoma (QCMG, Nature 2016)</a> <em>383 samples</em></li>
<li><a href="https://www.cbioportal.org/study?id=hnc_mskcc_2016">Recurrent and Metastatic Head &amp; Neck Cancer (JAMA Oncology, 2016)</a> <em>151 samples</em></li>
</ul>
</li>
<li><strong>New TCGA study</strong>:
<ul>
<li><a href="https://www.cbioportal.org/study?id=nsclc_tcga_broad_2016">Pan-Lung Cancer (TCGA, Nat Genet 2016)</a> <em>1144 samples</em></li>
</ul>
</li>
<li>Updated <strong>TCGA provisional studies</strong>
<ul>
<li>updated to the Firehose run of January 28, 2016</li>
<li>RPPA data updated with the latest data from MD Anderson</li>
<li><a href="http://oncotree.mskcc.org/">OncoTree</a> codes assigned per sample</li>
</ul>
</li>
</ul>
<doc-anchor-target id="june-6-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-6-2016">#</doc-anchor-trigger>
        <span>June 6, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Annotation of mutation effect and drug sensitivity on the Mutations tab and the patient view pages (via <a href="https://www.oncokb.org">OncoKB</a>)<br />
<img src="https://cloud.githubusercontent.com/assets/840895/15825344/54bbbc44-2bd1-11e6-8b65-e265e03f453a.png" alt="oncokb-screenshot" /></li>
</ul>
</li>
<li><strong>Improvements</strong>:
<ul>
<li>Improved OncoPrint visualization using WebGL: faster, more zooming flexibility, visualization of recurrent variants</li>
<li>Improved Network tab with SBGN view for a single interaction</li>
<li>Performance improvement of tables in the study view page</li>
<li>Mutation type summary on the Mutations tab</li>
</ul>
</li>
</ul>
<doc-anchor-target id="march-31-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-31-2016">#</doc-anchor-trigger>
        <span>March 31, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Visualization of &quot;Enrichments Analysis&quot; results via volcano plots</li>
<li>Improved performance of the cross cancer expression view  by switching to Plot.ly graphs</li>
<li>Improvements to the &quot;Clinical Data&quot; tab on the study view page</li>
<li>More customization options for the cross-cancer histograms</li>
<li>Performance improvements in the study view and query result tabs</li>
</ul>
</li>
<li><strong>Added data</strong> of 1235 samples from 3 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=lgggbm_tcga_pub">Merged Cohort of LGG and GBM (TCGA, 2016)</a></li>
<li><a href="https://www.cbioportal.org/study?id=luad_mskcc_2015">Lung Adenocarcinoma (MSKCC, 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=thyroid_mskcc_2016">Poorly-Differentiated and Anaplastic Thyroid Cancers (MSKCC, JCI 2016)</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="january-12-2016">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-12-2016">#</doc-anchor-trigger>
        <span>January 12, 2016</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Visualization of multiple samples in a patient</li>
<li>Visualization of timeline data of a patient (<a href="https://www.cbioportal.org/patient?studyId=lgg_ucsf_2014&amp;caseId=P04">example</a>)<br/>
<img src="https://cloud.githubusercontent.com/assets/840895/12055606/cca26160-aefc-11e5-93f9-2ecfe7e95caf.png" alt="timeline-example" /></li>
</ul>
</li>
<li>All <strong>TCGA data</strong> updated to the latest Firehose run of August 21, 2015</li>
<li><strong>New TCGA studies</strong>:
<ul>
<li><a href="https://www.cbioportal.org/study?id=chol_tcga">Cholangiocarcinoma (TCGA, Provisional)</a></li>
<li><a href="https://www.cbioportal.org/study?id=meso_tcga">Mesothelioma (TCGA, Provisional)</a></li>
<li><a href="https://www.cbioportal.org/study?id=tgct_tcga">Testicular Germ Cell Cancer (TCGA, Provisional)</a></li>
<li><a href="https://www.cbioportal.org/study?id=thym_tcga">Thymoma (TCGA, Provisional)</a></li>
</ul>
</li>
<li><strong>Added data</strong> of 650 samples from 10 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=nbl_amc_2012">Neuroblastoma (AMC Amsterdam, Nature 2012)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ccrcc_utokyo_2013">Clear Cell Renal Cell Carcinoma (U Tokyo, Nat Genet 2013)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ccrcc_irc_2014">Multiregion Sequencing of Clear Cell Renal Cell Carcinoma (IRC, Nat Genet 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=blca_dfarber_mskcc_2014">Bladder Urothelial Carcinoma (Dana Farber &amp; MSKCC, Cancer Discovery 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=lgg_ucsf_2014">Low-Grade Gliomas (UCSF, Science 2014)	</a></li>
<li><a href="https://www.cbioportal.org/study?id=escc_ucla_2014">Esophageal Squamous Cell Carcinoma (UCLA, Nat Genet 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=paac_jhu_2014">Acinar Cell Carcinoma of the Pancreas (Johns Hopkins, J Pathol 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=egc_tmucih_2015">Gastric Adenocarcinoma (TMUCIH, PNAS 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=pcnsl_mayo_2015">Primary Central Nervous System Lymphoma (Mayo Clinic, Clin Cancer Res 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=desm_broad_2015">Desmoplastic Melanoma (Broad Institute, Nat Genet 2015)</a></li>
</ul>
</li>
<li>All mutation data mapped to <a href="https://www.uniprot.org/help/canonical_and_isoforms">UniProt canonical isoforms</a></li>
</ul>
<doc-anchor-target id="december-23-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-23-2015">#</doc-anchor-trigger>
        <span>December 23, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Visualization of RNA-seq expression levels across TCGA studies (cross-cancer queries)<br/>
<img src="https://cloud.githubusercontent.com/assets/840895/11821643/8c9a6de4-a338-11e5-83d1-a9e92cb6dfb6.png" alt="cross cancer expression" /></li>
<li>Selection of genes in the study view to initiate queries<br/>
<img src="https://cloud.githubusercontent.com/assets/840895/11977489/5af87ca8-a951-11e5-87ad-5c31b0451cc1.png" alt="query gene in study view" /></li>
</ul>
</li>
<li><strong>Improvement</strong>:
<ul>
<li>3-D structures in the &quot;Mutations&quot; tab are now rendered by 3Dmol.js (previously JSmol)</li>
<li>Improved performance by code optimization and compressing large data by gzip</li>
</ul>
</li>
</ul>
<doc-anchor-target id="december-1-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-1-2015">#</doc-anchor-trigger>
        <span>December 1, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New feature</strong>: Annotated statistically recurrent hotspots, via new algorithm by <a href="https://www.nature.com/nbt/journal/vaop/ncurrent/full/nbt.3391.html">Chang et al. 2015</a></br>
<img src="https://cloud.githubusercontent.com/assets/840895/11794851/7729839e-a281-11e5-9413-12dc885b947d.png" alt="Annotate recurrent hotspots" /></li>
</ul>
<doc-anchor-target id="november-9-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-9-2015">#</doc-anchor-trigger>
        <span>November 9, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Links to MyCancerGenome.org for mutations<br/>
<img src="https://cloud.githubusercontent.com/assets/11892704/11050295/9745f78c-8712-11e5-8690-a9a04c4455fb.png" alt="Link to MyCancerGenome.org" /></li>
<li>Improved display of selection samples on the study view page</li>
</ul>
</li>
<li><strong>Improvements</strong>:
<ul>
<li>&quot;Enrichments&quot; analysis is now run across all genes</li>
<li>The &quot;Network&quot; tab is now using Cytoscape.js (Adobe Flash is no longer required)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-6-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-6-2015">#</doc-anchor-trigger>
        <span>October 6, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New TCGA data</strong>:
<ul>
<li><a href="https://www.cbioportal.org/study?id=brca_tcga_pub2015">Breast Invasive Carcinoma (TCGA, Cell 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=prad_tcga_pub">Prostate Adenocarcinoma (TCGA, in press)</a></li>
<li><a href="https://www.cbioportal.org/study?id=uvm_tcga">Uveal Melanoma (TCGA, Provisional)</a></li>
</ul>
</li>
<li><strong>Added data</strong> of 763 samples from 12 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=sclc_ucologne_2015">Small Cell Lung Cancer (U Cologne, Nature 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=ucs_jhu_2014">Uterine Carcinosarcoma (JHU, Nat Commun 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=paad_utsw_2015">Microdissected Pancreatic Cancer Whole Exome Sequencing (UTSW, Nat Commun 2015)</a></li>
<li><a href="https://www.cbioportal.org/https://www.cbioportal.org/study?id=panet_jhu_2011">Pancreatic Neuroendocrine Tumors (JHU, Science 2011)</a></li>
<li><a href="https://www.cbioportal.org/study?id=nccrcc_genentech_2014">Renal Non-Clear Cell Carcinoma (Genentech, Nat Genet 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=all_stjude_2015">Infant MLL-Rearranged Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=rms_nih_2014">Rhabdomyosarcoma (NIH, Cancer Discov 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=tet_nci_2014">Thymic epithelial tumors (NCI, Nat Genet 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=es_dfarber_broad_2014">Pediatric Ewing Sarcoma (DFCI, Cancer Discov 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=es_iocurie_2014">Ewing Sarcoma (Institut Cuire, Cancer Discov 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=cscc_dfarber_2015">Cutaneous squamous cell carcinoma (DFCI, Clin Cancer Res 2015)</a></li>
<li><a href="https://www.cbioportal.org/study?id=gbc_shanghai_2014">Gallbladder Carcinoma (Shanghai, Nat Genet 2014)</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="august-21-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-21-2015">#</doc-anchor-trigger>
        <span>August 21, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All <strong>TCGA data</strong> updated to the Firehose run of April 16, 2015.</li>
<li><strong>New feature</strong>: Enrichments Analysis finds alterations that are enriched in either altered or unaltered samples.</li>
<li><strong>Improvement</strong>: improved OncoPrint with better performance.</li>
</ul>
<doc-anchor-target id="june-3-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-3-2015">#</doc-anchor-trigger>
        <span>June 3, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>Improvements</strong>:
<ul>
<li>Allowed downloading data in each chart/table in study summary page.</li>
<li>Added log-rank test <em>p</em>-values to the survival plots in study summary page.</li>
<li>Improved visualization of patient clinical data in patient-centric view.</li>
<li>Added option to merge multiple samples for the same patient in OncoPrint.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="april-28-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-28-2015">#</doc-anchor-trigger>
        <span>April 28, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New features</strong>:
<ul>
<li>Redesigned query interface to allow selecting multiple cancer studies</li>
<li>Redesigned <strong>Plots</strong> tab</li>
</ul>
</li>
</ul>
<doc-anchor-target id="january-20-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-20-2015">#</doc-anchor-trigger>
        <span>January 20, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All <strong>TCGA data</strong> updated to the Firehose run of October 17, 2014</li>
<li><strong>COSMIC data</strong> updated to V71</li>
<li><strong>New features</strong>:
<ul>
<li>Query page: better search functions to find cancer studies</li>
<li>OncoPrints now support color coding of different mutation types</li>
<li>OncoPrints now support multiple clinical annotation tracks</li>
<li><a href="https://www.cbioportal.org/oncoprinter.jsp"><strong>OncoPrinter tool</strong></a> now supports mRNA expression changes<br />
<img src="https://user-images.githubusercontent.com/1334004/47188323-9f8e4b00-d305-11e8-8fe9-972104518545.png" alt="Oncoprint with multiple clinical tracks" /></li>
</ul>
</li>
</ul>
<doc-anchor-target id="january-6-2015">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-6-2015">#</doc-anchor-trigger>
        <span>January 6, 2015</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New feature</strong>: You can now view <strong>frequencies of mutations and copy-number alterations</strong> in the study view. These tables are updated dynamically when selecting subsets of samples.<br />
<img src="https://user-images.githubusercontent.com/1334004/47188333-9f8e4b00-d305-11e8-8c64-de5cbda0fa54.png" alt="Alterations in heavily copy-number altered endometrial cancer cases" /></li>
</ul>
<doc-anchor-target id="december-9-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-9-2014">#</doc-anchor-trigger>
        <span>December 9, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li><strong>New TCGA data</strong>:
<ul>
<li>Added complete and up-to-date <strong>clinical data</strong> for all <strong>TCGA</strong> provisional studies</li>
<li>All TCGA data updated to the Firehose run of July 15, 2014</li>
<li>New TCGA provisional studies: Esophageal cancer, Pheochromocytoma and Paraganglioma (PCPG)</li>
<li>New published TCGA studies: <a href="https://www.cbioportal.org/study?id=thca_tcga_pub">Thyroid Cancer</a> and <a href="https://www.cbioportal.org/study?id=kich_tcga_pub">Kidney Chromophobe</a></li>
</ul>
</li>
<li><strong>Added data</strong> of 172 samples from 4 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=chol_nus_2012">Cholangiocarcinoma (National University of Singapore, Nature Genetics 2012)</a></li>
<li><a href="https://www.cbioportal.org/study?id=chol_nccs_2013">Cholangiocarcinoma (National Cancer Centre of Singapore, Nature Genetics 2013)</a></li>
<li><a href="https://www.cbioportal.org/study?id=chol_jhu_2013">Intrahepatic Cholangiocarcinoma (Johns Hopkins University, Nature Genetics 2013)</a></li>
<li><a href="https://www.cbioportal.org/study?id=blca_mskcc_solit_2014">Bladder Cancer (MSKCC, Eur Urol 2014)</a></li>
</ul>
</li>
<li><strong>New features</strong>:
<ul>
<li>Redesigned <strong>Mutual Exclusivity</strong> tab</li>
<li>Added <strong>correlation scores</strong> for scatter plots on the Plots tab</li>
<li>Download links to <a href="https://www.genomespace.org/"><strong>GenomeSpace</strong></a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-24-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-24-2014">#</doc-anchor-trigger>
        <span>October 24, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Added data of 885 samples from 11 published studies:
<ul>
<li><a href="https://www.cbioportal.org/study?id=coadread_mskcc">Colorectal Adenocarcinoma Triplets (MSKCC, Genome Biology 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=escc_icgc">Esophageal Squamous Cell Carcinoma (ICGC, Nature 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=mpnst_mskcc">Malignant Peripheral Nerve Sheath Tumor (MSKCC, Nature Genetics 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=skcm_broad_dfarber">Melanoma (Broad/Dana Farber, Nature 2012)</a></li>
<li><a href="https://www.cbioportal.org/study?id=npc_nusingapore">Nasopharyngeal Carcinoma (National University Singapore, Nature Genetics 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=prad_mskcc_2014">Prostate Adenocarcinoma CNA study (MSKCC, PNAS 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=prad_mskcc_cheny1_organoids_2014">Prostate Adenocarcinoma Organoids (MSKCC, Cell 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=stad_tcga_pub">Stomach Adenocarcinoma (TCGA, Nature 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=stad_pfizer_uhongkong">Stomach Adenocarcinoma (Pfizer and University of Hong Kong, Nature Genetics 2014)</a></li>
<li><a href="https://www.cbioportal.org/study?id=stad_uhongkong">Stomach Adenocarcinoma (University of Hong Kong, Nature Genetics 2011)</a></li>
<li><a href="https://www.cbioportal.org/study?id=stad_utokyo">Stomach Adenocarcinoma (University of Tokyo, Nature Genetics 2014)</a></li>
</ul>
</li>
</ul>
<doc-anchor-target id="august-8-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#august-8-2014">#</doc-anchor-trigger>
        <span>August 8, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Released two new tools
<ul>
<li><a href="https://www.cbioportal.org/oncoprinter.jsp">Oncoprinter</a> lets you create Oncoprints from your own, custom data</li>
<li><a href="https://www.cbioportal.org/mutation_mapper.jsp">MutationMapper</a> draws mutation diagrams (lollipop plots) from your custom data</li>
</ul>
</li>
</ul>
<doc-anchor-target id="may-21-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-21-2014">#</doc-anchor-trigger>
        <span>May 21, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the Firehose run of April 16, 2014</li>
</ul>
<doc-anchor-target id="may-12-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-12-2014">#</doc-anchor-trigger>
        <span>May 12, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Improved study summary page including survival analysis based on clinical attributes<br />
e.g. <a href="https://www.cbioportal.org/study?id=ucec_tcga_pub">TCGA Endometrial Cancer cohort</a><br />
<img src="https://user-images.githubusercontent.com/1334004/47188334-a026e180-d305-11e8-964c-6c417bb06ca2.png" alt="Study view" /></li>
</ul>
<doc-anchor-target id="march-27-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-27-2014">#</doc-anchor-trigger>
        <span>March 27, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>New features:
<ul>
<li>Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)</li>
<li>Gene expression correlation analysis (find all genes with expression correlation to your query genes)</li>
<li>The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation</li>
<li>The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation<br />
<img src="https://user-images.githubusercontent.com/1334004/47188325-9f8e4b00-d305-11e8-96db-67936cd89f8f.png" alt="3D viewer &amp; Co-expression" /></li>
</ul>
</li>
</ul>
<doc-anchor-target id="march-18-2014">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-18-2014">#</doc-anchor-trigger>
        <span>March 18, 2014</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the Firehose run of January 15, 2014</li>
<li>Updated to the latest COSMIC data (v68)</li>
<li>Added two new provisional TCGA studies:
<ul>
<li>Adrenocortical Carcinoma</li>
<li>Uterine Carcinosarcoma</li>
</ul>
</li>
<li>Added mutation data of 898 samples from 11 published studies:
<ul>
<li>Hepatocellular Carcinoma (RIKEN, Nature Genetics 2012)</li>
<li>Hepatocellular Carcinoma (AMC, Hepatology in press)</li>
<li>Medulloblastoma (Broad, Nature 2012)</li>
<li>Medulloblastoma (ICGC, Nature 2012)</li>
<li>Medulloblastoma (PCGP, Nature 2012)</li>
<li>Multiple Myeloma (Broad, Cancer Cell 2014)</li>
<li>Pancreatic Adenocarcinoma (ICGC, Nature 2012)</li>
<li>Small Cell Carcinoma of the Ovary (MSKCC, Nature Genetics in press)</li>
<li>Small Cell Lung Cancer (CLCGP, Nature Genetics 2012)</li>
<li>Small Cell Lung Cancer (Johns Hopkins, Nature Genetics 2012)</li>
<li>NCI-60 Cell Lines (NCI, Cancer Res. 2012)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="december-9-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-9-2013">#</doc-anchor-trigger>
        <span>December 9, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Added mutation data of 99 bladder cancer samples (BGI, Nature Genetics 2013)</li>
</ul>
<doc-anchor-target id="december-6-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-6-2013">#</doc-anchor-trigger>
        <span>December 6, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Data sets matching four recently submitted or published TCGA studies are now available
<ul>
<li>Glioblastoma (Cell 2013)</li>
<li>Bladder carcinoma (Nature, in press)</li>
<li>Head &amp; neck squamous cell carcinoma (submitted)</li>
<li>Lung adenocarcinoma (submitted)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="november-8-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-8-2013">#</doc-anchor-trigger>
        <span>November 8, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the Firehose run of September 23, 2013.</li>
<li>Updated to the latest COSMIC data (v67).</li>
<li>Added mutation data of 792 samples from 9 published cancer studies:
<ul>
<li>Esophageal Adenocarcinoma (Broad, Nature Genetics 2013)</li>
<li>Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)</li>
<li>Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)</li>
<li>Kidney Renal Clear Cell Carcinoma (BGI, Nature Genetics 2012)</li>
<li>Prostate Adenocarcinoma, Metastatic (Michigan, Nature 2012)</li>
<li>Prostate Adenocarcinoma (Broad/Cornell, Nature Genetics 2012)</li>
<li>Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)</li>
<li>Skin Cutaneous Melanoma (Yale, Nature Genetics 2012)</li>
<li>Skin Cutaneous Melanoma (Broad, Cell 2012)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="october-21-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-21-2013">#</doc-anchor-trigger>
        <span>October 21, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Improved interface for survival plots, including information on individual samples via mouse-over</li>
<li>New fusion glyph in OncoPrints    <a href="https://www.cbioportal.org/index.do?cancer_study_id=hnsc_tcga_pub&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;case_set_id=hnsc_tcga_pub_sequenced&amp;case_ids=&amp;gene_set_choice=user-defined-list&amp;gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&amp;clinical_param_selection=null&amp;tab_index=tab_visualize&amp;Action=Submit"><img src="https://user-images.githubusercontent.com/1334004/47188320-9f8e4b00-d305-11e8-9d8c-09a7ec28d921.png" alt="FGFR3 fusions in head and neck carcinoma" /></a></li>
<li>Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
<img src="https://user-images.githubusercontent.com/1334004/47188319-9ef5b480-d305-11e8-9281-fb31993b7b42.png" alt="Cross Cancer Query" /></li>
</ul>
<doc-anchor-target id="september-9-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-9-2013">#</doc-anchor-trigger>
        <span>September 9, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Updated COSMIC data (v66 Release)</li>
<li>Improved / interactive visualization on the &quot;Protein changes&quot; tab</li>
<li>Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters</li>
<li>Addition of DNA cytoband information in the patient view of copy-number changes</li>
<li>OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
<img src="https://user-images.githubusercontent.com/1334004/47188326-9f8e4b00-d305-11e8-9cf5-52e459d56310.png" alt="Oncoprint with clinical track" /></li>
</ul>
<doc-anchor-target id="july-25-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-25-2013">#</doc-anchor-trigger>
        <span>July 25, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Multi-gene correlation plots.</li>
<li>Variant allele frequency distribution plots for individual tumor samples.</li>
<li>Tissue images for TCGA samples in the patient view, via <a href="https://cancer.digitalslidearchive.net/">Digital Slide Archive</a>. <a href="https://www.cbioportal.org/patient?studyId=ucec_tcga&amp;caseId=TCGA-BK-A0CC&amp;tab=tissueImageTab">Example</a>.</li>
</ul>
<doc-anchor-target id="july-16-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-16-2013">#</doc-anchor-trigger>
        <span>July 16, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the May Firehose run (May 23, 2013).
<ul>
<li>TCGA Pancreatic Cancer study (provisional) added.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="july-4-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-4-2013">#</doc-anchor-trigger>
        <span>July 4, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Improved rendering of mutation diagrams, including ability to download in PDF format.</li>
<li>Improved home page: Searchable cancer study &amp; gene set selectors, data sets selector.</li>
</ul>
<doc-anchor-target id="june-17-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-17-2013">#</doc-anchor-trigger>
        <span>June 17, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Improved interface for correlation plots, including information on individual samples via mouse-over.</li>
<li>Gene Details from Biogene are now available in the Network view.</li>
<li>Added mutation and copy number data from a new adenoid cystic carcinoma study: Ho et al., Nature Genetics 2013.</li>
<li>Added mutation data from 6 cancer studies.
<ul>
<li>Breast Invasive Carcinoma (Shah et al., Nature 2012)</li>
<li>Breast Invasive Carcinoma (Banerji et al., Nature 2012)</li>
<li>Breast Invasive Carcinoma (Stephens et al., Nature 2012)</li>
<li>Lung Adenocarcinoma (Imielinksi et al., Cell 2012)</li>
<li>Lung Adenocarcinoma (Ding et al., Nature 2008)</li>
<li>Colorectal Cancer (Seshagiri et al., Nature 2012)</li>
</ul>
</li>
</ul>
<doc-anchor-target id="june-4-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#june-4-2013">#</doc-anchor-trigger>
        <span>June 4, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the April Firehose run (April 21, 2012).</li>
</ul>
<doc-anchor-target id="may-14-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-14-2013">#</doc-anchor-trigger>
        <span>May 14, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Added a published TCGA study: Acute Myeloid Leukemia (TCGA, NEJM 2013).</li>
</ul>
<doc-anchor-target id="april-28-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-28-2013">#</doc-anchor-trigger>
        <span>April 28, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the March Firehose run (March 26, 2012).</li>
<li>mRNA percentiles for altered genes shown in patient view.</li>
</ul>
<doc-anchor-target id="april-2-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#april-2-2013">#</doc-anchor-trigger>
        <span>April 2, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the February Firehose run (February 22, 2012).</li>
</ul>
<doc-anchor-target id="march-28-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-28-2013">#</doc-anchor-trigger>
        <span>March 28, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All TCGA data updated to the January Firehose run (January 16, 2012).</li>
<li>Data from a new bladder cancer study from MSKCC has been added (97 samples, Iyer et al., JCO in press).</li>
</ul>
<doc-anchor-target id="february-16-2013">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-16-2013">#</doc-anchor-trigger>
        <span>February 16, 2013</span>
    </h2>
</doc-anchor-target>
<ul>
<li>The cBio Portal now contains mutation data from all provisional TCGA projects. Please adhere to <a href="https://cancergenome.nih.gov/abouttcga/policies/publicationguidelines">the TCGA publication guidelines</a> when using these and any TCGA data in your publications.</li>
<li>All data updated to the October Firehose run (October 24, 2012).</li>
<li><strong>Sequencing read counts and frequencies</strong> are now shown in the Mutation Details table when available.</li>
<li>Improved OncoPrints, resulting in performance improvements.</li>
</ul>
<doc-anchor-target id="november-21-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-21-2012">#</doc-anchor-trigger>
        <span>November 21, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Major new feature: Users can now visualize <strong>genomic alterations and clinical data of individual tumors</strong>, including:
<ul>
<li>Summary of <strong>mutations</strong> and <strong>copy-number alterations</strong> of interest</li>
<li><strong>Clinical trial</strong> information</li>
<li>TCGA <strong>Pathology Reports</strong></li>
</ul>
</li>
<li>New <strong>cancer summary view</strong> (Example <a href="https://www.cbioportal.org/study?id=ucec_tcga">Endometrial Cancer</a>)</li>
<li><strong>Updated drug data</strong> from KEGG DRUG and NCI Cancer Drugs (aggregated by <a href="https://bitbucket.org/armish/pihelper">PiHelper</a>)</li>
</ul>
<doc-anchor-target id="october-22-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#october-22-2012">#</doc-anchor-trigger>
        <span>October 22, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All data updated to the <strong>Broad Firehose</strong> run from July 25, 2012.</li>
<li><strong>COSMIC data</strong> added to Mutation Details (via Oncotator).</li>
<li>All predicted functional impact scores are updated to <strong>Mutation Assessor 2.0</strong>.</li>
<li>Users can now base queries on genes in recurrent regions of copy-number alteration (from <strong>GISTIC</strong> via Firehose).</li>
<li>The <a href="https://www.cbioportal.org/onco_query_lang_desc.jsp">Onco Query Language (OQL)</a> now supports queries for specific mutations or mutation types.</li>
<li>Data sets added that match the data of all TCGA publications (GBM, ovarian, colorectal, and lung squamous).</li>
</ul>
<doc-anchor-target id="july-18-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#july-18-2012">#</doc-anchor-trigger>
        <span>July 18, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Mutation data for the TCGA lung squamous cell carcinoma and breast cancer projects (manuscripts in press at Nature).</li>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (May 25, 2012).</li>
<li><strong>Drug information</strong> added to the network view (via Drugbank).</li>
<li><strong>Improved cross-cancer</strong> queries: Option to select data types, export of summary graphs.</li>
<li>Users can now base queries on frequently mutated genes (from <strong>MutSig</strong> via Firehose).</li>
</ul>
<doc-anchor-target id="may-16-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#may-16-2012">#</doc-anchor-trigger>
        <span>May 16, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (March 21, 2012).</li>
<li>Extended <strong>cross-cancer</strong> functionality, enabling users to query across all cancer studies in our database.</li>
<li>New <strong>&quot;build a case&quot;</strong> functionality, enabling users to generate custom case sets, based on one or more clinical attributes.</li>
<li>New OncoPrint features, including more <strong>compact OncoPrints</strong>, and support for <strong>RPPA</strong> visualization.</li>
</ul>
<doc-anchor-target id="february-27-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#february-27-2012">#</doc-anchor-trigger>
        <span>February 27, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (January 24, 2012).</li>
<li>Validated mutation data for colorectal cancer.</li>
<li>New feature: <strong>Mutation Diagrams</strong> that show mutations in the context of protein domains.
<img src="https://user-images.githubusercontent.com/1334004/47188347-a0bf7800-d305-11e8-8c92-7c8763711338.png" alt="TP53 Mutations in Ovarian Cancer" /></li>
</ul>
<doc-anchor-target id="january-30-2012">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#january-30-2012">#</doc-anchor-trigger>
        <span>January 30, 2012</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Updated data for several TCGA cancer studies.</li>
<li>Some small bug-fixes.</li>
</ul>
<doc-anchor-target id="december-22-2011">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-22-2011">#</doc-anchor-trigger>
        <span>December 22, 2011</span>
    </h2>
</doc-anchor-target>
<ul>
<li><p><strong>Fourteen new <a href="https://cancergenome.nih.gov/">TCGA</a> cancer studies</strong>: This includes complete data for TCGA Colorectal Carcinoma and provisional data for thirteen other cancer types in the TCGA production pipeline. Please note that data from these thirteen new cancer types are provisional, not final and do not yet include mutation data. As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.
<img src="https://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/TCGA_54px-Logo.png" alt="TCGA" /></p>
</li>
<li><p><strong>Four new data types</strong>:</p>
<ul>
<li>Reverse-phase protein array (RPPA) data.</li>
<li>microRNA expression and copy-number (including support for multiple loci)</li>
<li>RNA-Seq based expression data.</li>
<li>log2 copy-number data.</li>
</ul>
</li>
<li><p>Updated TCGA GBM copy-number, expression, and methylation data.</p>
</li>
<li><p>New <strong>gene symbol validation</strong> service. You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</p>
</li>
<li><p><strong>Links to IGV</strong> for visualization of DNA copy-number changes.</p>
</li>
<li><p>Background information from the <a href="https://www.sanger.ac.uk/genetics/CGP/Census/">Sanger Cancer Gene Census</a>.</p>
</li>
<li><p>Two <strong>new <a href="https://www.cbioportal.org/tutorials">Tutorials</a></strong> to get you quickly started in using the portal.</p>
</li>
</ul>
<doc-anchor-target id="november-14-2011">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-14-2011">#</doc-anchor-trigger>
        <span>November 14, 2011</span>
    </h2>
</doc-anchor-target>
<ul>
<li>New and <strong>improved mutation details</strong>, with sorting and filtering capabilities.</li>
<li>In collaboration with Bilkent University, we have added a <strong>new Network tab</strong> to our results pages. The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="https://www.pathwaycommons.org">Pathway Commons</a>.
<img src="https://user-images.githubusercontent.com/1334004/47188328-9f8e4b00-d305-11e8-87d7-6336285f7572.png" alt="GBM Network" /></li>
</ul>
<doc-anchor-target id="september-3-2011">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#september-3-2011">#</doc-anchor-trigger>
        <span>September 3, 2011</span>
    </h2>
</doc-anchor-target>
<ul>
<li>You can now query across different cancer studies (feature available directly from the home page).</li>
<li>Our <a href="https://www.cbioportal.org/cgds_r.jsp">MATLAB CGDS Cancer Genomics Toolbox</a> is now available. The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="https://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>. If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<doc-anchor-target id="march-2-2011">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#march-2-2011">#</doc-anchor-trigger>
        <span>March 2, 2011</span>
    </h2>
</doc-anchor-target>
<p>New plotting features and other improvements:</p>
<ul>
<li>Correlation plots that show the relationship between different data types for individual genes.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Updated <a href="https://www.cbioportal.org/cgds_r.jsp">R Package</a> with support for correlation plots and general improvements for retrieving and accessing data in R data frames.</li>
<li>The <a href="https://www.cbioportal.org/web_api.jsp">Web Interface</a> now supports basic clinical data, e.g. survival data.</li>
<li><a href="https://www.cbioportal.org/networks.jsp">Networks</a> for pathway analysis are now available for download.
<img src="https://user-images.githubusercontent.com/1334004/47188329-9f8e4b00-d305-11e8-83e2-0ad816431888.png" alt="Survival Analysis" /></li>
</ul>
<doc-anchor-target id="december-15-2010">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#december-15-2010">#</doc-anchor-trigger>
        <span>December 15, 2010</span>
    </h2>
</doc-anchor-target>
<p>Several new features, including:</p>
<ul>
<li>Redesigned and streamlined user interface, based on user feedback and usability testing.</li>
<li>Advanced support for gene-specific alterations. For example, users can now view mutations within TP53, and ignore copy number alterations, or only view amplifications of EGFR, and ignore deletions.</li>
<li>Improved performance.</li>
<li><a href="faq">Frequently Asked Questions</a> document released.</li>
<li>Updated <del>Video Tutorial</del> (update: old link no longer functional. Now see: <a href="https://www.youtube.com/results?search_query=cbioportal">YouTube</a></li>
</ul>
<doc-anchor-target id="november-4-2010">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#november-4-2010">#</doc-anchor-trigger>
        <span>November 4, 2010</span>
    </h2>
</doc-anchor-target>
<ul>
<li>Enhanced <a href="https://www.cbioportal.org/faq.jsp#what-are-oncoprints">Oncoprints</a>, enabling users to quickly visualize genomic alterations across many cases. Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.</li>
<li>Official release of our <a href="https://www.cbioportal.org/web_api.jsp">Web Interface</a>, enabling programmatic access to all data.</li>
<li>Official release of our <a href="https://www.cbioportal.org/cgds_r.jsp">R Package</a>, enabling programmatic access to all data from the R platform for statistical computing.
<img src="https://user-images.githubusercontent.com/1334004/47188322-9f8e4b00-d305-11e8-9fae-ca188de91267.png" alt="OncoPrints" /></li>
</ul>

                                
                                <!-- Required only on API pages -->
                                <doc-toolbar-member-filter-no-results></doc-toolbar-member-filter-no-results>
                            </div>
                            <footer class="clear-both">
                                <div class="flex flex-wrap items-center justify-between mt-14">
                                    <a class="my-2.5 inline-flex items-center text-sm whitespace-nowrap text-blue-500 dark:text-blue-400 hover:text-blue-700 hover:underline" href="https://github.com/cbioportal/cbioportal/blob/master/docs/News.md" target="_blank" rel="noopener">
                                        <svg class="mr-1.5" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M20 12c-.55 0-1 .45-1 1v7c0 .55-.45 1-1 1H4c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h7c.55 0 1-.45 1-1s-.45-1-1-1H4C2.35 3 1 4.35 1 6v14c0 1.65 1.35 3 3 3h14c1.65 0 3-1.35 3-3v-7c0-.55-.45-1-1-1z" /><path d="M22.21 1.79c-1.18-1.18-3.24-1.18-4.41 0l-9.5 9.5c-.13.13-.22.29-.26.46l-1 4c-.08.34.01.7.26.95.18.2.44.3.7.3.08 0 .16-.01.24-.03l4-1c.18-.04.34-.13.46-.26l9.5-9.5c1.22-1.22 1.22-3.2.01-4.42zm-1.42 3l-9.3 9.3-2.11.53.53-2.11 9.3-9.3c.42-.42 1.16-.42 1.59 0 .43.43.43 1.15-.01 1.58z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        <span>Edit this page</span>
                                    </a>
                                </div>
                            
                                <nav class="print:hidden flex mt-14">
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 h-full flex items-center break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-l-lg transition-colors duration-150 relative hover:z-5" href="../user-guide/oql/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mr-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19 11H7.41l5.29-5.29a.996.996 0 10-1.41-1.41l-7 7a1 1 0 000 1.42l7 7a1.024 1.024 0 001.42-.01.996.996 0 000-1.41L7.41 13H19c.55 0 1-.45 1-1s-.45-1-1-1z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                            <span>
                                                <span class="block text-xs font-normal text-gray-400 dark:text-dark-400">Previous</span>
                                                <span class="block mt-1">Onco Query Language (OQL)</span>
                                            </span>
                                        </a>
                                    </div>
                            
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 -mx-px h-full flex items-center justify-end break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-r-lg transition-colors duration-150 relative hover:z-5" href="../news-genie/">
                                            <span>
                                                <span class="block text-xs font-normal text-right text-gray-400 dark:text-dark-400">Next</span>
                                                <span class="block mt-1">Genie News</span>
                                            </span>
                                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19.92 12.38a1 1 0 00-.22-1.09l-7-7a.996.996 0 10-1.41 1.41l5.3 5.3H5c-.55 0-1 .45-1 1s.45 1 1 1h11.59l-5.29 5.29a.996.996 0 000 1.41c.19.2.44.3.7.3s.51-.1.71-.29l7-7c.09-.09.16-.21.21-.33z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        </a>
                                    </div>
                                </nav>
                            </footer>
                        </main>
                
                        <div class="print:border-none border-t dark:border-dark-650 pt-6 mb-8">
                            <footer class="flex flex-wrap items-center justify-between print:justify-center">
                                <div class="print:hidden">
                                    <ul class="flex flex-wrap items-center text-sm">
                                    </ul>
                                </div>
                                <div class="print:justify-center docs-copyright py-2 text-gray-500 dark:text-dark-350 text-sm leading-relaxed"><p>© Copyright 2025. All rights reserved.</p></div>
                            </footer>
                        </div>
                    </div>
                
                    <!-- Rendered if sidebar right is enabled -->
                    <!-- Sidebar right skeleton-->
                    <div v-cloak class="fixed top-0 bottom-0 right-0 translate-x-full bg-white border-gray-200 lg:sticky lg:border-l lg:shrink-0 lg:pt-6 lg:transform-none sm:w-1/2 lg:w-64 lg:z-0 md:w-104 sidebar-right skeleton dark:bg-dark-850 dark:border-dark-650">
                        <div class="pl-5">
                            <div class="w-32 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-48 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-40 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                        </div>
                    </div>
                
                    <!-- User should be able to hide sidebar right -->
                    <doc-sidebar-right v-cloak></doc-sidebar-right>
                </div>

            </div>
        </div>
    
        <doc-search-mobile></doc-search-mobile>
        <doc-back-to-top></doc-back-to-top>
    </div>


    <div id="docs-overlay-target"></div>

    <script data-cfasync="false">window.__DOCS__ = { "title": "News", level: 1, icon: "file", hasPrism: false, hasMermaid: false, hasMath: false, tocDepth: 23 }</script>
</body>
</html>
