 [TOC]

# What is the cBio SU2 Cancer Genomics Portal?

The cBio SU2C Cancer Genomics Portal currently contains data from all women's cancer projects run by the TCGA. The portal has all the same features as the public [cBio Cancer Genomics Portal](http://cbioportal.org/), but contains additional protein data for some genes and tumor samples.

# What are the main functionalities of the cBio SU2C Cancer Genomics Portal?

The portal stores genomic data from large scale, integrated cancer genomic data sets. It allows explorative data analysis (e.g.: Is my gene of interest altered in a specific cancer type? Do mutations of BRCA1 and BRCA2 in ovarian cancer co-occur?) and provides simple download of small data slices (user-defined gene and sample sets, no need to download entire data sets).

# What data types are in the portal?

The portal currently stores DNA copy-number data (putative, discrete values per gene, e.g. "homozygously deleted" or "amplified"), mRNA and microRNA expression data, non-synonymous mutations, and clinical data.

# What kind of clinical data is stored in the portal?

The portal currently stores overall and disease-free survival data, when available.

# Does the portal contain methylation data?

Yes, the portal currently contains methylation data for ovarian cancer, and we are working on adding data for the other cancer types.

# Does the portal store raw or probe-level data?

No, the portal only contains gene-level data. Data for different isoforms of a given gene are merged. Raw and probe-level data for all date sets is available via NCBI GEO or through the TCGA Data Portal. See the cancer type description on the main query page for links to the raw data.

# What are Oncoprints?

OncoPrints are compact means of visualizing distinct genomic alterations, including somatic mutations, copy number alterations, and mRNA expression changes across a set of cases.  They are extremely useful for visualizing gene set and pathway alterations across a set of cases, and for visually identifying trends, such as trends in mutual exclusivity or co-occurence between gene pairs within a gene set. Individual genes are represented as rows, and individual cases or patients are represented as columns. ![Example OncoPrint](images/oncoprint_example.png)

# Can I change the order of genes in the Oncoprint?

The order of genes in the Oncoprint is determined by the order entered into the initial query field. Simply change the initial gene order, resubmit your query, and the change will be reflected in the Oncoprint.

# Does the portal work on all browsers and operating systems?

The portal should work on all major browsers and operating systems. We have tested it using current versions of Firefox, Safari, and Chrome, in Mac OS X and Windows. However, SVG images, which are used for downloadable oncoprints, are not supported by Internet Explorer versions older than 9. If you notice any other incompatibilities, please let us know.

# What are mRNA and microRNA Z-Scores?

For mRNA and microRNA expression data, we typically compute the relative expression of an individual gene and tumor to the gene's expression distribution in a reference population. That reference population is either all tumors that are diploid for the gene in question, or, when available, normal adjacent tissue. The returned value indicates the number of standard deviations away from the mean of expression in the reference population (Z-score). This measure is useful to determine whether a gene is up- or down-regulated relative to the normal samples or all other tumor samples.

# What do "-2", "-1", "0", "1", and "2" mean in the copy-number data?

These levels are derived from the copy-number analysis algorithms GISTIC or RAE, and indicate the copy-number level per gene. "-2" is a deep loss, possibly a homozygous deletion, "-1" is a single-copy loss (heterozygous deletion), "0" is diploid, "1" indicates a low-level gain, and "2" is a high-level amplification. Note that these calls are putative.

# What is GISTIC?  What is RAE?

Copy number data sets within the portal are generated by [GISTIC](http://www.ncbi.nlm.nih.gov/sites/entrez?term=18077431) or [RAE](http://www.ncbi.nlm.nih.gov/sites/entrez?term=18784837)
algorithms.  Both algorithms attempt to identify significantly altered regions of amplification or deletion across sets of patients. Both algorithms also generate putative gene/patient copy number specific calls, which are then input into the portal.  
