<h1 id="march-27-2014">March 27, 2014</h1>
<ul>
<li>New features:<ul>
<li>Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)</li>
<li>Gene expression correlation analysis (find all genes with expression correlation to your query genes)</li>
<li>The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation</li>
<li>The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation
<br/><img alt="3D viewer &amp; Co-expression" src="images/previews/news_3d_coexp.png" /></li>
</ul>
</li>
</ul>
<h1 id="march-18-2014">March 18, 2014</h1>
<ul>
<li>All TCGA data updated to the Firehose run of January 15, 2014</li>
<li>Updated to the latest COSMIC data (v68)</li>
<li>Added two new provisional TCGA studies:<ul>
<li>Adrenocortical Carcinoma</li>
<li>Uterine Carcinosarcoma</li>
</ul>
</li>
<li>Added mutation data of 898 samples from 11 published studies:<ul>
<li>Hepatocellular Carcinoma (RIKEN, Nature Genetics 2012)</li>
<li>Hepatocellular Carcinoma (LGGM, Hepatology in press)</li>
<li>Medulloblastoma (Broad, Nature 2012)</li>
<li>Medulloblastoma (ICGC, Nature 2012)</li>
<li>Medulloblastoma (WashU, Nature 2012)</li>
<li>Multiple Myeloma (Broad, Cancer Cell 2014)</li>
<li>Pancreatic Adenocarcinoma (ICGC, Nature 2012)</li>
<li>Small Cell Carcinoma of the Ovary (MSKCC, Nature Genetics in press)</li>
<li>Small Cell Lung Cancer (CLCGP, Nature Genetics 2012)</li>
<li>Small Cell Lung Cancer (Johns Hopkins, Nature Genetics 2012)</li>
<li>NCI-60 Cell Lines (NCI, Cancer Res. 2012)</li>
</ul>
</li>
</ul>
<h1 id="december-9-2013">December 9, 2013</h1>
<ul>
<li>Added mutation data of 99 bladder cancer samples (BGI, Nature Genetics 2013)</li>
</ul>
<h1 id="december-6-2013">December 6, 2013</h1>
<ul>
<li>Data sets matching four recently submitted or published TCGA studies are now available<ul>
<li>Glioblastoma (Cell 2013)</li>
<li>Bladder carcinoma (Nature, in press)</li>
<li>Head &amp; neck squamous cell carcinoma (submitted)</li>
<li>Lung adenocarcinoma (submitted)</li>
</ul>
</li>
</ul>
<h1 id="november-8-2013">November 8, 2013</h1>
<ul>
<li>All TCGA data updated to the Firehose run of September 23, 2013.</li>
<li>Updated to the latest COSMIC data (v67).</li>
<li>Added mutation data of 792 samples from 9 published cancer studies:<ul>
<li>Esophageal Adenocarcinoma (Broad, Nature Genetics 2013)</li>
<li>Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)</li>
<li>Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)</li>
<li>Kidney Renal Clear Cell Carcinoma (BGI, Nature Genetics 2012)</li>
<li>Prostate Adenocarcinoma, Metastatic (Michigan, Nature 2012)</li>
<li>Prostate Adenocarcinoma (Broad/Cornell, Nature Genetics 2012)</li>
<li>Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)</li>
<li>Skin Cutaneous Melanoma (Yale, Nature Genetics 2012)</li>
<li>Skin Cutaneous Melanoma (Broad, Cell 2012)</li>
</ul>
</li>
</ul>
<h1 id="october-21-2013">October 21, 2013</h1>
<ul>
<li>Improved interface for survival plots, including information on individual samples via mouse-over</li>
<li>New fusion glyph in OncoPrints &nbsp;&nbsp;&nbsp;<a href="index.do?cancer_study_id=hnsc_tcga_pub&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;case_set_id=hnsc_tcga_pub_sequenced&amp;case_ids=&amp;gene_set_choice=user-defined-list&amp;gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&amp;clinical_param_selection=null&amp;tab_index=tab_visualize&amp;Action=Submit"><img alt="FGFR3 fusions in head and neck carcinoma" src="images/previews/fusion-in-oncoprint.png" /></a></li>
<li>Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
<center><img alt="Cross Cancer Query" src="images/previews/cross_cancer.png" /></center></li>
</ul>
<h1 id="september-9-2013">September 9, 2013</h1>
<ul>
<li>Updated COSMIC data (v66 Release)</li>
<li>Improved / interactive visualization on the "Protein changes" tab</li>
<li>Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters</li>
<li>Addition of DNA cytoband information in the patient view of copy-number changes</li>
<li>OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
<center><img alt="Oncoprint with clinical track" src="images/previews/oncoprint_clinical_track.png" /></center></li>
</ul>
<h1 id="july-25-2013">July 25, 2013</h1>
<ul>
<li>Multi-gene correlation plots.</li>
<li>Variant allele frequency distribution plots for individual tumor samples.</li>
<li>Tissue images for TCGA samples in the patient view, via <a href="http://cancer.digitalslidearchive.net/">Digital Slide Archive</a>. <a href="case.do?cancer_study_id=ucec_tcga&amp;case_id=TCGA-BK-A0CC#images">Example</a>.</li>
</ul>
<h1 id="july-16-2013">July 16, 2013</h1>
<ul>
<li>All TCGA data updated to the May Firehose run (May 23, 2013).<ul>
<li>TCGA Pancreatic Cancer study (provisional) added.</li>
</ul>
</li>
</ul>
<h1 id="july-4-2013">July 4, 2013</h1>
<ul>
<li>Improved rendering of mutation diagrams, including ability to download in PDF format.</li>
<li>Improved home page: Searchable cancer study &amp; gene set selectors, data sets selector.</li>
</ul>
<h1 id="june-17-2013">June 17, 2013</h1>
<ul>
<li>Improved interface for correlation plots, including information on individual samples via mouse-over.</li>
<li>Gene Details from Biogene are now available in the Network view.</li>
<li>Added mutation and copy number data from a new adenoid cystic carcinoma study: Ho et al., Nature Genetics 2013.</li>
<li>Added mutation data from 6 cancer studies.<ul>
<li>Breast Invasive Carcinoma (Shah et al., Nature 2012)</li>
<li>Breast Invasive Carcinoma (Banerji et al., Nature 2012)</li>
<li>Breast Invasive Carcinoma (Stephens et al., Nature 2012)</li>
<li>Lung Adenocarcinoma (Imielinksi et al., Cell 2012)</li>
<li>Lung Adenocarcinoma (Ding et al., Nature 2008)</li>
<li>Colorectal Cancer (Seshagiri et al., Nature 2012)</li>
</ul>
</li>
</ul>
<h1 id="june-4-2013">June 4, 2013</h1>
<ul>
<li>All TCGA data updated to the April Firehose run (April 21, 2012).</li>
</ul>
<h1 id="may-14-2013">May 14, 2013</h1>
<ul>
<li>Added a published TCGA study: Acute Myeloid Leukemia (TCGA, NEJM 2013).</li>
</ul>
<h1 id="april-28-2013">April 28, 2013</h1>
<ul>
<li>All TCGA data updated to the March Firehose run (March 26, 2012).</li>
<li>mRNA percentiles for altered genes shown in patient view.</li>
</ul>
<h1 id="april-2-2013">April 2, 2013</h1>
<ul>
<li>All TCGA data updated to the February Firehose run (February 22, 2012).</li>
</ul>
<h1 id="march-28-2013">March 28, 2013</h1>
<ul>
<li>All TCGA data updated to the January Firehose run (January 16, 2012).</li>
<li>Data from a new bladder cancer study from MSKCC has been added (97 samples, Iyer et al., JCO in press).</li>
</ul>
<h1 id="february-16-2013">February 16, 2013</h1>
<ul>
<li>The cBio Portal now contains mutation data from all provisional TCGA projects. Please adhere to <a href="http://cancergenome.nih.gov/abouttcga/policies/publicationguidelines">the TCGA publication guidelines</a> when using these and any TCGA data in your publications.</li>
<li>All data updated to the October Firehose run (October 24, 2012).</li>
<li><strong>Sequencing read counts and frequencies</strong> are now shown in the Mutation Details table when available.</li>
<li>Improved OncoPrints, resulting in performance improvements.</li>
</ul>
<h1 id="november-21-2012">November 21, 2012</h1>
<ul>
<li>Major new feature: Users can now visualize <strong>genomic alterations and clinical data of individual tumors</strong>, including:<ul>
<li>Summary of <strong>mutations</strong> and <strong>copy-number alterations</strong> of interest</li>
<li><strong>Clinical trial</strong> information</li>
<li>TCGA <strong>Pathology Reports</strong></li>
</ul>
</li>
<li>New <strong>cancer summary view</strong> (Example <a href="study.do?cancer_study_id=ucec_tcga">Endometrial Cancer</a>)</li>
<li><strong>Updated drug data</strong> from KEGG DRUG and NCI Cancer Drugs (aggregated by <a href="https://bitbucket.org/armish/pihelper">PiHelper</a>)</li>
</ul>
<h1 id="october-22-2012">October 22, 2012</h1>
<ul>
<li>All data updated to the <strong>Broad Firehose</strong> run from July 25, 2012.</li>
<li><strong>COSMIC data</strong> added to Mutation Details (via Oncotator).</li>
<li>All predicted functional impact scores are updated to <strong>Mutation Assessor 2.0</strong>.</li>
<li>Users can now base queries on genes in recurrent regions of copy-number alteration (from <strong>GISTIC</strong> via Firehose).</li>
<li>The <a href="onco_query_lang_desc.jsp">Onco Query Language (OQL)</a> now supports queries for specific mutations or mutation types.</li>
<li>Data sets added that match the data of all TCGA publications (GBM, ovarian, colorectal, and lung squamous).</li>
</ul>
<h1 id="july-18-2012">July 18, 2012</h1>
<ul>
<li>Mutation data for the TCGA lung squamous cell carcinoma and breast cancer projects (manuscripts in press at Nature).</li>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (May 25, 2012).</li>
<li><strong>Drug information</strong> added to the network view (via Drugbank).</li>
<li><strong>Improved cross-cancer</strong> queries: Option to select data types, export of summary graphs.</li>
<li>Users can now base queries on frequently mutated genes (from <strong>MutSig</strong> via Firehose).</li>
</ul>
<h1 id="may-16-2012">May 16, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (March 21, 2012).</li>
<li>Extended <strong>cross-cancer</strong> functionality, enabling users to query across all cancer studies in our database.</li>
<li>New <strong>"build a case"</strong> functionality, enabling users to generate custom case sets, based on one or more clinical attributes.</li>
<li>New OncoPrint features, including more <strong>compact OncoPrints</strong>, and support for <strong>RPPA</strong> visualization.</li>
</ul>
<h1 id="february-27-2012">February 27, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (January 24, 2012).</li>
<li>Validated mutation data for colorectal cancer.</li>
<li>New feature: <strong>Mutation Diagrams</strong> that show mutations in the context of protein domains.
<center><img alt="TP53 Mutations in Ovarian Cancer" src="images/previews/tp53_mutations.png" /></center></li>
</ul>
<h1 id="january-30-2012">January 30, 2012</h1>
<ul>
<li>Updated data for several TCGA cancer studies.</li>
<li>Some small bug-fixes.</li>
</ul>
<h1 id="december-22-2011">December 22, 2011</h1>
<ul>
<li><strong>Fourteen new <a href="http://cancergenome.nih.gov/">TCGA</a> cancer studies</strong>:  This includes complete data for TCGA Colorectal Carcinoma
and provisional data for thirteen other cancer types in the TCGA production pipeline.  Please note that data from these
thirteen new cancer types are provisional, not final and do not yet include mutation data.
As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.  </li>
</ul>
<p><center><img alt="TCGA" src="http://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/tcga_logo.png" /></center></p>
<ul>
<li><strong>Four new data types</strong>:<ul>
<li>Reverse-phase protein array (RPPA) data.</li>
<li>microRNA expression and copy-number (including support for multiple loci)</li>
<li>RNA-Seq based expression data.</li>
<li>log2 copy-number data.</li>
</ul>
</li>
<li>Updated TCGA GBM copy-number, expression, and methylation data.</li>
<li>New <strong>gene symbol validation</strong> service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</li>
<li><strong>Links to IGV</strong> for visualization of DNA copy-number changes.</li>
<li>Background information from the <a href="http://www.sanger.ac.uk/genetics/CGP/Census/">Sanger Cancer Gene Census</a>.</li>
<li>Two <strong>new <a href="tutorial.jsp">Tutorials</a></strong> to get you quickly started in using the portal.
<BR/>
<BR/></li>
</ul>
<h1 id="november-14-2011">November 14, 2011</h1>
<ul>
<li>New and <strong>improved mutation details</strong>, with sorting and filtering capabilities.</li>
<li>In collaboration with Bilkent University, we have added a <strong>new Network tab</strong> to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
</ul>
<p><center><img alt="GBM Network" src="images/previews/ova_network.png" /></center>
<BR/>
<BR/></p>
<h1 id="september-3-2011">September 3, 2011</h1>
<ul>
<li>You can now query across different cancer studies (feature available directly from the home page).</li>
<li>Our <a href="cgds_r.jsp">MATLAB CGDS Cancer Genomics Toolbox</a> is now available.  The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.
<BR/>
<BR/></li>
</ul>
<h1 id="march-2-2011">March 2, 2011</h1>
<p>New plotting features and other improvements:</p>
<ul>
<li>Correlation plots that show the relationship between different data types for individual genes.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Updated <a href="cgds_r.jsp">R Package</a> with support for correlation plots and general improvements for retrieving and accessing data in R data frames.</li>
<li>The <a href="web_api.jsp">Web Interface</a> now supports basic clinical data, e.g. survival data.</li>
<li><a href="networks.jsp">Networks</a> for pathway analysis are now available for download.</li>
</ul>
<p><center><img alt="Survival Analysis" src="images/previews/ova_survival_60_percent.png" /></center>
<BR/>
<BR/></p>
<h1 id="december-15-2010">December 15, 2010</h1>
<p>Several new features, including:</p>
<ul>
<li>Redesigned and streamlined user interface, based on user feedback and usability testing.</li>
<li>Advanced support for gene-specific alterations.  For example, users
can now view mutations within TP53, and ignore copy number alterations, or
only view amplifications of EGFR, and ignore deletions.</li>
<li>Improved performance.</li>
<li><a href="faq.jsp">Frequently Asked Questions</a> document released.</li>
<li>Updated <a href="video.jsp">Video Tutorial</a>.
<BR/>
<BR/></li>
</ul>
<h1 id="november-4-2010">November 4, 2010</h1>
<ul>
<li>Enhanced <a href="faq.jsp#what-are-oncoprints">Oncoprints</a>, enabling users to quickly visualize genomic alterations across many cases.  Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.</li>
<li>Official release of our <a href="web_api.jsp">Web Interface</a>, enabling programmatic access to all data.</li>
<li>Official release of our <a href="cgds_r.jsp">R Package</a>, enabling programmatic access to all data from the R platform for statistical computing.</li>
</ul>
<p><center><img alt="OncoPrints" src="images/previews/gbm_oncoprint.png" /></center></p>