<h1 id="march-27-2014">March 27, 2014</h1>
<ul>
<li>New features:<ul>
<li>Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)</li>
<li>Gene expression correlation analysis (find all genes with expression correlation to your query genes)</li>
<li>The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation</li>
<li>The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation
<br/><img alt="3D viewer &amp; Co-expression" src="images/previews/news_3d_coexp.png" /></li>
</ul>
</li>
</ul>
<h1 id="october-18-2013">October 18, 2013</h1>
<ul>
<li>Improved interface for survival plots, including information on individual samples via mouse-over</li>
<li>New fusion glyph in OncoPrints</li>
<li>Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
<center><img alt="Cross Cancer Query" src="images/previews/cross_cancer.png" /></center></li>
</ul>
<h1 id="september-5-2013">September 5, 2013:</h1>
<ul>
<li><a href="tools.jsp">New standalone oncoprint tool</a></li>
<li>Updated COSMIC data (v66 Release)</li>
<li>Improved / interactive visualization on the "Protein changes" tab</li>
<li>Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters</li>
<li>Addition of DNA cytoband information in the patient view of copy-number changes</li>
<li>OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
<center><img alt="Oncoprint with clinical track" src="images/previews/oncoprint_clinical_track.png" /></center></li>
</ul>
<h1 id="june-4-2013">June 4, 2013</h1>
<ul>
<li>All TCGA data updated to the April Firehose run (April 21, 2012).</li>
</ul>
<h1 id="april-28-2013">April 28, 2013</h1>
<ul>
<li>All TCGA data updated to the March Firehose run (March 26, 2012).</li>
<li>mRNA percentiles for altered genes shown in patient view.</li>
</ul>
<h1 id="april-2-2013">April 2, 2013</h1>
<ul>
<li>New study added: Breast Cancer (British Columbia, Nature 2012).</li>
<li>All TCGA data updated to the February Firehose run (February 22, 2012).</li>
<li><strong>Sequencing read counts and frequencies</strong> are now shown in the Mutation Details table when available.</li>
<li>Improved OncoPrints, resulting in performance improvements.</li>
<li>Major new feature: Users can now visualize <strong>genomic alterations and clinical data of individual tumors</strong>, including:<ul>
<li>Summary of <strong>mutations</strong> and <strong>copy-number alterations</strong> of interest</li>
<li><strong>Clinical trial</strong> information</li>
<li>TCGA <strong>Pathology Reports</strong></li>
</ul>
</li>
<li>New <strong>cancer summary view</strong> (Example <a href="study.do?cancer_study_id=ucec_tcga">Endometrial Cancer</a>)</li>
<li><strong>Updated drug data</strong> from KEGG DRUG and NCI Cancer Drugs (aggregated by <a href="https://bitbucket.org/armish/pihelper">PiHelper</a>)</li>
</ul>
<h1 id="october-10-2012">October 10, 2012</h1>
<ul>
<li>New studies added: Breast cancer sequencing studies from the Broad and the Sanger, Cancer Cell Line Encyclopedia (CCLE), and cervical cancer TCGA.</li>
<li>All data updated to the <strong>Broad Firehose</strong> run from July 25, 2012.</li>
<li><strong>COSMIC data</strong> added to Mutation Details (via Oncotator)</li>
<li>All predicted functional impact scores are updated to <strong>Mutation Assessor 2.0</strong></li>
<li>Users can now base queries on genes in recurrent regions of copy-number alteration (from <strong>GISTIC</strong> via Firehose).</li>
<li>The <a href="onco_query_lang_desc.jsp">Onco Query Language (OQL)</a> now supports queries for specific mutations or mutation types.</li>
<li><strong>Drug information</strong> added to the network view (via Drugbank).</li>
<li>Extended <strong>cross-cancer</strong> functionality, enabling users to query across all cancer studies in our database.</li>
<li>Users can now base queries on frequently mutated genes (from <strong>MutSig</strong> via Firehose).</li>
<li>New <strong>"build a case"</strong> functionality, enabling users to generate custom case sets, based on one or more clinical attributes.</li>
<li>New OncoPrint features, including more <strong>compact OncoPrints</strong>, and support for <strong>RPPA</strong> visualization.</li>
<li>New feature: <strong>Mutation Diagrams</strong> that show mutations in the context of protein domains.
<center><img alt="TP53 Mutations in Ovarian Cancer" src="images/previews/tp53_mutations.png" /></center></li>
</ul>
<h1 id="january-27-2011">January 27, 2011</h1>
<ul>
<li><strong>Updated TCGA data sets</strong> for Breast Cancer, Ovarian, and Uterine Corpus Endometrioid Carcinoma.</li>
<li>New <strong>gene symbol validation</strong> service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</li>
<li>New and <strong>improved mutation details</strong>, with sorting and filtering capabilities.</li>
<li>In collaboration with Bilkent University, we have added a <strong>new Network tab</strong> to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
</ul>
<p><center><img alt="GBM Network" src="images/previews/ova_network.png" /></center></p>
<h1 id="september-23-2011">September 23, 2011</h1>
<ul>
<li>Updated data for breast and endometrial cancer from TCGA, including 499 fully sequenced breast cancer samples.</li>
<li>You can now query across different cancer studies (feature available directly from the home page).</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<h1 id="april-1-2011">April 1, 2011</h1>
<p>This is the initial release of the cBio SU2C Cancer Genomics Portal. The portal contains genomic data from the TCGA projects on serous ovarian cancer, breast cancer, and uterine cancer, as well as limited protein data from TCGA ovarian cancer samples. </p>