<h1 id="november-9-2015">November 9, 2015</h1>
<ul>
<li><strong>New features</strong>:<ul>
<li>Links to MyCancerGenome.org for mutations</li>
<li>Improved display of selection samples on the study view page</li>
</ul>
</li>
<li><strong>Improvements</strong>:<ul>
<li>"Enrichments" analysis is now run across all genes</li>
<li>The "Network" tab is now using Cytoscape.js (Adobe Flash is no longer required)</li>
</ul>
</li>
</ul>
<h1 id="august-21-2015">August 21, 2015</h1>
<ul>
<li>All <strong>TCGA data</strong> updated to the Firehose run of April 16, 2015.</li>
<li><strong>New feature</strong>: Enrichments Analysis finds alterations that are enriched in either altered or unaltered samples.</li>
<li><strong>Improvement</strong>: improved OncoPrint with better performance.</li>
</ul>
<h1 id="june-3-2015">June 3, 2015</h1>
<ul>
<li><strong>Improvements</strong>:<ul>
<li>Allowed downloading data in each chart/table in study summary page.</li>
<li>Added log-rank test <em>p</em>-values to the survival plots in study summary page.</li>
<li>Improved visualization of patient clinical data in patient-centric view.</li>
<li>Added option to merge multiple samples for the same patient in OncoPrint.</li>
</ul>
</li>
</ul>
<h1 id="april-28-2015">April 28, 2015</h1>
<ul>
<li><strong>New features</strong>:<ul>
<li>Redesigned query interface to allow selecting multiple cancer studies</li>
<li>Redesigned <strong>Plots</strong> tab</li>
</ul>
</li>
</ul>
<h1 id="january-14-2015">January 14, 2015</h1>
<ul>
<li>All TCGA data updated to the Firehose run of October 17, 2014</li>
</ul>
<h1 id="november-26-2014">November 26, 2014</h1>
<ul>
<li>Complete and updated TCGA clinical data</li>
<li>Redesigned mutual exclusivity tab</li>
</ul>
<h1 id="september-26-2014">September 26, 2014</h1>
<ul>
<li>All TCGA data updated to the Firehose run of July 15, 2014</li>
</ul>
<h1 id="may-21-2014">May 21, 2014</h1>
<ul>
<li>All TCGA data updated to the Firehose run of April 16, 2014</li>
</ul>
<h1 id="may-12-2014">May 12, 2014</h1>
<ul>
<li>Improved study summary page including survival analysis based on clinical attributes
<br/>e.g. <a href="study?id=ucec_tcga_pub">TCGA Endometrial Cancer cohort</a>
<br/><img alt="Study view" src="images/previews/study_view.png" /></li>
</ul>
<h1 id="march-27-2014">March 27, 2014</h1>
<ul>
<li>New features:<ul>
<li>Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)</li>
<li>Gene expression correlation analysis (find all genes with expression correlation to your query genes)</li>
<li>The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation</li>
<li>The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation
<br/><img alt="3D viewer &amp; Co-expression" src="images/previews/news_3d_coexp.png" /></li>
</ul>
</li>
</ul>
<h1 id="march-18-2014">March 18, 2014</h1>
<ul>
<li>All TCGA data updated to the Firehose run of January 15, 2014</li>
<li>Updated to the latest COSMIC data (v68)</li>
<li>Added two new provisional TCGA studies<ul>
<li>Adrenocortical Carcinoma</li>
<li>Uterine Carcinosarcoma</li>
</ul>
</li>
<li>Added genomic data from multiple studies from the literature</li>
</ul>
<h1 id="november-8-2013">November 8, 2013</h1>
<ul>
<li>All TCGA data updated to the Firehose run of September 23, 2013.</li>
<li>Updated to the latest COSMIC data (v67).</li>
</ul>
<h1 id="october-18-2013">October 18, 2013</h1>
<ul>
<li>Improved interface for survival plots, including information on individual samples via mouse-over</li>
<li>New fusion glyph in OncoPrints &nbsp;&nbsp;&nbsp;<a href="index.do?cancer_study_id=hnsc_tcga_pub&amp;genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&amp;genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&amp;Z_SCORE_THRESHOLD=2.0&amp;RPPA_SCORE_THRESHOLD=2.0&amp;data_priority=0&amp;case_set_id=hnsc_tcga_pub_sequenced&amp;case_ids=&amp;gene_set_choice=user-defined-list&amp;gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&amp;clinical_param_selection=null&amp;tab_index=tab_visualize&amp;Action=Submit"><img alt="FGFR3 fusions in head and neck carcinoma" src="images/previews/fusion-in-oncoprint.png" /></a></li>
<li>Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
<center><img alt="Cross Cancer Query" src="images/previews/cross_cancer.png" /></center></li>
</ul>
<h1 id="september-5-2013">September 5, 2013:</h1>
<ul>
<li>Updated COSMIC data (v66 Release)</li>
<li>Improved / interactive visualization on the "Protein changes" tab</li>
<li>Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters</li>
<li>Addition of DNA cytoband information in the patient view of copy-number changes</li>
<li>OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
<center><img alt="Oncoprint with clinical track" src="images/previews/oncoprint_clinical_track.png" /></center></li>
</ul>
<h1 id="july-23-2013">July 23, 2013</h1>
<ul>
<li>Multi-gene correlation plots.</li>
<li>Variant allele frequency distribution plots for individual tumor samples.</li>
<li>Tissue images for TCGA samples in the patient view, via <a href="http://cancer.digitalslidearchive.net/">Digital Slide Archive</a>. <a href="case.do#/patient?studyId=ucec_tcga&amp;caseId=TCGA-BK-A0CC&tab=tissueImageTab">Example</a>.</li>
</ul>
<h1 id="july-16-2013">July 16, 2013</h1>
<ul>
<li>All TCGA data updated to the May Firehose run (May 23, 2012).<ul>
<li>TCGA Pancreatic Cancer study (provisional) added.</li>
</ul>
</li>
</ul>
<h1 id="july-4-2013">July 4, 2013</h1>
<ul>
<li>Improved rendering of mutation diagrams, including ability to download in PDF format.</li>
<li>Improved home page: Searchable cancer study &amp; gene set selectors, data sets selector.</li>
</ul>
<h1 id="june-13-2013">June 13, 2013</h1>
<ul>
<li>Improved interface for correlation plots, including information on individual samples via mouse-over.</li>
<li>Gene Details from Biogene are now available in the Network view.</li>
</ul>
<h1 id="june-4-2013">June 4, 2013</h1>
<ul>
<li>All TCGA data updated to the April Firehose run (April 21, 2012).</li>
</ul>
<h1 id="april-28-2013">April 28, 2013</h1>
<ul>
<li>All TCGA data updated to the March Firehose run (March 26, 2012).</li>
</ul>
<h1 id="april-2-2013">April 2, 2013</h1>
<ul>
<li>All TCGA data updated to the February Firehose run (February 22, 2012).</li>
<li>mRNA percentiles for altered genes shown in patient view.</li>
</ul>
<h1 id="march-26-2013">March 26, 2013</h1>
<ul>
<li>All TCGA data updated to the <strong>January Firehose run</strong> (January 16, 2012).</li>
</ul>
<h1 id="january-21-2013">January 21, 2013</h1>
<ul>
<li>All data updated to the <strong>October Firehose run</strong> (October 24, 2012), plus updated mutation data for many tumor types.</li>
<li><strong>Sequencing read counts and frequencies</strong> are now shown in the Mutation Details table when available.</li>
<li><strong>Support for gene fusions</strong> in lung adenocarcinoma (RET/ROS1/ALK fusions are shown as mutations, other tumor types coming soon).</li>
<li>Improved visualization of drugs and clinical trials for individual patients</li>
<li>Improved OncoPrints, resulting in performance improvements</li>
</ul>
<h1 id="november-21-2012">November 21, 2012</h1>
<ul>
<li>Major new feature: Users can now visualize <strong>genomic alterations and clinical data of individual tumors</strong>, including:<ul>
<li>Summary of <strong>mutations</strong> and <strong>copy-number alterations</strong> of interest</li>
<li><strong>Clinical trial</strong> information</li>
<li>TCGA <strong>Pathology Reports</strong></li>
</ul>
</li>
<li>New <strong>cancer summary view</strong> (Example <a href="study?id=ucec_tcga">Endometrial Cancer</a>)</li>
<li><strong>Updated drug data</strong> from KEGG DRUG and NCI Cancer Drugs (aggregated by <a href="https://bitbucket.org/armish/pihelper">PiHelper</a>)</li>
</ul>
<h1 id="october-10-2012">October 10, 2012</h1>
<ul>
<li>All data updated to the <strong>Broad Firehose</strong> run from July 25, 2012.</li>
<li><strong>COSMIC data</strong> added to Mutation Details (via Oncotator)</li>
<li>All predicted functional impact scores are updated to <strong>Mutation Assessor 2.0</strong></li>
<li>Users can now base queries on genes in recurrent regions of copy-number alteration (from <strong>GISTIC</strong> via Firehose).</li>
<li>New studies added: Cancer Cell Line Encyclopedia (CCLE) and Lung Adenocarcinoma from the Broad Institute.</li>
<li>The <a href="onco_query_lang_desc.jsp">Onco Query Language (OQL)</a> now supports queries for specific mutations or mutation types.</li>
</ul>
<h1 id="july-18-2012">July 18, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (May 25, 2012).</li>
<li><strong>Drug information</strong> added to the network view (via Drugbank).</li>
<li><strong>Improved cross-cancer</strong> queries: Option to select data types, export of summary graphs.</li>
<li>Users can now base queries on frequently mutated genes (from <strong>MutSig</strong> via Firehose).</li>
</ul>
<h1 id="may-16-2012">May 16, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Broad Firehose run</strong> (March 21, 2012).</li>
<li>Extended <strong>cross-cancer</strong> functionality, enabling users to query across all cancer studies in our database.</li>
<li>New <strong>"build a case"</strong> functionality, enabling users to generate custom case sets, based on one or more clinical attributes.</li>
<li>New OncoPrint features, including more <strong>compact OncoPrints</strong>, and support for <strong>RPPA</strong> visualization.</li>
</ul>
<h1 id="february-27-2012">February 27, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Firehose run</strong> (January 24).</li>
<li>Validated mutation data for colorectal cancer.</li>
<li>Updated mutation data for several TCGA studies.</li>
<li>New feature: <strong>Mutation Diagrams</strong> that show mutations in the context of protein domains.
<center><img alt="TP53 Mutations in Ovarian Cancer" src="images/previews/tp53_mutations.png" /></center></li>
</ul>
<h1 id="january-26-2012">January 26, 2012</h1>
<ul>
<li>All data updated to the <strong>latest Firehose run</strong> (December 30).</li>
<li><strong>New RNA-Seq expression data</strong> for many tumor types.</li>
<li><strong>New mutation data</strong>: All new bladder cancer data and small updates for lung squamous carcinoma.</li>
</ul>
<h1 id="january-5-2012">January 5, 2012</h1>
<ul>
<li>Now <strong>19 tumor types</strong>, all data updated to the <strong>latest Firehose run</strong> (November 28).</li>
<li>Support for <strong>microRNA</strong> expression and copy-number (including multiple loci).</li>
<li>New <strong>gene symbol validation</strong> service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.</li>
<li>Improved <strong>links to IGV</strong> for visualization of DNA copy-number changes.</li>
<li>Background information from the <a href="http://www.sanger.ac.uk/genetics/CGP/Census/">Sanger Cancer Gene Census</a>.</li>
<li>Two <strong>new <a href="tutorial.jsp">Tutorials</a></strong> to get you quickly started in using the portal.</li>
</ul>
<h1 id="december-5-2011">December 5, 2011</h1>
<p>New data types and updated data:</p>
<ul>
<li>Added support for RNA-Seq expression data (COADREAD, KIRC, LUSC).</li>
<li>Added support for log2 copy-number data (available on the Plots tab).</li>
<li>Updated LAML and LUSC MAF files.</li>
<li>Updated LUSC case lists (reflecting the manuscript sample set).</li>
</ul>
<h1 id="november-15-2011">November 15, 2011</h1>
<p>Another major update, with the following new features and data:</p>
<ul>
<li>In collaboration with Bilkent University, we have added a new <em>network</em> tab to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from <a href="http://www.pathwaycommons.org">Pathway Commons</a>.</li>
<li>New and improved mutation details, with sorting and filtering capabilities.</li>
<li>Survival analysis - assess survival differences between altered and non-altered patient sets.</li>
<li>Links to the <a href="http://www.broadinstitute.org/igv/">Integrative Genomics Viewer (IGV)</a> to view copy-number alterations.</li>
<li>All data updated to the October 2011 TCGA Broad Firehose run.</li>
<li>RPPA data is now available for 2154 samples in six tumor types.</li>
<li>Many new and updated MAF files - now 9 tumor types with mutation data.</li>
</ul>
<p><img alt="GBM Network" src="images/previews/ova_network.png" /></p>
<h1 id="october-18-2011">October 18, 2011</h1>
<p>This release adds the following new data:</p>
<ul>
<li>Based on the September Firehose run with new tumor types (Bladder, Liver, Low grade glioma).</li>
<li>Latest MAF files from the COADREAD and LUSC analysis working groups.</li>
<li>Mutation Assessor links for all missense mutations.</li>
<li>RPPA data for kidney clear cell and endometrial cancer.</li>
<li>GBM case lists for transcriptional subtypes and CIMP.</li>
</ul>
<h1 id="september-12-2011">September 12, 2011</h1>
<p>This is a major update, which includes the following main changes:</p>
<ul>
<li>New tumor types and data: Now ~3500 samples from 12 tumor types, including mutation data for 8 tumor types.</li>
<li>RPPA data support - the Portal currently has data for ovarian and breast cancer.</li>
<li>Access control via <a href="http://openid.net/">OpenID</a>.</li>
<li>The code for the cBio Portal has now been fully open sourced, and made available at <a href="http://code.google.com/p/cbio-cancer-genomics-portal/">Google Code</a>.  If you would like to join our open source efforts and make the portal even better, drop us an email.</li>
</ul>
<h1 id="april-26-2011">April 26, 2011</h1>
<p>This is the initial release of the cBio TCGA Cancer Genomics Portal. The portal contains data from all completed and ongoing TCGA projects. All data is automatically imported from the output of the Broad Firehose.</p>
<p>This release includes data on 2531 samples from 11 tumor types.</p>
