<!DOCTYPE html>
<html lang="en" class="h-full">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <meta name="turbo-cache-control" content="no-cache" data-turbo-track="reload" data-track-token="3.5.0.766544874491">

    <!-- See retype.com -->
    <meta name="generator" content="Retype 3.5.0">

    <!-- Primary Meta Tags -->
    <title>More application.properties Settings</title>
    <meta name="title" content="More application.properties Settings">
    <meta name="description" content="This page describes the main properties within application.properties.">

    <!-- Canonical -->
    <link rel="canonical" href="https://docs.cbioportal.org/deployment/customization/application.properties-reference/">

    <!-- Open Graph / Facebook -->
    <meta property="og:type" content="website">
    <meta property="og:url" content="https://docs.cbioportal.org/deployment/customization/application.properties-reference/">
    <meta property="og:title" content="More application.properties Settings">
    <meta property="og:description" content="This page describes the main properties within application.properties.">
    <meta property="og:image" content="https://docs.cbioportal.org/images/previews/quick_search_example.png">

    <!-- Twitter -->
    <meta property="twitter:card" content="summary_large_image">
    <meta property="twitter:url" content="https://docs.cbioportal.org/deployment/customization/application.properties-reference/">
    <meta property="twitter:title" content="More application.properties Settings">
    <meta property="twitter:description" content="This page describes the main properties within application.properties.">
    <meta property="twitter:image" content="https://docs.cbioportal.org/images/previews/quick_search_example.png">

    <script data-cfasync="false">(function () { var el = document.documentElement, m = localStorage.getItem("doc_theme"), wm = window.matchMedia; if (m === "dark" || (!m && wm && wm("(prefers-color-scheme: dark)").matches)) { el.classList.add("dark") } else { el.classList.remove("dark") } })();</script>

    <link href="../../../images/cbio-logo.png" rel="icon">
    <link href="../../../resources/css/retype.css?v=3.5.0.766544874491" rel="stylesheet">

    <script data-cfasync="false" src="../../../resources/js/config.js?v=3.5.0.766544874491" data-turbo-eval="false" defer></script>
    <script data-cfasync="false" src="../../../resources/js/retype.js?v=3.5.0" data-turbo-eval="false" defer></script>
    <script id="lunr-js" data-cfasync="false" src="../../../resources/js/lunr.js?v=3.5.0.766544874491" data-turbo-eval="false" defer></script>
    <script id="prism-js" data-cfasync="false" src="../../../resources/js/prism.js?v=3.5.0.766544874491" defer></script>
    <link href="../../../resources/css/katex.css?v=3.5.0" rel="stylesheet">
    <script id="katex-js" data-cfasync="false" src="../../../resources/js/katex.js?v=3.5.0" defer></script>

    <link rel="stylesheet" href="/static/custom.css"></link>
</head>
<body>
    <div id="docs-app" class="relative text-base antialiased text-gray-700 bg-white font-body dark:bg-dark-850 dark:text-dark-300">
        <div class="absolute bottom-0 left-0 bg-gray-100 dark:bg-dark-800" style="top: 5rem; right: 50%"></div>
    
        <header id="docs-site-header" class="sticky top-0 z-30 flex w-full h-16 bg-white border-b border-gray-200 md:h-20 dark:bg-dark-850 dark:border-dark-650">
            <div class="container relative flex items-center justify-between pr-6 grow md:justify-start">
                <!-- Mobile menu button skeleton -->
                <button v-cloak class="skeleton docs-mobile-menu-button flex items-center justify-center shrink-0 overflow-hidden dark:text-white focus:outline-none rounded-full w-10 h-10 ml-3.5 md:hidden"><svg xmlns="http://www.w3.org/2000/svg" class="mb-px shrink-0" width="24" height="24" viewBox="0 0 24 24" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor"><path d="M2 4h20v2H2zM2 11h20v2H2zM2 18h20v2H2z"></path></g></svg></button>
                <div v-cloak id="docs-sidebar-toggle"></div>
        
                <!-- Logo -->
                <div class="flex items-center justify-between h-full py-2 md:w-75">
                    <div class="flex items-center px-2 md:px-6">
                        <a id="docs-site-logo" href="../../../" class="flex items-center leading-snug text-xl">
                            <span class="w-10 mr-2 grow-0 shrink-0 overflow-hidden">
                                <img class="max-h-10 dark:hidden md:inline-block" src="../../../images/cbio-logo.png">
                                <img class="max-h-10 hidden dark:inline-block" src="../../../images/cbio-logo.png">
                            </span>
                            <span class="dark:text-white font-semibold line-clamp-1 md:line-clamp-2">cBioPortal Docs</span>
                        </a>
                    </div>
        
                    <span class="hidden h-8 border-r md:inline-block dark:border-dark-650"></span>
                </div>
        
                <div class="flex justify-between md:grow">
                    <!-- Top Nav -->
                    <nav class="hidden md:flex">
                        <ul class="flex flex-col mb-4 md:pl-16 md:mb-0 md:flex-row md:items-center">
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.cbioportal.org">
                                    <img class="mb-px mr-1" width="18" alt="" src="../../../images/cbio-logo.png">
                                    <span>cBioPortal</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <symbol id="mark-github-icon-symbol" viewBox="0 0 16 16"><path d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"/></symbol><use xlink:href="#mark-github-icon-symbol" />
                                        </g>
                                    </svg>
                                    <span>Github</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/issues/">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                        </g>
                                    </svg>
                                    <span>Issues</span>
                                </a>
                            </li>
                            <li class="mr-6">
                                <a class="py-2 md:mb-0 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://groups.google.com/g/cbioportal">
                                    <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                        <g fill="currentColor">
                                            <path d="M1.75 1h12.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 14.25 14H8.061l-2.574 2.573A1.458 1.458 0 0 1 3 15.543V14H1.75A1.75 1.75 0 0 1 0 12.25v-9.5C0 1.784.784 1 1.75 1ZM1.5 2.75v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25H1.75a.25.25 0 0 0-.25.25Z"/><path d="M22.5 8.75a.25.25 0 0 0-.25-.25h-3.5a.75.75 0 0 1 0-1.5h3.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 22.25 20H21v1.543a1.457 1.457 0 0 1-2.487 1.03L15.939 20H10.75A1.75 1.75 0 0 1 9 18.25v-1.465a.75.75 0 0 1 1.5 0v1.465c0 .138.112.25.25.25h5.5a.75.75 0 0 1 .53.22l2.72 2.72v-2.19a.75.75 0 0 1 .75-.75h2a.25.25 0 0 0 .25-.25v-9.5Z"/>
                                        </g>
                                    </svg>
                                    <span>Google Group</span>
                                </a>
                            </li>
        
                        </ul>
                    </nav>
        
                    <!-- Header Right Skeleton -->
                    <div v-cloak class="flex justify-end grow skeleton">
        
                        <!-- Search input mock -->
                        <div class="relative hidden w-40 lg:block lg:max-w-sm lg:ml-auto">
                            <div class="absolute flex items-center justify-center h-full pl-3 dark:text-dark-300">
                                <svg xmlns="http://www.w3.org/2000/svg" class="icon-base" width="16" height="16" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 1px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                            </div>
                            <input class="w-full h-10 placeholder-gray-400 transition-colors duration-200 ease-in bg-gray-200 border border-transparent rounded md:text-sm hover:bg-white hover:border-gray-300 focus:outline-none focus:bg-white focus:border-gray-500 dark:bg-dark-600 dark:border-dark-600 dark:placeholder-dark-400" style="padding: 0.625rem 0.75rem 0.625rem 2rem" type="text" placeholder="Search">
                        </div>
        
                        <!-- Mobile search button -->
                        <div class="flex items-center justify-center w-10 h-10 lg:hidden">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="20" height="20" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><path d="M21.71 20.29l-3.68-3.68A8.963 8.963 0 0020 11c0-4.96-4.04-9-9-9s-9 4.04-9 9 4.04 9 9 9c2.12 0 4.07-.74 5.61-1.97l3.68 3.68c.2.19.45.29.71.29s.51-.1.71-.29c.39-.39.39-1.03 0-1.42zM4 11c0-3.86 3.14-7 7-7s7 3.14 7 7c0 1.92-.78 3.66-2.04 4.93-.01.01-.02.01-.02.01-.01.01-.01.01-.01.02A6.98 6.98 0 0111 18c-3.86 0-7-3.14-7-7z" ></path></g></svg>
                        </div>
        
                        <!-- Dark mode switch placeholder -->
                        <div class="w-10 h-10 lg:ml-2"></div>
        
                        <!-- History button -->
                        <div class="flex items-center justify-center w-10 h-10" style="margin-right: -0.625rem;">
                            <svg xmlns="http://www.w3.org/2000/svg" class="shrink-0 icon-base" width="22" height="22" viewBox="0 0 24 24" aria-labelledby="icon" role="presentation" style="margin-bottom: 0px;"><g fill="currentColor" ><g ><path d="M12.01 6.01c-.55 0-1 .45-1 1V12a1 1 0 00.4.8l3 2.22a.985.985 0 001.39-.2.996.996 0 00-.21-1.4l-2.6-1.92V7.01c.02-.55-.43-1-.98-1z"></path><path d="M12.01 1.91c-5.33 0-9.69 4.16-10.05 9.4l-.29-.26a.997.997 0 10-1.34 1.48l1.97 1.79c.19.17.43.26.67.26s.48-.09.67-.26l1.97-1.79a.997.997 0 10-1.34-1.48l-.31.28c.34-4.14 3.82-7.41 8.05-7.41 4.46 0 8.08 3.63 8.08 8.09s-3.63 8.08-8.08 8.08c-2.18 0-4.22-.85-5.75-2.4a.996.996 0 10-1.42 1.4 10.02 10.02 0 007.17 2.99c5.56 0 10.08-4.52 10.08-10.08.01-5.56-4.52-10.09-10.08-10.09z"></path></g></g></svg>
                        </div>
                    </div>
        
                    <div v-cloak class="flex justify-end grow">
                        <div id="docs-mobile-search-button"></div>
                        <doc-search-desktop></doc-search-desktop>
        
                        <doc-theme-switch class="lg:ml-2"></doc-theme-switch>
                        <doc-history></doc-history>
                    </div>
                </div>
            </div>
        </header>
    
        <div class="container relative flex bg-white">
            <!-- Sidebar Skeleton -->
            <div v-cloak class="fixed flex flex-col shrink-0 duration-300 ease-in-out bg-gray-100 border-gray-200 sidebar top-20 w-75 border-r h-screen md:sticky transition-transform skeleton dark:bg-dark-800 dark:border-dark-650">
            
                <!-- Render this div, if config.showSidebarFilter is `true` -->
                <div class="flex items-center h-16 px-6">
                    <input class="w-full h-8 px-3 py-2 transition-colors duration-200 ease-linear bg-white border border-gray-200 rounded shadow-none text-sm focus:outline-none focus:border-gray-600 dark:bg-dark-600 dark:border-dark-600" type="text" placeholder="Filter">
                </div>
            
                <div class="pl-6 mt-1 mb-4">
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-32 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-48 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                    <div class="w-40 h-3 mb-4 bg-gray-200 rounded-full loading dark:bg-dark-600"></div>
                </div>
            
                <div class="shrink-0 mt-auto bg-transparent dark:border-dark-650">
                    <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                        <span class="text-xs whitespace-nowrap">Powered by</span>
                        <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                    </a>
                </div>
            </div>
            
            <!-- Sidebar component -->
            <doc-sidebar v-cloak>
                <template #sidebar-footer>
                    <div class="shrink-0 mt-auto border-t md:bg-transparent md:border-none dark:border-dark-650">
            
                        <div class="py-3 px-6 md:hidden border-b dark:border-dark-650">
                            <nav>
                                <ul class="flex flex-wrap justify-center items-center">
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://www.cbioportal.org">
                                            <img class="mb-px mr-1" width="18" alt="" src="../../../images/cbio-logo.png">
                                            <span>cBioPortal</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <symbol id="mark-github-icon-symbol" viewBox="0 0 16 16"><path d="M8 0c4.42 0 8 3.58 8 8a8.013 8.013 0 0 1-5.45 7.59c-.4.08-.55-.17-.55-.38 0-.27.01-1.13.01-2.2 0-.75-.25-1.23-.54-1.48 1.78-.2 3.65-.88 3.65-3.95 0-.88-.31-1.59-.82-2.15.08-.2.36-1.02-.08-2.12 0 0-.67-.22-2.2.82-.64-.18-1.32-.27-2-.27-.68 0-1.36.09-2 .27-1.53-1.03-2.2-.82-2.2-.82-.44 1.1-.16 1.92-.08 2.12-.51.56-.82 1.28-.82 2.15 0 3.06 1.86 3.75 3.64 3.95-.23.2-.44.55-.51 1.07-.46.21-1.61.55-2.33-.66-.15-.24-.6-.83-1.23-.82-.67.01-.27.38.01.53.34.19.73.9.82 1.13.16.45.68 1.31 2.69.94 0 .67.01 1.3.01 1.49 0 .21-.15.45-.55.38A7.995 7.995 0 0 1 0 8c0-4.42 3.58-8 8-8Z"/></symbol><use xlink:href="#mark-github-icon-symbol" />
                                                </g>
                                            </svg>
                                            <span>Github</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://github.com/cbioportal/cbioportal/issues/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M7.72.22a.75.75 0 0 1 1.06 0l1.204 1.203A4.98 4.98 0 0 1 12 1c.717 0 1.4.151 2.016.423L15.22.22a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-.971.972A4.991 4.991 0 0 1 17 6v1.104a2.755 2.755 0 0 1 1.917 1.974l1.998-.999a.75.75 0 0 1 .67 1.342L19 10.714V13.5l3.25.003a.75.75 0 0 1 0 1.5L19 15.001V16c0 .568-.068 1.134-.204 1.686l.04.018 2.75 1.375a.75.75 0 1 1-.671 1.342l-2.638-1.319A6.998 6.998 0 0 1 12 23a6.998 6.998 0 0 1-6.197-3.742l-2.758 1.181a.752.752 0 0 1-1.064-.776.752.752 0 0 1 .474-.602l2.795-1.199A6.976 6.976 0 0 1 5 16v-.996H1.75a.75.75 0 0 1 0-1.5H5v-2.79L2.415 9.42a.75.75 0 0 1 .67-1.342l1.998.999A2.756 2.756 0 0 1 7 7.104V6a4.99 4.99 0 0 1 1.69-3.748l-.97-.972a.75.75 0 0 1 0-1.06ZM6.5 9.75V16a5.5 5.5 0 1 0 11 0V9.75c0-.69-.56-1.25-1.25-1.25h-8.5c-.69 0-1.25.56-1.25 1.25ZM8.5 7h7V6a3.5 3.5 0 1 0-7 0Z"/>
                                                </g>
                                            </svg>
                                            <span>Issues</span>
                                        </a>
                                    </li>
                                    <li class="mr-6">
                                        <a class="block py-1 inline-flex items-center text-sm whitespace-nowrap transition-colors duration-200 ease-linear md:text-blue-500 dark:text-blue-400 hover:text-blue-800 dark:hover:text-blue-200" href="https://groups.google.com/g/cbioportal">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mb-px mr-1" width="18" height="18" viewBox="0 0 24 24" role="presentation">
                                                <g fill="currentColor">
                                                    <path d="M1.75 1h12.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 14.25 14H8.061l-2.574 2.573A1.458 1.458 0 0 1 3 15.543V14H1.75A1.75 1.75 0 0 1 0 12.25v-9.5C0 1.784.784 1 1.75 1ZM1.5 2.75v9.5c0 .138.112.25.25.25h2a.75.75 0 0 1 .75.75v2.19l2.72-2.72a.749.749 0 0 1 .53-.22h6.5a.25.25 0 0 0 .25-.25v-9.5a.25.25 0 0 0-.25-.25H1.75a.25.25 0 0 0-.25.25Z"/><path d="M22.5 8.75a.25.25 0 0 0-.25-.25h-3.5a.75.75 0 0 1 0-1.5h3.5c.966 0 1.75.784 1.75 1.75v9.5A1.75 1.75 0 0 1 22.25 20H21v1.543a1.457 1.457 0 0 1-2.487 1.03L15.939 20H10.75A1.75 1.75 0 0 1 9 18.25v-1.465a.75.75 0 0 1 1.5 0v1.465c0 .138.112.25.25.25h5.5a.75.75 0 0 1 .53.22l2.72 2.72v-2.19a.75.75 0 0 1 .75-.75h2a.25.25 0 0 0 .25-.25v-9.5Z"/>
                                                </g>
                                            </svg>
                                            <span>Google Group</span>
                                        </a>
                                    </li>
            
                                </ul>
                            </nav>
                        </div>
            
                        <a class="flex items-center justify-center flex-nowrap h-16 text-gray-400 dark:text-dark-400 hover:text-gray-700 dark:hover:text-dark-300 transition-colors duration-150 ease-in docs-powered-by" target="_blank" href="https://retype.com/" rel="noopener">
                            <span class="text-xs whitespace-nowrap">Powered by</span>
                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-2" fill="currentColor" width="96" height="20" overflow="visible"><path d="M0 0v20h13.59V0H0zm11.15 17.54H2.44V2.46h8.71v15.08zM15.8 20h2.44V4.67L15.8 2.22zM20.45 6.89V20h2.44V9.34z"/><g><path d="M40.16 8.44c0 1.49-.59 2.45-1.75 2.88l2.34 3.32h-2.53l-2.04-2.96h-1.43v2.96h-2.06V5.36h3.5c1.43 0 2.46.24 3.07.73s.9 1.27.9 2.35zm-2.48 1.1c.26-.23.38-.59.38-1.09 0-.5-.13-.84-.4-1.03s-.73-.28-1.39-.28h-1.54v2.75h1.5c.72 0 1.2-.12 1.45-.35zM51.56 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92h4.74v1.83h-6.79V5.36h6.64zM60.09 7.15v7.48h-2.06V7.15h-2.61V5.36h7.28v1.79h-2.61zM70.81 14.64h-2.06v-3.66l-3.19-5.61h2.23l1.99 3.45 1.99-3.45H74l-3.19 5.61v3.66zM83.99 6.19c.65.55.97 1.4.97 2.55s-.33 1.98-1 2.51-1.68.8-3.04.8h-1.23v2.59h-2.06V5.36h3.26c1.42 0 2.45.28 3.1.83zm-1.51 3.65c.25-.28.37-.69.37-1.22s-.16-.92-.48-1.14c-.32-.23-.82-.34-1.5-.34H79.7v3.12h1.38c.68 0 1.15-.14 1.4-.42zM95.85 5.36V7.2h-4.59v1.91h4.13v1.76h-4.13v1.92H96v1.83h-6.79V5.36h6.64z"/></g></svg>
                        </a>
                    </div>
                </template>
            </doc-sidebar>
    
            <div class="grow min-w-0 dark:bg-dark-850">
                <!-- Render "toolbar" template here on api pages --><!-- Render page content -->
                <div class="flex">
                    <div class="min-w-0 p-4 grow md:px-16">
                        <main class="relative pb-12 lg:pt-2">
                            <div class="docs-markdown" id="docs-content">
                                <!-- Rendered if sidebar right is enabled -->
                                <div id="docs-sidebar-right-toggle"></div>
                                <!-- Page content  -->
<doc-anchor-target id="more-applicationproperties-settings" class="break-words">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#more-applicationproperties-settings">#</doc-anchor-trigger>
        <span>More application.properties Settings</span>
    </h1>
</doc-anchor-target>
<p>This page describes the main properties within application.properties.</p>
<doc-anchor-target id="database-settings">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#database-settings">#</doc-anchor-trigger>
        <span>Database Settings</span>
    </h2>
</doc-anchor-target>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">db.user=
db.password=
db.connection_string=
db.driver=[this is the name of your JDBC driver, e.g., com.mysql.jdbc.Driver]</code></pre>
</doc-codeblock></div>
<p>The format of the <code v-pre>db.connection_string</code> is:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">jdbc:mysql://&lt;host&gt;:&lt;port&gt;/&lt;database name&gt;?&lt;parameter1&gt;&amp;&lt;parameter2&gt;&amp;&lt;parameter...&gt;</code></pre>
</doc-codeblock></div>
<p>For example:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">jdbc:mysql://localhost:3306/cbiodb?zeroDateTimeBehavior=convertToNull&amp;useSSL=false</code></pre>
</doc-codeblock></div>
<p><span class="docs-emoji">&#x26A0;&#xFE0F;</span> The fields <code v-pre>db.host</code> and <code v-pre>db.portal_db_name</code> and <code v-pre>db.use_ssl</code> are deprecated. It is required to configure the database connection using
the <code v-pre>db.connection_string</code> instead.</p>
<p><code v-pre>db.tomcat_resource_name</code> is required in order to work with the tomcat database connection pool and should have the default value jdbc/cbioportal in order to work correctly with the your WAR file.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">db.tomcat_resource_name=jdbc/cbioportal</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="cbioportal-customization">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#cbioportal-customization">#</doc-anchor-trigger>
        <span>cBioPortal Customization</span>
    </h2>
</doc-anchor-target>
<doc-anchor-target id="hide-tabs-pages">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#hide-tabs-pages">#</doc-anchor-trigger>
        <span>Hide tabs (pages)</span>
    </h3>
</doc-anchor-target>
<p>Settings controlling which tabs (pages) to hide. Set them to <code v-pre>false</code> if you want to hide those tabs, otherwise set the properties to <code v-pre>true</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.show_data_tab=
skin.show_web_api_tab=
skin.show_r_matlab_tab=
skin.show_tutorials_tab=
skin.show_faqs_tab=
skin.show_news_tab=
skin.show_tools_tab=
skin.show_about_tab=</code></pre>
</doc-codeblock></div>
<p><strong>Note:</strong> <code v-pre>skin.show_tools_tab</code> refers to the <code v-pre>Visualize Your Data</code> tab, while <code v-pre>skin.show_data_tab</code> refers to the <code v-pre>Data Sets</code> tab.</p>
<doc-anchor-target id="cross-cancer-study-query-default">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#cross-cancer-study-query-default">#</doc-anchor-trigger>
        <span>Cross Cancer Study Query Default</span>
    </h3>
</doc-anchor-target>
<p>The cross cancer study query default is a list of studies used when querying one or more genes and not specifying a specific study or list of studies. There are two ways in which the default cross cancer study list is used:</p>
<ol>
<li>When using the linkout links without a study e.g. <code v-pre>/ln?q=TP53:MUT</code>. Those links are used mostly used to allow for easy linking to particular queries. One can&#x27;t get those links using the cBioPortal user interface itself, they are only mentioned in the documentation of the Web API (<a href="https://www.cbioportal.org/webAPI">https://www.cbioportal.org/webAPI</a>).</li>
<li>In the quick search when querying for a gene. Quick search is disabled by default. It is a beta feature. See the <doc-anchor-trigger to="#quick-search-beta">quick search documentation</doc-anchor-trigger>.</li>
</ol>
<p>The configuration is set with the following if you have session service enabled:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">default_cross_cancer_study_session_id=</code></pre>
</doc-codeblock></div>
<p>The title will be pulled from the virtual study. Make sure to create a <code v-pre>virtual_study</code> with studies that everybody has access to and don&#x27;t use a <code v-pre>main_session</code> id.</p>
<p>If session service is disabled one can use the following instead:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none"># query this comma separated list of studies
default_cross_cancer_study_list=
default_cross_cancer_study_list_name=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="quick-search-beta">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#quick-search-beta">#</doc-anchor-trigger>
        <span>Quick Search (BETA)</span>
    </h3>
</doc-anchor-target>
<p><figure class="content-center">
    <img src="../../../images/previews/quick_search_example.png" alt="Quick search example">
    <figcaption class="caption">Quick search example</figcaption>
</figure>
</p>
<p>Enable or disable the quick search with the following:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none"># Enable/Disable quick search (default is false)
quick_search.enabled=true</code></pre>
</doc-codeblock></div>
<p>The default studies queried when searching for a single gene is defined with the <code v-pre>default_cross_cancer_study_session_id</code> or <code v-pre>default_cross_cancer_study_list</code> properties as described in the <doc-anchor-trigger to="#cross-cancer-study-query-default">cross cancer study query default section</doc-anchor-trigger>.</p>
<doc-anchor-target id="hide-sections-in-the-right-navigation-bar">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#hide-sections-in-the-right-navigation-bar">#</doc-anchor-trigger>
        <span>Hide sections in the right navigation bar</span>
    </h3>
</doc-anchor-target>
<p>Settings controlling what to show in the right navigation bar. Set them to <code v-pre>false</code> if you want to hide those sections, otherwise set the properties to <code v-pre>true</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">#Cancer Studies section:
skin.right_nav.show_data_sets=

#Example Queries section:
skin.right_nav.show_examples=

#Testimonials section:
skin.right_nav.show_testimonials=

#What's New section
skin.right_nav.show_whats_new=

#Twitter feed in What's New section
skin.right_nav.show_twitter=</code></pre>
</doc-codeblock></div>
<p>#Interactive tours section
skin.right_nav.show_web_tours=</p>
<doc-anchor-target id="control-the-content-of-specific-sections">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#control-the-content-of-specific-sections">#</doc-anchor-trigger>
        <span>Control the content of specific sections</span>
    </h3>
</doc-anchor-target>
<p>Setting controlling the blurb: you can add any HTML code here that you want to visualize. This will be shown between the cBioPortal menu and the Query selector in the main page.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.blurb=</code></pre>
</doc-codeblock></div>
<p>Setting controlling the citation below the blurb: you can add any HTML code here that you want to visualize. If the field is left empty, this HTML code will be shown: <code v-pre>Please cite: &lt;a href=&quot;https://cancerdiscovery.aacrjournals.org/content/2/5/401.abstract&quot; target=&quot;_blank&quot;&gt;Cerami et al., 2012&lt;/a&gt; &amp;amp; &lt;a href=&quot;https://www.ncbi.nlm.nih.gov/pubmed/23550210&quot; target=&quot;_blank&quot;&gt;Gao et al., 2013&lt;/a&gt;</code>).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.citation_rule_text=</code></pre>
</doc-codeblock></div>
<p>Setting controlling the footer: you can add any HTML code here that you want to visualize. If the field is left empty, the default footer (from <a href="http://www.cbioportal.org">www.cbioportal.org</a>) will be shown.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.footer=</code></pre>
</doc-codeblock></div>
<p>Setting controlling listing the development channels in the site&#x27;s footer section</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.footer_show_dev=</code></pre>
</doc-codeblock></div>
<p>Settings controlling the &quot;What&#x27;s New&quot; blurb in the right navigation bar: you can add any HTML code here that you want to visualize. If the field is left empty, the Twitter timeline will be shown (as long as <code v-pre>skin.right_nav.show_whats_new</code> is <code v-pre>true</code>, otherwise this section will not be displayed).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.right_nav.whats_new_blurb=</code></pre>
</doc-codeblock></div>
<p>Add custom logos to the left or right side of the header section. Place here the full name of the logo file (e.g. <code v-pre>logo.png</code>). This file should be present in <code v-pre>$PORTAL_HOME/portal/images/</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.left_logo=
skin.right_logo=</code></pre>
</doc-codeblock></div>
<p>Prevent users from saving data by removing all Download tabs and download and copy-to-clipboard controls.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.hide_download_controls=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="control-default-setting-for-filtering-of-genes-in-mutation-and-cna-tables-of-patient-view">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#control-default-setting-for-filtering-of-genes-in-mutation-and-cna-tables-of-patient-view">#</doc-anchor-trigger>
        <span>Control default setting for filtering of genes in mutation and CNA tables of patient view</span>
    </h3>
</doc-anchor-target>
<p>Different samples of a patient may have been analyzed with different gene panels. In patient view mutations and discrete CNA&#x27;s can be filtered based on whether the gene of respective mutations/CNA&#x27;s was profiled in all samples of the patient (mutations profiled in <code v-pre>all samples</code>), or not (mutations profiled in <code v-pre>any sample</code>). Setting this field to <code v-pre>true</code> will make patient view select the <code v-pre>all samples</code> filter at startup. When set to false or left blank the patient view will default to the <code v-pre>any samples</code> filter setting.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.patientview.filter_genes_profiled_all_samples=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="control-default-settings-of-the-vaf-line-chart-in-the-genomic-evolution-tab-of-patient-view">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#control-default-settings-of-the-vaf-line-chart-in-the-genomic-evolution-tab-of-patient-view">#</doc-anchor-trigger>
        <span>Control default settings of the VAF line chart in the genomic evolution tab of patient view</span>
    </h3>
</doc-anchor-target>
<p>If you want to enable log scale and sequential mode by default, set this property to <code v-pre>true</code>:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">vaf.log_scale.default=true|false
vaf.sequential_mode.default=true|false</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="control-unauthorized-studies-to-be-displayed-on-the-home-page">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#control-unauthorized-studies-to-be-displayed-on-the-home-page">#</doc-anchor-trigger>
        <span>Control unauthorized studies to be displayed on the home page</span>
    </h3>
</doc-anchor-target>
<p>By default, on an authenticated portal the home page will only show studies for which the current user is authorized. By setting the <em>skin.home_page.show_unauthorized_studies</em> property to <em>true</em> the home page will also show unauthorized studies. The unauthorized studies will appear greyed out and cannot be selected for downstream analysis in Results View or Study View.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.home_page.show_unauthorized_studies=</code></pre>
</doc-codeblock></div>
<p>If <em>show_unauthorized_studies</em> feature has been enabled, a message (template) can be defined by the property below that informs the user of insufficient permissions. This message will appear inside a tooltip when hovering the lock icon next to the study name on the Study Selection page. This message may contain placeholders for study-specific information derived from <a href="../../../file-formats/#study-tags-file">study tags data</a>. The information in the study tags JSON file can be accessed using Json Path placeholders. For example <code v-pre>{$.Owner.email}</code> points to member <code v-pre>{Owner: {email: &quot;me@myself.org&quot;}}</code>. For the studies that don&#x27;t have this information available in the study tags, the default message &quot;The study is unauthorized. You need to request access.&quot; will be displayed. In addition to the study tags information, the cancer study identifier can be included in the message using {<span class="math">.studyId} placeholder (does not have to be present in study tags file). For example: skin.home\_page.unauthorized\_studies\_global\_message=You do not have access to this study. You can request access with {</span>.Owner.email} (please mention the &#x27;{$.studyId}&#x27; study identifier).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.home_page.unauthorized_studies_global_message=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="show-badge-with-reference-genome">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#show-badge-with-reference-genome">#</doc-anchor-trigger>
        <span>Show badge with reference genome</span>
    </h3>
</doc-anchor-target>
<p>In instances with hg19 and hg38 studies you can show the reference genome in the home page next to the number of samples. This can be done setting this property to <code v-pre>true</code> (<code v-pre>false</code> by default):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.home_page.show_reference_genome=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="control-the-appearance-of-the-settings-menu-in-study-view-and-group-comparison-that-controls-custom-annotation-based-filtering">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#control-the-appearance-of-the-settings-menu-in-study-view-and-group-comparison-that-controls-custom-annotation-based-filtering">#</doc-anchor-trigger>
        <span>Control the appearance of the settings menu in study view and group comparison that controls custom annotation-based filtering</span>
    </h3>
</doc-anchor-target>
<p>A settings menu that allows the user to filter alterations in study view and group comparison may be used when <a href="../../../file-formats/#custom-driver-annotations">custom driver annotations</a> were loaded for the study or studies displayed in these sections. This menu will only appear, when setting the property <em>skin.show_settings_menu</em> to <em>true</em>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.show_settings_menu=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="hide-logout-button">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#hide-logout-button">#</doc-anchor-trigger>
        <span>Hide logout button</span>
    </h3>
</doc-anchor-target>
<p>When the user is logged-in, a button with logout and data access token options is shown at the right side of the header section. This button can be hidden by setting the property _skin.hide_logout_button to <em>true</em> (default is <em>false</em>). Hiding the logout button will remove the ability to perform a manual logout and download data access tokens by the user.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.hide_logout_button=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="namespace-columns-visible-in-mutation-table-by-default">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#namespace-columns-visible-in-mutation-table-by-default">#</doc-anchor-trigger>
        <span>Namespace columns visible in Mutation Table by default</span>
    </h3>
</doc-anchor-target>
<p>Namespace columns are custom columns in the MAF file that can be shown in Mutation Table components. By setting this property to <em>true</em> these columns will be visible when the mutation table is created. By default, the namespace column is hidden and can be made visible using the column selection menu.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.mutation_table.namespace_column.show_by_default=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="default-visible-columns-on-init-in-mutation-copy-number-and-structural-variant-tables">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#default-visible-columns-on-init-in-mutation-copy-number-and-structural-variant-tables">#</doc-anchor-trigger>
        <span>Default visible columns on init in Mutation, Copy-Number and Structural Variant Tables</span>
    </h3>
</doc-anchor-target>
<p>Define the columns that are going to be visible in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View. The <code v-pre>skin.mutation_table.namespace_column.show_by_default</code> takes precedence over these settings for namespace columns.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.results_view.mutation_table.columns.show_on_init=
skin.patient_view.mutation_table.columns.show_on_init=
skin.patient_view.copy_number_table.columns.show_on_init=
skin.patient_view.structural_variant_table.columns.show_on_init=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="default-sort-columns-on-mutation-copy-number-and-structural-variant-tables">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#default-sort-columns-on-mutation-copy-number-and-structural-variant-tables">#</doc-anchor-trigger>
        <span>Default sort columns on Mutation, Copy-Number and Structural Variant Tables</span>
    </h3>
</doc-anchor-target>
<p>Define the column that are going to sort be default in the Mutation, Copy-Number and Structural Variant Tables on the Patient View and the Mutation Table in the Results View.
Column name should be exactly the same as shown in tables.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.results_view.tables.default_sort_column=
skin.patient_view.tables.default_sort_column=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="define-custom-sample-type-colors">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#define-custom-sample-type-colors">#</doc-anchor-trigger>
        <span>Define custom sample type colors</span>
    </h3>
</doc-anchor-target>
<p>Define the colors of custom sample types in the patient view using a json object with for each sample type a color:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.patient_view.custom_sample_type_colors_json=classpath:/skin-patient-view-custom-sample-type-colors.json</code></pre>
</doc-codeblock></div>
<p>Example of json file contents:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-json"><code v-pre class="language-json">{
    &quot;Primary&quot;: &quot;green&quot;,
    &quot;Biopsy 3&quot;: &quot;#00c040ff&quot;
}</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="choose-the-display-name-for-authenticated-users">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#choose-the-display-name-for-authenticated-users">#</doc-anchor-trigger>
        <span>Choose the display name for authenticated users</span>
    </h3>
</doc-anchor-target>
<p>By default the display name for authenticated users is email, but it can be changed for the user name:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.user_display_name=username</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="hide-p--and-q-values-in-survival-types-table">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#hide-p--and-q-values-in-survival-types-table">#</doc-anchor-trigger>
        <span>Hide p- and q-values in survival types table</span>
    </h3>
</doc-anchor-target>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none"># Show/hide p- and q-values in survival types table (default is true)
survival.show_p_q_values_in_survival_type_table=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="set-the-initial-x-axis-limit-for-the-survival-plot">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#set-the-initial-x-axis-limit-for-the-survival-plot">#</doc-anchor-trigger>
        <span>Set the initial x-axis limit for the survival plot</span>
    </h3>
</doc-anchor-target>
<p>By default, the initial x-axis limit for a survival plot is the time of the latest event in the data. With this setting, you can make the initial x-axis limit be smaller than that. <img src="../../../images/previews/survival_x_axis_limit.png" alt="Survival x-axis limit examples" /></p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none"># Set initial x-axis limit for survival plot (by default, initial limit will be the latest event in the data)
survival.initial_x_axis_limit=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="display-installation-map">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#display-installation-map">#</doc-anchor-trigger>
        <span>Display installation map</span>
    </h3>
</doc-anchor-target>
<p>This setting specifies the URL for the iframes of a given portal&#x27;s installation map on the homepage and standalone page. By default, the installation map configuration is commented out with the iframes and standalone page hidden. The configuration below is used for <a href="https://www.cbioportal.org/">https://www.cbioportal.org/</a>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">installation_map_url=https://installationmap.netlify.app/</code></pre>
</doc-codeblock></div>
<p>To set up an installation map instance, one may consult the source code for the installation map <a href="https://github.com/cbioportal/installation-map">here</a>.</p>
<doc-anchor-target id="show-structural-variants-table-on-study-view">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#show-structural-variants-table-on-study-view">#</doc-anchor-trigger>
        <span>Show structural variants table on study view</span>
    </h3>
</doc-anchor-target>
<p>The structural variants table widget on study view allows users to define cohorts based on gene-orientation specific structural variant data format (see <a href="../../../file-formats/#structural-variant-data">here</a>).
The structural variants table widget supports cohort selection based on <em>gene1</em>, <em>gene2</em> and <em>gene1/gene2</em> orientation specific genomic events.
This property enables the structural variants table widget on Study View.</p>
<p><span class="docs-emoji">&#x26A0;&#xFE0F;</span> Although <em>gene1</em> and <em>gene2</em> specific queries may be used to investigate up- and downstream fusion partners, respectively, the validity of this
depends on supports for this interpretation in the underlying data.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.study_view.show_sv_table=true</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="ensembl-transcript-lookup-url">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#ensembl-transcript-lookup-url">#</doc-anchor-trigger>
        <span>Ensembl transcript lookup URL</span>
    </h2>
</doc-anchor-target>
<p>The Mutations tab contains various links, redirecting the user to external information resources regarding the displayed transcript. The Ensembl template URL can be customized by modifying the property:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">ensembl.transcript_url=</code></pre>
</doc-codeblock></div>
<p>The default setting is <code v-pre>https://ensembl.org/homo_sapiens/Transcript/Summary?t=&lt;%= transcriptId %&gt;</code>. The <code v-pre>&lt;%= transcriptId %&gt;</code> is substituted by the frontend code into respective transcript ID.</p>
<doc-anchor-target id="segment-file-url">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#segment-file-url">#</doc-anchor-trigger>
        <span>Segment File URL</span>
    </h2>
</doc-anchor-target>
<p>This is a root URL to where segment files can be found. This is used when you want to provide segment file viewing via external tools such as <a href="https://www.broadinstitute.org/igv/">IGV</a>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">segfile.url=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="bitly-api-username-and-key">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#bitly-api-username-and-key">#</doc-anchor-trigger>
        <span>Bitly API Username and Key</span>
    </h2>
</doc-anchor-target>
<p>The following properties are used to provide shortened bookmarks to the cBioPortal:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">bitly.user=
bitly.api_key=</code></pre>
</doc-codeblock></div>
<p>To obtain a bitly username and key, first register at: <a href="https://bitly.com/">https://bitly.com/</a></p>
<p>Then, go to: <a href="https://bitly.com/a/your%5C_api%5C_key">https://bitly.com/a/your\_api\_key</a></p>
<p><strong>Note:</strong> If you are developing on a local machine, and using localhost, the bitly URL shortening service will not work. This is because bitly will not shorten URLs for localhost. Once you deploy to your final server, the issue should go away.</p>
<doc-anchor-target id="google-analytics">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#google-analytics">#</doc-anchor-trigger>
        <span>Google Analytics</span>
    </h2>
</doc-anchor-target>
<p>When the following property is defined, Google Analytics will track site usage.</p>
<p><span class="docs-emoji">&#x26A0;&#xFE0F;</span> In contrast to what this property&#x27;s name suggests, this should be the <strong>GA Universal Analytics ID</strong> (UA-XXXXX-YY), not to be confused with the new <a href="https://support.google.com/analytics/answer/10089681">GA-4 ID, introduced in 2020</a>. A Universal Analytics GA property can be created at <a href="https://analytics.google.com/analytics/web/provision/">Google Analytics</a>, create a new GA property, and make sure to click <strong>Show advanced options</strong> and switch on <strong>Create a Universal Analytics property</strong>. After creating a Universal Analytics property, the tracking ID can be found at <strong>Admin</strong>, column <strong>property</strong>, under <strong>Property Settings</strong></p>
<p><span class="docs-emoji">&#x26A0;&#xFE0F;</span> Adblockers may block site tracking by GA</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">google_analytics_profile_id</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="password-authentication">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#password-authentication">#</doc-anchor-trigger>
        <span>Password Authentication</span>
    </h2>
</doc-anchor-target>
<p>The portal supports password authentication via Google+. Before you start you need to setup a google account that will own the authentication API. Follow <a href="https://developers.google.com/identity/sign-in/web/devconsole-project">https://developers.google.com/identity/sign-in/web/devconsole-project</a> to get clientID and secret. Fill it in application.properties:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">googleplus.consumer.key=195047654890-499gl89hj65j8d2eorqe0jvjnfaxcln0.apps.googleusercontent.com
googleplus.consumer.secret=2jCfg4SPWdGfXF44WC588dK</code></pre>
</doc-codeblock></div>
<p>(note: these are just examples, you need to get your own) You will also need to go to &quot;Google+ API&quot; and click Enable button. In case of problems make sure to enable DEBUG logging for org.springframework.social and org.springframework.security.web.authentication.</p>
<p>To activate password authentication follow the <a href="../../deploy-without-docker/deploying/#required-login">Deployment with authentication steps</a> and set <code v-pre>authenticate=googleplus</code>.</p>
<p>In addition, set this property in <code v-pre>application.properties</code>:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">app.name=cbioportal</code></pre>
</doc-codeblock></div>
<p>app.name should be set to the name of the portal instance referenced in the &quot;AUTHORITY&quot; column of the &quot;AUTHORITIES&quot; table. See the <a href="../../authorization-and-authentication/user-authorization/">User Authorization</a> for more information.</p>
<doc-anchor-target id="oncokb-integration">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oncokb-integration">#</doc-anchor-trigger>
        <span>OncoKB integration</span>
    </h2>
</doc-anchor-target>
<p>OncoKB integration can be turned on or off with the following property (default: true):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">show.oncokb=true|false</code></pre>
</doc-codeblock></div>
<p>A private token is required to access the OncoKB Data (for details see the section <a href="../../integration-with-other-webservices/oncokb-data-access/">OncoKB Data Access</a>):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncokb.token=</code></pre>
</doc-codeblock></div>
<p>&quot;cBioPortal&gt;=&quot; driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">show.cbioportal=true|false</code></pre>
</doc-codeblock></div>
<p>&quot;COSMIC&gt;=&quot; driver annotation sources in the settings menu of Results View can be hidden by turning off the following property (default: true):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">show.cosmic=true|false</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="civic-integration">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#civic-integration">#</doc-anchor-trigger>
        <span>CIViC integration</span>
    </h2>
</doc-anchor-target>
<p>CIViC integration can be turned on or off with the following property (default: true):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">show.civic=true|false</code></pre>
</doc-codeblock></div>
<p>The CIViC API url is set to <a href="https://civic.genome.wustl.edu/api/">https://civic.genome.wustl.edu/api/</a> by default. It can be overridden using the following property:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">civic.url=</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="genome-nexus-integration">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#genome-nexus-integration">#</doc-anchor-trigger>
        <span>Genome Nexus Integration</span>
    </h2>
</doc-anchor-target>
<p>Genome Nexus provides annotations of mutations in cBioPortal. The mutations tab relies heavily on the Genome Nexus API, therefore that tab won&#x27;t work well without it. By default cBioPortal will use the public Genome Nexus API, such that no extra configuration is necessary.</p>
<doc-anchor-target id="genome-build">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#genome-build">#</doc-anchor-trigger>
        <span>Genome Build</span>
    </h3>
</doc-anchor-target>
<p>Genome Nexus supports both GRCh37 and GRCh38, but support for the latter is limited. Several annotation sources served by Genome Nexus might not have official GRCh38 support yet i.e. <a href="https://www.oncokb.org">OncoKB</a>, <a href="https://civicdb.org">CIViC</a>, <a href="https://www.cancerhotspots.org">Cancer Hotspots</a>, <a href="https://www.mycancergenome.org">My Cancer Genome</a> and <a href="https://g2s.genomenexus.org">3D structures</a>. Although most of the time the canonical transcript for a gene will be the same between GRCh37 and GRCh38 there might be some that cause issues. In addition the complete integration of cBioPortal with Genome Nexus&#x27; GRCh38 is not complete yet. That is cBioPortal currently only connects to one Genome Nexus API by default (the GRCh37 one), so it&#x27;s not possible to have multiple genome builds in one instance of cBioPortal and get the correct annotations from Genome Nexus for both. Currently only the <a href="https://www.cbioportal.org/mutation_mapper">mutation mapper tool page</a> is able to handle both.</p>
<doc-anchor-target id="properties">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#properties">#</doc-anchor-trigger>
        <span>Properties</span>
    </h3>
</doc-anchor-target>
<p>By default the Genome Nexus API url is set to <a href="https://v1.genomenexus.org/">https://v1.genomenexus.org/</a>, which uses GRCh37. It can be overridden using the following property:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">genomenexus.url=</code></pre>
</doc-codeblock></div>
<p>Genome Nexus provides a set of mappings from Hugo genes names to Ensembl transcript IDs. There are two mappings: <code v-pre>mskcc</code> and <code v-pre>uniprot</code>.
You can read more about the difference between those in the <a href="../../../mutation-data-transcript-annotation/">Mutation Data Annotation Section</a>. The default is currently <code v-pre>uniprot</code>, but we recommend new installers to use <code v-pre>mskcc</code> and people with older installations to consider migrating. The property
can be changed with:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">genomenexus.isoform_override_source=mskcc</code></pre>
</doc-codeblock></div>
<p>The mutation mapper tool page can annotate GRCh38 coordinates. By default it uses <a href="https://grch38.genomenexus.org">https://grch38.genomenexus.org</a>. It can be overridden by setting:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">genomenexus.url.grch38=</code></pre>
</doc-codeblock></div>
<p>The GRCh38 annotation in mutation mapper can be hidden by setting <code v-pre>show.mutation_mappert_tool.grch38=false</code>, by default it&#x27;s set to <code v-pre>true</code>.</p>
<doc-anchor-target id="mdacc-heatmap-integration">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#mdacc-heatmap-integration">#</doc-anchor-trigger>
        <span>MDACC Heatmap Integration</span>
    </h2>
</doc-anchor-target>
<p>MDACC Heatmap integration (button in OncoPrint heatmap dropdown and tab on Study page can be turned on or off by setting the following property:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">show.mdacc.heatmap=true</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="oncoprint">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#oncoprint">#</doc-anchor-trigger>
        <span>OncoPrint</span>
    </h2>
</doc-anchor-target>
<p>The default view in OncoPrint (&quot;patient&quot; or &quot;sample&quot;) can be set with the following option. The default is &quot;patient&quot;.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.defaultview=sample</code></pre>
</doc-codeblock></div>
<p>Configuration of tracks that will be visible by default in the oncoprint. It points to a JSON file on the classpath.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.clinical_tracks.config_json=classpath:/oncoprint-default-tracks.json</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="custom-annotation-of-driver-and-passenger-mutations">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#custom-annotation-of-driver-and-passenger-mutations">#</doc-anchor-trigger>
        <span>Custom annotation of driver and passenger mutations</span>
    </h2>
</doc-anchor-target>
<p>cBioPortal supports 2 formats to add custom annotations for driver and passenger mutations.</p>
<ol>
<li><strong>cbp_driver</strong>: This will define whether a mutation is a driver or not.</li>
<li><strong>cbp_driver_tiers</strong>: This can be used to define multiple classes of driver mutations.</li>
</ol>
<p>These data formats are described in the <a href="../../../file-formats/#extended-maf-format">cBioPortal MAF specifications</a>.</p>
<p><strong>Enabling custom annotations in the OncoPrint</strong></p>
<p>To enable functionality for one or both types of custom annotations, enter values for the following properties. These values will appear in the OncoPrint&#x27;s &quot;Mutation color&quot; menu, Patient View&#x27;s (mutation, CNA, SV) tables, Results View&#x27;s mutation table, and Group Comparison View&#x27;s mutation table.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.custom_driver_annotation.binary.menu_label=Custom Driver
oncoprint.custom_driver_annotation.binary.menu_description=Custom driver tiers
oncoprint.custom_driver_annotation.tiers.menu_label=Custom Driver Tiers
oncoprint.custom_driver_annotation.tiers.menu_description=Custom driver tiers</code></pre>
</doc-codeblock></div>
<p><strong>Automatic selection of OncoKB, hotspots and custom annotations</strong></p>
<p>OncoKB and Hotspots are by default automatically selected as annotation source, if <code v-pre>show.oncokb</code> and <code v-pre>show.hotspots</code> are set to <code v-pre>true</code>. To add automatic selection of custom driver or custom driver tiers annotations, set the respective property to <code v-pre>true</code>. Default is <code v-pre>false</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.custom_driver_annotation.binary.default=true|false
oncoprint.custom_driver_annotation.tiers.default=true|false</code></pre>
</doc-codeblock></div>
<p>If you want to disable the automatic selection of OncoKB and hotspots as annotation source, set these properties to <code v-pre>false</code>:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.oncokb.default=true|false
oncoprint.hotspots.default=true|false</code></pre>
</doc-codeblock></div>
<p>If you want to enable oncoprint heatmap clustering by default, set this property to <code v-pre>true</code>:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.clustered.default=true|false</code></pre>
</doc-codeblock></div>
<p><strong>Automatic hiding of variants of unknown significance (VUS)</strong></p>
<p>By default, the selection box to hide VUS mutations is unchecked. If you want to automatically hide VUS, set this property to <code v-pre>true</code>. Default is <code v-pre>false</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">oncoprint.hide_vus.default=true|false</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="gene-sets-used-for-gene-querying">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#gene-sets-used-for-gene-querying">#</doc-anchor-trigger>
        <span>Gene sets used for gene querying</span>
    </h2>
</doc-anchor-target>
<p>To change the gene sets used for gene querying, create a JSON file and add gene sets, following the format specified in the examples below. Set the path to this file (e.g. <code v-pre>file:/cbioportal/custom_gene_sets.json</code>) in the following property and restart Tomcat to apply the update. The default gene sets will be replaced by the ones in <code v-pre>custom_gene_sets.json</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">querypage.setsofgenes.location=file:/&lt;path&gt;</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="example-with-gene-names">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#example-with-gene-names">#</doc-anchor-trigger>
        <span>Example with gene names</span>
    </h3>
</doc-anchor-target>
<p>In this example, two gene sets will appear in the query page, under the names &quot;Prostate Cancer: AR Signaling&quot; and &quot;Prostate Cancer: AR and steroid synthesis enzymes&quot;.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">[{
	&quot;id&quot;: &quot;Prostate Cancer: AR Signaling&quot;,
	&quot;genes&quot;: [&quot;SOX9&quot;, &quot;RAN&quot;, &quot;TNK2&quot;, &quot;EP300&quot;, &quot;PXN&quot;, &quot;NCOA2&quot;, &quot;AR&quot;, &quot;NRIP1&quot;, &quot;NCOR1&quot;, &quot;NCOR2&quot;]
}, {
	&quot;id&quot;: &quot;Prostate Cancer: AR and steroid synthesis enzymes&quot;,
	&quot;genes&quot;: [&quot;AKR1C3&quot;, &quot;AR&quot;, &quot;CYB5A&quot;, &quot;CYP11A1&quot;, &quot;CYP11B1&quot;, &quot;CYP11B2&quot;, &quot;CYP17A1&quot;, &quot;CYP19A1&quot;, &quot;CYP21A2&quot;, &quot;HSD17B1&quot;, &quot;HSD17B10&quot;, &quot;HSD17B11&quot;, &quot;HSD17B12&quot;, &quot;HSD17B13&quot;, &quot;HSD17B14&quot;, &quot;HSD17B2&quot;, &quot;HSD17B3&quot;, &quot;HSD17B4&quot;, &quot;HSD17B6&quot;, &quot;HSD17B7&quot;, &quot;HSD17B8&quot;, &quot;HSD3B1&quot;, &quot;HSD3B2&quot;, &quot;HSD3B7&quot;, &quot;RDH5&quot;, &quot;SHBG&quot;, &quot;SRD5A1&quot;, &quot;SRD5A2&quot;, &quot;SRD5A3&quot;, &quot;STAR&quot;]
}]</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="example-with-specific-alterations">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#example-with-specific-alterations">#</doc-anchor-trigger>
        <span>Example with specific alterations</span>
    </h3>
</doc-anchor-target>
<p>In this example, only one gene set will appear in the query page, under the name &quot;Genes with alterations&quot;, which will add the different genetic alterations stated below in the query box.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">[{
	&quot;id&quot;: &quot;Genes with alterations&quot;,
	&quot;genes&quot;: [&quot;TP53: MUT=R273C;&quot;, &quot;KRAS: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;&quot;]
}]</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="example-with-merged-gene-tracks">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#example-with-merged-gene-tracks">#</doc-anchor-trigger>
        <span>Example with merged gene tracks</span>
    </h3>
</doc-anchor-target>
<p>In this example, only one gene set will appear in the query page, under the name &quot;BRCA genes test&quot;, containing the merged gene track called &quot;BRCA genes&quot;.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">[{
	&quot;id&quot;: &quot;BRCA genes test&quot;,
	&quot;genes&quot;: [&quot;[\\\&quot;BRCA genes\\\&quot; BRCA1: MUT=E1258D;&quot;, &quot;BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;]&quot;]
}]</code></pre>
</doc-codeblock></div>
<p>This gene set will add the following in the query box:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">&quot;BRCA genes&quot; BRCA1: MUT=E1258D; BRCA2: HOMDEL MUT=NONSENSE MUT=NONSTART MUT=NONSTOP MUT=FRAMESHIFT MUT=SPLICE MUT=TRUNC;</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="cache-settings">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#cache-settings">#</doc-anchor-trigger>
        <span>Cache Settings</span>
    </h2>
</doc-anchor-target>
<p>cBioPortal is supported on the backend with Ehcache or Redis. These caches are configurable from within application.properties through the following properties.</p>
<p>The cache type is set using <code v-pre>persistence.cache_type</code>. Valid values are <code v-pre>no-cache</code>, <code v-pre>redis</code> (redis), <code v-pre>ehache-heap</code> (ehcache heap-only), <code v-pre>ehache-disk</code> (ehcache disk-only), and <code v-pre>ehache-hybrid</code> (ehcache disk + heap). By default, <code v-pre>persistence.cache_type</code> is set to <code v-pre>no-cache</code> which disables the cache. When the cache is disabled, no responses will be stored in the cache.</p>
<p><span class="docs-emoji">&#x26A0;&#xFE0F;</span> the &#x27;redis&#x27; caching option will likely cause a conflict when installing the portal in a Tomcat installation which uses redisson for session management. If you plan to deploy cbioportal to such a system, avoid the &#x27;redis&#x27; caching option for <code v-pre>persistence.cache_type</code> and be sure to build cbioportal.war with the maven option <code v-pre>-Dexclude-redisson</code> (see <a href="../../deploy-without-docker/build-from-source/#building-with-maven">Building with Maven</a>).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">persistence.cache_type=[no-cache or ehache-heap or ehcache-disk or ehcache-hybrid or redis]</code></pre>
</doc-codeblock></div>
<p>Logged metrics and additional information such as cache size and cached keys are available through an optional endpoint. The optional endpoint is turned off by default but can be turned on by setting <code v-pre>cache.statistics_endpoint_enabled</code> to true.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">cache.statistics_endpoint_enabled=false[true or false]</code></pre>
</doc-codeblock></div>
<p>The cache statistics endpoint is hidden on the api page; users must directly access the URL to view the response. The cache statistics endpoint can be accessed in the following ways.</p>
<p>For a list of all keys in the cache:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">/api/[name of cache]/keysInCache</code></pre>
</doc-codeblock></div>
<p>For a list of counts of keys in cache per repository class:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">/api/[name of cache]/keyCountsPerClass</code></pre>
</doc-codeblock></div>
<p>For general statistics about the cache such as memory usage (not currently implemented for Redis):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">/api/cacheStatistics</code></pre>
</doc-codeblock></div>
<p><strong>WARNING</strong>: It must be noted that since cache statistics endpoint returns data on cache keys, the endpoint may expose otherwise hidden database query parameters such as sample identifiers, study names, etc. Generally, it is recommended that the endpoint only be turned on during cache-related development for testing. Deployers of a protected portal where users only have authorities to a subset of studies should carefully consider whether or not to turn on the cache statistics endpoint, as it does not filter the results.</p>
<p>For more information on how caching is implemented in cBioPortal refer to the <a href="../caching/">Caching</a> documentation.</p>
<doc-anchor-target id="redis">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#redis">#</doc-anchor-trigger>
        <span>Redis</span>
    </h3>
</doc-anchor-target>
<p>To cache with Redis set <code v-pre>persistence.cache_type</code> to <code v-pre>redis</code>.</p>
<p>To setup the Redis cache servers the following properties are required:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">redis.name=unique_cbioportal_instance_name
redis.leader_address=redis://{host/servicename}:6379
redis.follower_address=redis://{host/servicename}:6379
redis.database=0
redis.password=password</code></pre>
</doc-codeblock></div>
<p>If you are running one redis instance for multiple instances of cBioPortal, one can use the properties <code v-pre>redis.name</code> and <code v-pre>redis.database</code> to avoid clashes. If you are running only one instance of cBioPortal any value for name/database will do.</p>
<p>There are also some optional parameters:</p>
<p><code v-pre>redis.clear_on_startup</code>: If <code v-pre>true</code>, the caches will clear on startup. This is important to do to avoid reading old study data from the cache. You may want to turn it off and clear redis yourself if you are running in a clustered environments, as you&#x27;ll have frequent restarts that do not require you to clear the redis cache.<br />
<code v-pre>redis.ttl_mins</code>: The time to live of items in the general cache, in minutes. The default value is 10000, or just under 7 days.</p>
<p>For more information on Redis, refer to the official documentation <a href="https://redis.io/documentation">here</a></p>
<doc-anchor-target id="ehcache">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#ehcache">#</doc-anchor-trigger>
        <span>Ehcache</span>
    </h3>
</doc-anchor-target>
<p>To cache with Ehcache set <code v-pre>persistence.cache_type</code> to <code v-pre>ehache-heap</code> (ehcache heap-only), <code v-pre>ehache-disk</code> (ehcache disk-only), or <code v-pre>ehache-hybrid</code> (ehcache disk + heap).</p>
<p>Ehcache initializes caches using a template found in an Ehcache xml configuration file. When caching is enabled, set <code v-pre>ehcache.xml_configuration</code> to the name of the Ehcache xml configuration file. The default provided is <code v-pre>ehcache.xml</code>; to change the cache template, directly edit this file. Alternatively, you can create your own Ehcache xml configuration file, place it under <code v-pre>/persistence/persistence-api/src/main/resources/</code> and set <code v-pre>ehcache.xml_configuration</code> to <code v-pre>/[Ehcache xml configuration filename]</code>.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">ehcache.xml_configuration=</code></pre>
</doc-codeblock></div>
<p>If the cache is configured to use disk resources, users must make a directory available and set it with the <code v-pre>ehcache.persistence_path</code> property. Ehcache will create separate directories under the provided path for each cache defined in the ehcache.xml_configuration file.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">ehcache.persistence_path=[location on the disk filesystem where Ehcache can write the cache to /tmp/]</code></pre>
</doc-codeblock></div>
<p>Cache size must be set for heap and/or disk depending on which are in use; Ehcache requires disk size to be greater than heap size in a hybrid configuration. Zero is not a supported size and will cause an exception. Units are in megabytes. Default values are provided. The general repository cache is specified to use 1024MB of heap and 4096MB of disk. The static repository cache is specified to use 30MB of heap and 32MB of disk. For installations with increased traffic or data, cache sizes can be increased to further improve performance.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">ehcache.general_repository_cache.max_mega_bytes_heap=
ehcache.general_repository_cache.max_mega_bytes_local_disk=

ehcache.static_repository_cache_one.max_mega_bytes_heap=
ehcache.static_repository_cache_one.max_mega_bytes_local_disk=</code></pre>
</doc-codeblock></div>
<p>For more information on Ehcache, refer to the official documentation <a href="https://www.ehcache.org/documentation/3.7/index.html">here</a></p>
<doc-anchor-target id="evict-caches-with-the-apicache-endpoint">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#evict-caches-with-the-apicache-endpoint">#</doc-anchor-trigger>
        <span>Evict caches with the /api/cache endpoint</span>
    </h2>
</doc-anchor-target>
<p><code v-pre>DELETE</code> http requests to the <code v-pre>/api/cache</code> endpoint will flush the cBioPortal caches, and serves as an alternative to restarting the cBioPortal application.</p>
<p>By default the endpoint is disabled. The endpoint can be enabled by setting:</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">cache.endpoint.enabled=true</code></pre>
</doc-codeblock></div>
<p>Access to the endpoint is not regulated by the configured user authorization mechanism. Instead, an API key should be passed with the <code v-pre>X-API-KEY</code> header. The accepted value for the API key can be configured by setting (for example):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">cache.endpoint.api-key=7d70fecb-cda8-490f-9ea2-ef874b6512f4</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="delegate-user-authorization-cache-to-spring-managed-cache">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#delegate-user-authorization-cache-to-spring-managed-cache">#</doc-anchor-trigger>
        <span>Delegate user-authorization cache to Spring-managed cache</span>
    </h2>
</doc-anchor-target>
<p>For evaluation fo user permissions cBioPortal uses a user-authorization cache that is populated at startup. By setting the <code v-pre>cache.cache-map-utils.spring-managed</code> property to <em>true</em> this cache will be managed by the Spring caching solution such as EHCache or Redis. For more extended information, see <a href="../caching/#user-authorization-cache">here</a></p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">cache.cache-map-utils.spring-managed=false</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="enable-gsva-functionality">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#enable-gsva-functionality">#</doc-anchor-trigger>
        <span>Enable GSVA functionality</span>
    </h2>
</doc-anchor-target>
<p><a href="https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#gene-set-data">GSVA functionality</a> can be enabled by uncommenting this line (and making sure it is set to <code v-pre>true</code>):</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.show_gsva=true</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="set-default-thresholds-for-geneset-hierarchy">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#set-default-thresholds-for-geneset-hierarchy">#</doc-anchor-trigger>
        <span>Set default thresholds for geneset hierarchy</span>
    </h2>
</doc-anchor-target>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.geneset_hierarchy.default_gsva_score=0.3
skin.geneset_hierarchy.default_p_value=0.02</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="collapses-the-tree-widget-of-the-geneset-hierarchy-dialog-on-initialization">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#collapses-the-tree-widget-of-the-geneset-hierarchy-dialog-on-initialization">#</doc-anchor-trigger>
        <span>Collapses the tree widget of the geneset hierarchy dialog on initialization</span>
    </h2>
</doc-anchor-target>
<p>By default, the tree is expanded (property value is <code v-pre>false</code>).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">skin.geneset_hierarchy.collapse_by_default = true</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="cross-study-expression-and-protein-data">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#cross-study-expression-and-protein-data">#</doc-anchor-trigger>
        <span>Cross study expression and protein data</span>
    </h2>
</doc-anchor-target>
<p>By default we hide expression data for multi-study queries as they are
usually not normalized across studies. For the public cBioPortal for instance, only TCGA Pancan Atlas studies expression data has been normalized.</p>
<p>If you know the expression data in your instance is comparable, or is comparable for a subset of
studies, you can configure a rule as follows.</p>
<p>The value of this property can be boolean (true|false) or a javascript function
which executes at runtime and is passed the list of study objects being queried
by the user and evaluates whether expression data can be safely displayed.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">// a function that accepts the users's selected studies and
// returns whether or not to allow expression data from the involved studies to be mixed
enable_cross_study_expression = (selectedStudies)=&gt;{ [your logic] return true|false }</code></pre>
</doc-codeblock></div>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">// boolean
enable_cross_study_expression = true|false</code></pre>
</doc-codeblock></div>
<doc-anchor-target id="combined-study-view-summary-limits">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#combined-study-view-summary-limits">#</doc-anchor-trigger>
        <span>Combined Study View Summary Limits</span>
    </h2>
</doc-anchor-target>
<doc-anchor-target id="background">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#background">#</doc-anchor-trigger>
        <span>Background</span>
    </h3>
</doc-anchor-target>
<p>A limit is added to prevent poor performance of Study View when selecting too large sample numbers.</p>
<doc-anchor-target id="properties-1">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#properties-1">#</doc-anchor-trigger>
        <span>Properties</span>
    </h3>
</doc-anchor-target>
<ul>
<li><code v-pre>studyview.max_samples_selected</code>: Limit is disabled when not set</li>
</ul>
<doc-anchor-target id="behavior">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#behavior">#</doc-anchor-trigger>
        <span>Behavior</span>
    </h3>
</doc-anchor-target>
<p>When these limits are exceeded the &quot;Explore Selected Studies&quot; button will be disabled on the Study View Page.</p>
<doc-anchor-target id="request-body-compression">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#request-body-compression">#</doc-anchor-trigger>
        <span>Request Body Compression</span>
    </h2>
</doc-anchor-target>
<doc-anchor-target id="background-1">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#background-1">#</doc-anchor-trigger>
        <span>Background</span>
    </h3>
</doc-anchor-target>
<p>Some REST endpoints that the cBioPortal frontend uses have request bodies that scale as your dataset increases. In portals where users commonly query more than 100,000 samples, we found that some of these request bodies could get as large as 20 Mb. These large request bodies pose a significant problem for users with poor upload speeds - some users experienced upload times of more than five minutes for these requests. Request body compression is our temporary solution to this problem. When this feature is toggled on, we compress the request bodies of a few problematic endpoints.</p>
<doc-anchor-target id="properties-2">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#properties-2">#</doc-anchor-trigger>
        <span>Properties</span>
    </h3>
</doc-anchor-target>
<p>There are two <code v-pre>portal.property</code> values related to this feature:</p>
<ul>
<li><code v-pre>enable_request_body_gzip_compression</code>: when <code v-pre>true</code>, the feature will be enabled.</li>
<li><code v-pre>request_gzip_body_size_bytes</code>: the maximum allowable unzipped request body in bytes. Defaults to 80000000 (80 Mb).</li>
</ul>
<doc-anchor-target id="behavior-1">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#behavior-1">#</doc-anchor-trigger>
        <span>Behavior</span>
    </h3>
</doc-anchor-target>
<ul>
<li>This is a nonbreaking change. Any consumers of the cBioPortal API you have that send requests with uncompressed request bodies will continue to work, regardless of whether you turn this feature on or off.</li>
<li>If you turn this feature on, the cBioPortal API will now be able to handle any request with a gzipped request body, provided:
<ul>
<li>It is a POST request.</li>
<li>It has a <code v-pre>Content-Encoding: gzip</code> header.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="reasons-to-enable-this-feature">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#reasons-to-enable-this-feature">#</doc-anchor-trigger>
        <span>Reasons to Enable This Feature</span>
    </h3>
</doc-anchor-target>
<ul>
<li>You have studies with tens of thousands of samples.</li>
<li>You have users with poor upload speeds (&lt; 1mb up).</li>
</ul>
<doc-anchor-target id="reasons-to-disable-this-feature">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#reasons-to-disable-this-feature">#</doc-anchor-trigger>
        <span>Reasons to Disable This Feature</span>
    </h3>
</doc-anchor-target>
<ul>
<li>It is harder to debug gzipped requests
<ul>
<li>Chrome&#x27;s <code v-pre>copy request as CURL</code> will not work.</li>
<li>The compressed request body is not human-readable.</li>
</ul>
</li>
<li>It is a potential vector for denial of memory attacks.
<ul>
<li>Any request that has a body that takes significantly more space in memory than it does in the request body is potentially dangerous. We try to mitigate this by limiting the size of the unzipped request body via the <code v-pre>request_gzip_body_size_bytes</code> property, but at a fundamental level, this is still a concern.</li>
<li>Along these lines, if you do enable this feature, setting <code v-pre>request_gzip_body_size_bytes</code> to an arbitrarily large number would be unwise.</li>
</ul>
</li>
<li>This is not a cure-all for performance issues
<ul>
<li>Most requests the cBioPortal makes do not have large request bodies, so most requests will not be compressed, and will see no performance improvement.</li>
<li>Users with good upload speeds will see minimal performance improvements, as their upload speed is not a bottleneck.</li>
</ul>
</li>
</ul>
<doc-anchor-target id="datasets-tab-study-download-links">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#datasets-tab-study-download-links">#</doc-anchor-trigger>
        <span>DataSets Tab (Study Download Links)</span>
    </h1>
</doc-anchor-target>
<doc-anchor-target id="background-2">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#background-2">#</doc-anchor-trigger>
        <span>Background</span>
    </h3>
</doc-anchor-target>
<p>The DataSets tab has the ability to create a <code v-pre>download</code> button that allows users to quickly download &quot;raw&quot; public studies.</p>
<doc-anchor-target id="properties-3">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#properties-3">#</doc-anchor-trigger>
        <span>Properties</span>
    </h3>
</doc-anchor-target>
<ul>
<li><code v-pre>study_download_url</code> : when set, the feature will be enabled</li>
</ul>
<doc-anchor-target id="behavior-2">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#behavior-2">#</doc-anchor-trigger>
        <span>Behavior</span>
    </h3>
</doc-anchor-target>
<p>For private instances that want to replicate the<code v-pre>public-portal</code> they first must set up their studies
they want available for download in a similar format to what is described in the Example section below.
The studies are located on the <code v-pre>public-portal</code> at  <code v-pre>https://cbioportal-datahub.s3.amazonaws.com/</code>.
Then there is a <code v-pre>study_list.json</code> defined that list the studies that can be downloaded.
The studies to be downloaded need to be compressed with the extension <code v-pre>tar.gz</code></p>
<doc-anchor-target id="example">
    <h3>
        <doc-anchor-trigger class="header-anchor-trigger" to="#example">#</doc-anchor-trigger>
        <span>Example</span>
    </h3>
</doc-anchor-target>
<ul>
<li>We have set <code v-pre>study_download_url</code> property to <code v-pre>https://cbioportal-datahub.s3.amazonaws.com/</code></li>
<li><code v-pre>study_list.json</code> resides <code v-pre>https://cbioportal-datahub.s3.amazonaws.com/study_list.json</code>
<ul>
<li><code v-pre>[ &quot;acbc_mskcc_2015&quot;, &quot;acc_2019&quot;]</code> Example of contents</li>
</ul>
</li>
<li><code v-pre>acbc_mskcc_2015.tar.gz</code> resides <code v-pre>https://cbioportal-datahub.s3.amazonaws.com/acbc_mskcc_2015.tar.gz</code></li>
</ul>
<doc-anchor-target id="prioritized-studies-on-study-selector-view">
    <h1>
        <doc-anchor-trigger class="header-anchor-trigger" to="#prioritized-studies-on-study-selector-view">#</doc-anchor-trigger>
        <span>Prioritized studies on study selector view</span>
    </h1>
</doc-anchor-target>
<p>By default, the studies loaded into a local cBioPortal instance are organized based on their cancer type (i.e. Breast &gt;&gt; Other).</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">priority_studies=</code></pre>
</doc-codeblock></div>
<p>The value of this variable will create a custom category with studies on the top of the study selector view. The format for the string should be category1#study1a,study1b,study1c;category2#study2 (e.g., PanCancer Studies#msk_impact_2017), where the <code v-pre>category</code> can be any string and the <code v-pre>study</code> should be the study ID of the required uploaded study.</p>
<doc-anchor-target id="study-tag-functionality">
    <h2>
        <doc-anchor-trigger class="header-anchor-trigger" to="#study-tag-functionality">#</doc-anchor-trigger>
        <span>Study Tag functionality</span>
    </h2>
</doc-anchor-target>
<p>Study Tags allow portal maintainers to define miscellaneous descriptive meta data to studies, which will be shown to users in tooltips and are also searchable. This feature
is on by default but can be disabled using the following property.</p>
<div class="codeblock-wrapper"><doc-codeblock>
<pre class="language-none"><code v-pre class="language-none">//boolean
enable_study_tags=true|false</code></pre>
</doc-codeblock></div>

                                
                                <!-- Required only on API pages -->
                                <doc-toolbar-member-filter-no-results></doc-toolbar-member-filter-no-results>
                            </div>
                            <footer class="clear-both">
                                <div class="flex flex-wrap items-center justify-between mt-14">
                                    <a class="my-2.5 inline-flex items-center text-sm whitespace-nowrap text-blue-500 dark:text-blue-400 hover:text-blue-700 hover:underline" href="https://github.com/cbioportal/cbioportal/blob/master/docs/deployment/customization/application.properties-Reference.md" target="_blank" rel="noopener">
                                        <svg class="mr-1.5" xmlns="http://www.w3.org/2000/svg" width="18" height="18" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M20 12c-.55 0-1 .45-1 1v7c0 .55-.45 1-1 1H4c-.55 0-1-.45-1-1V6c0-.55.45-1 1-1h7c.55 0 1-.45 1-1s-.45-1-1-1H4C2.35 3 1 4.35 1 6v14c0 1.65 1.35 3 3 3h14c1.65 0 3-1.35 3-3v-7c0-.55-.45-1-1-1z" /><path d="M22.21 1.79c-1.18-1.18-3.24-1.18-4.41 0l-9.5 9.5c-.13.13-.22.29-.26.46l-1 4c-.08.34.01.7.26.95.18.2.44.3.7.3.08 0 .16-.01.24-.03l4-1c.18-.04.34-.13.46-.26l9.5-9.5c1.22-1.22 1.22-3.2.01-4.42zm-1.42 3l-9.3 9.3-2.11.53.53-2.11 9.3-9.3c.42-.42 1.16-.42 1.59 0 .43.43.43 1.15-.01 1.58z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        <span>Edit this page</span>
                                    </a>
                                </div>
                            
                                <nav class="flex mt-14">
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 h-full flex items-center break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-l-lg transition-colors duration-150 relative hover:z-5" href="../../../deployment/customization/customizing-your-instance-of-cbioportal/">
                                            <svg xmlns="http://www.w3.org/2000/svg" class="mr-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19 11H7.41l5.29-5.29a.996.996 0 10-1.41-1.41l-7 7a1 1 0 000 1.42l7 7a1.024 1.024 0 001.42-.01.996.996 0 000-1.41L7.41 13H19c.55 0 1-.45 1-1s-.45-1-1-1z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                            <span>
                                                <span class="block text-xs font-normal text-gray-400 dark:text-dark-400">Previous</span>
                                                <span class="block mt-1">Customizing your instance of cBioPortal</span>
                                            </span>
                                        </a>
                                    </div>
                            
                                    <div class="w-1/2">
                                        <a class="px-5 py-4 -mx-px h-full flex items-center justify-end break-normal font-medium text-blue-500 dark:text-blue-400 border border-gray-300 hover:border-gray-400 dark:border-dark-650 dark:hover:border-dark-450 rounded-r-lg transition-colors duration-150 relative hover:z-5" href="../../../deployment/customization/security.properties-reference/">
                                            <span>
                                                <span class="block text-xs font-normal text-right text-gray-400 dark:text-dark-400">Next</span>
                                                <span class="block mt-1">Application Properties</span>
                                            </span>
                                            <svg xmlns="http://www.w3.org/2000/svg" class="ml-3" width="24" height="24" viewBox="0 0 24 24" fill="currentColor" overflow="visible"><path d="M19.92 12.38a1 1 0 00-.22-1.09l-7-7a.996.996 0 10-1.41 1.41l5.3 5.3H5c-.55 0-1 .45-1 1s.45 1 1 1h11.59l-5.29 5.29a.996.996 0 000 1.41c.19.2.44.3.7.3s.51-.1.71-.29l7-7c.09-.09.16-.21.21-.33z" /><path fill="none" d="M0 0h24v24H0z" /></svg>
                                        </a>
                                    </div>
                                </nav>
                            </footer>
                        </main>
                
                        <div class="border-t dark:border-dark-650 pt-6 mb-8">
                            <footer class="flex flex-wrap items-center justify-between">
                                <div>
                                    <ul class="flex flex-wrap items-center text-sm">
                                    </ul>
                                </div>
                                <div class="docs-copyright py-2 text-gray-500 dark:text-dark-350 text-sm leading-relaxed"><p> Copyright 2024. All rights reserved.</p></div>
                            </footer>
                        </div>
                    </div>
                
                    <!-- Rendered if sidebar right is enabled -->
                    <!-- Sidebar right skeleton-->
                    <div v-cloak class="fixed top-0 bottom-0 right-0 translate-x-full bg-white border-gray-200 lg:sticky lg:border-l lg:shrink-0 lg:pt-6 lg:transform-none sm:w-1/2 lg:w-64 lg:z-0 md:w-104 sidebar-right skeleton dark:bg-dark-850 dark:border-dark-650">
                        <div class="pl-5">
                            <div class="w-32 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-48 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                            <div class="w-40 h-3 mb-4 bg-gray-200 dark:bg-dark-600 rounded-full loading"></div>
                        </div>
                    </div>
                
                    <!-- User should be able to hide sidebar right -->
                    <doc-sidebar-right v-cloak></doc-sidebar-right>
                </div>

            </div>
        </div>
    
        <doc-search-mobile></doc-search-mobile>
        <doc-back-to-top></doc-back-to-top>
    </div>


    <div id="docs-overlay-target"></div>

    <script data-cfasync="false">window.__DOCS__ = { "title": "More application.properties Settings", level: 3, icon: "file", hasPrism: true, hasMermaid: false, hasMath: true, tocDepth: 23 }</script>
</body>
</html>
