# November 9, 2015
* __New features__:
    * Links to MyCancerGenome.org for mutations
    * Improved display of selection samples on the study view page
* __Improvements__:
    * "Enrichments" analysis is now run across all genes
    * The "Network" tab is now using Cytoscape.js (Adobe Flash is no longer required)

# October 6, 2015
* __New TCGA data__:
    * [Breast Invasive Carcinoma (TCGA, Cell 2015)](study?id=brca_tcga_pub2015)
    * [Prostate Adenocarcinoma (TCGA, in press)](study?id=prad_tcga_pub)
    * [Uveal Melanoma (TCGA, Provisional)](study?id=uvm_tcga)
* __Added data__ of 763 samples from 12 published studies:
    * [Small Cell Lung Cancer (U Cologne, Nature 2015)](study?id=sclc_ucologne_2015)
    * [Uterine Carcinosarcoma (JHU, Nat Commun 2014)](study?id=ucs_jhu_2014)
    * [Microdissected Pancreatic Cancer Whole Exome Sequencing (UTSW, Nat Commun 2015)](study?id=paad_utsw_2015)
    * [Pancreatic Neuroendocrine Tumors (JHU, Science 2011)](study?id=panet_jhu_2011)
    * [Renal Non-Clear Cell Carcinoma (Genentech, Nat Genet 2014)](study?id=nccrcc_genentech_2014)
    * [Infant MLL-Rearranged Acute Lymphoblastic Leukemia (St Jude, Nat Genet 2015)](study?id=all_stjude_2015)
    * [Rhabdomyosarcoma (NIH, Cancer Discov 2014)](study?id=rms_nih_2014)
    * [Thymic epithelial tumors (NCI, Nat Genet 2014)](study?id=tet_nci_2014)
    * [Pediatric Ewing Sarcoma (DFCI, Cancer Discov 2014)](study?id=es_dfarber_broad_2014)
    * [Ewing Sarcoma (Institut Cuire, Cancer Discov 2014)](study?id=es_iocurie_2014)
    * [Cutaneous squamous cell carcinoma (DFCI, Clin Cancer Res 2015)](study?id=cscc_dfarber_2015)
    * [Gallbladder Carcinoma (Shanghai, Nat Genet 2014)](study?id=gbc_shanghai_2014)

# August 21, 2015
* All __TCGA data__ updated to the Firehose run of April 16, 2015.
* __New feature__: Enrichments Analysis finds alterations that are enriched in either altered or unaltered samples.
* __Improvement__: improved OncoPrint with better performance.

# June 3, 2015
* __Improvements__:
    * Allowed downloading data in each chart/table in study summary page.
    * Added log-rank test _p_-values to the survival plots in study summary page.
    * Improved visualization of patient clinical data in patient-centric view.
    * Added option to merge multiple samples for the same patient in OncoPrint.

# April 28, 2015
* __New features__:
    * Redesigned query interface to allow selecting multiple cancer studies
    * Redesigned __Plots__ tab

# January 20, 2015
* All __TCGA data__ updated to the Firehose run of October 17, 2014
* __COSMIC data__ updated to V71
* __New features__:
    * Query page: better search functions to find cancer studies
    * OncoPrints now support color coding of different mutation types
    * OncoPrints now support multiple clinical annotation tracks
    * [__OncoPrinter tool__](oncoprinter.jsp) now supports mRNA expression changes 
<br/>![Oncoprint with multiple clinical tracks](images/previews/multi-clinical-track-oncoprint.png)

# January 6, 2015
* __New feature__: You can now view __frequencies of mutations and copy-number alterations__ in the study view. These tables are updated dynamically when selecting subsets of samples.<br>
<br/>![Alterations in heavily copy-number altered endometrial cancer cases](images/previews/study_view_alt_frequencies.png)</center>

# December 9, 2014
* __New TCGA data__:
    * Added complete and up-to-date __clinical data__ for all __TCGA__ provisional studies
    * All TCGA data updated to the Firehose run of July 15, 2014
    * New TCGA provisional studies: Esophageal cancer, Pheochromocytoma and Paraganglioma (PCPG)
    * New published TCGA studies: [Thyroid Cancer](study?id=thca_tcga_pub) and [Kidney Chromophobe](study?id=kich_tcga_pub)
* __Added data__ of 172 samples from 4 published studies:
    * [Cholangiocarcinoma (National University of Singapore, Nature Genetics 2012)](study?id=chol_nus_2012)
    * [Cholangiocarcinoma (National Cancer Centre of Singapore, Nature Genetics 2013)](study?id=chol_nccs_2013)
    * [Intrahepatic Cholangiocarcinoma (Johns Hopkins University, Nature Genetics 2013)](study?id=chol_jhu_2013)
    * [Bladder Cancer (MSKCC, Eur Urol 2014)](study?id=blca_mskcc_solit_2014)
* __New features__:
    * Redesigned __Mutual Exclusivity__ tab
    * Added __correlation scores__ for scatter plots on the Plots tab
    * Download links to [__GenomeSpace__](http://www.genomespace.org/)


# October 24, 2014
* Added data of 885 samples from 11 published studies:
    * [Colorectal Adenocarcinoma Triplets (MSKCC, Genome Biology 2014)](study?id=coadread_mskcc)
    * [Esophageal Squamous Cell Carcinoma (ICGC, Nature 2014)](study?id=escc_icgc)
    * [Malignant Peripheral Nerve Sheath Tumor (MSKCC, Nature Genetics 2014)](study?id=mpnst_mskcc)
    * [Melanoma (Broad/Dana Farber, Nature 2012)](study?id=skcm_broad_dfarber)
    * [Nasopharyngeal Carcinoma (National University Singapore, Nature Genetics 2014)](study?id=npc_nusingapore)
    * [Prostate Adenocarcinoma CNA study (MSKCC, PNAS 2014)](index.do?cancer_study_id=prad_mskcc_2014)
    * [Prostate Adenocarcinoma Organoids (MSKCC, Cell 2014)](study?id=prad_mskcc_cheny1_organoids_2014)
    * [Stomach Adenocarcinoma (TCGA, Nature 2014)](study?id=stad_tcga_pub)
    * [Stomach Adenocarcinoma (Pfizer and University of Hong Kong, Nature Genetics 2014)](study?id=stad_pfizer_uhongkong)
    * [Stomach Adenocarcinoma (University of Hong Kong, Nature Genetics 2011)](study?id=stad_uhongkong)
    * [Stomach Adenocarcinoma (University of Tokyo, Nature Genetics 2014)](study?id=stad_utokyo)

# August 8, 2014
* Released two new tools
    * [Oncoprinter](oncoprinter.jsp) lets you create Oncoprints from your own, custom data
    * [MutationMapper](mutation_mapper.jsp) draws mutation diagrams (lollipop plots) from your custom data

# May 21, 2014
* All TCGA data updated to the Firehose run of April 16, 2014

# May 12, 2014
* Improved study summary page including survival analysis based on clinical attributes
<br/>e.g. [TCGA Endometrial Cancer cohort](study?id=ucec_tcga_pub)
<br/>![Study view](images/previews/study_view.png)

# March 27, 2014
* New features:
    * Visualizing of mutations mapped on 3D structures (individual or multiple mutations, directly in the browser)
    * Gene expression correlation analysis (find all genes with expression correlation to your query genes)
    * The Patient-Centric View now displays mutation frequencies across all cohorts in cBioPortal for each mutation
    * The Mutation Details Tab and the Patient-Centric View now display the copy-number status of each mutation
<br/>![3D viewer & Co-expression](images/previews/news_3d_coexp.png)

# March 18, 2014
* All TCGA data updated to the Firehose run of January 15, 2014
* Updated to the latest COSMIC data (v68)
* Added two new provisional TCGA studies:
    * Adrenocortical Carcinoma
    * Uterine Carcinosarcoma
* Added mutation data of 898 samples from 11 published studies:
    * Hepatocellular Carcinoma (RIKEN, Nature Genetics 2012)
    * Hepatocellular Carcinoma (AMC, Hepatology in press)
    * Medulloblastoma (Broad, Nature 2012)
    * Medulloblastoma (ICGC, Nature 2012)
    * Medulloblastoma (PCGP, Nature 2012)
    * Multiple Myeloma (Broad, Cancer Cell 2014)
    * Pancreatic Adenocarcinoma (ICGC, Nature 2012)
    * Small Cell Carcinoma of the Ovary (MSKCC, Nature Genetics in press)
    * Small Cell Lung Cancer (CLCGP, Nature Genetics 2012)
    * Small Cell Lung Cancer (Johns Hopkins, Nature Genetics 2012)
    * NCI-60 Cell Lines (NCI, Cancer Res. 2012)

# December 9, 2013
* Added mutation data of 99 bladder cancer samples (BGI, Nature Genetics 2013)

# December 6, 2013
* Data sets matching four recently submitted or published TCGA studies are now available
    * Glioblastoma (Cell 2013)
    * Bladder carcinoma (Nature, in press)
    * Head & neck squamous cell carcinoma (submitted)
    * Lung adenocarcinoma (submitted)

# November 8, 2013

* All TCGA data updated to the Firehose run of September 23, 2013.
* Updated to the latest COSMIC data (v67).
* Added mutation data of 792 samples from 9 published cancer studies:
    * Esophageal Adenocarcinoma (Broad, Nature Genetics 2013)
    * Head and Neck Squamous Cell Carcinoma (Broad, Science 2011)
    * Head and Neck Squamous Cell Carcinoma (Johns Hopkins, Science 2011)
    * Kidney Renal Clear Cell Carcinoma (BGI, Nature Genetics 2012)
    * Prostate Adenocarcinoma, Metastatic (Michigan, Nature 2012)
    * Prostate Adenocarcinoma (Broad/Cornell, Nature Genetics 2012)
    * Prostate Adenocarcinoma (Broad/Cornell, Cell 2013)
    * Skin Cutaneous Melanoma (Yale, Nature Genetics 2012)
    * Skin Cutaneous Melanoma (Broad, Cell 2012)

# October 21, 2013
 
* Improved interface for survival plots, including information on individual samples via mouse-over
* New fusion glyph in OncoPrints &nbsp;&nbsp;&nbsp;[![FGFR3 fusions in head and neck carcinoma](images/previews/fusion-in-oncoprint.png)](index.do?cancer_study_id=hnsc_tcga_pub&genetic_profile_ids_PROFILE_MUTATION_EXTENDED=hnsc_tcga_pub_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=hnsc_tcga_pub_gistic&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&case_set_id=hnsc_tcga_pub_sequenced&case_ids=&gene_set_choice=user-defined-list&gene_list=FGFR3%3A+AMP+MUT%3B%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A%0D%0A&clinical_param_selection=null&tab_index=tab_visualize&Action=Submit)
* Improved cross-cancer query: new alteration frequency histogram (example below - query gene: CDKN2A) and mutation diagram
<center>![Cross Cancer Query](images/previews/cross_cancer.png)</center>

# September 9, 2013

* Updated COSMIC data (v66 Release)
* Improved / interactive visualization on the "Protein changes" tab
* Enhanced mutation diagrams: color-coding by mutation time and syncing with table filters
* Addition of DNA cytoband information in the patient view of copy-number changes
* OncoPrints now allow the display of an optional track with clinical annotation (Endometrial cancer example below)
<center>![Oncoprint with clinical track](images/previews/oncoprint_clinical_track.png)</center>

# July 25, 2013

* Multi-gene correlation plots.
* Variant allele frequency distribution plots for individual tumor samples.
* Tissue images for TCGA samples in the patient view, via [Digital Slide Archive](http://cancer.digitalslidearchive.net/). [Example](case.do#/patient?studyId=ucec_tcga&caseId=TCGA-BK-A0CC&tab=tissueImageTab).

# July 16, 2013

* All TCGA data updated to the May Firehose run (May 23, 2013).
    * TCGA Pancreatic Cancer study (provisional) added.

# July 4, 2013

* Improved rendering of mutation diagrams, including ability to download in PDF format.
* Improved home page: Searchable cancer study & gene set selectors, data sets selector.

# June 17, 2013

* Improved interface for correlation plots, including information on individual samples via mouse-over.
* Gene Details from Biogene are now available in the Network view.
* Added mutation and copy number data from a new adenoid cystic carcinoma study: Ho et al., Nature Genetics 2013.
* Added mutation data from 6 cancer studies.
    * Breast Invasive Carcinoma (Shah et al., Nature 2012)
    * Breast Invasive Carcinoma (Banerji et al., Nature 2012)
    * Breast Invasive Carcinoma (Stephens et al., Nature 2012)
    * Lung Adenocarcinoma (Imielinksi et al., Cell 2012)
    * Lung Adenocarcinoma (Ding et al., Nature 2008)
    * Colorectal Cancer (Seshagiri et al., Nature 2012)

# June 4, 2013

* All TCGA data updated to the April Firehose run (April 21, 2012).

# May 14, 2013

* Added a published TCGA study: Acute Myeloid Leukemia (TCGA, NEJM 2013).

# April 28, 2013

* All TCGA data updated to the March Firehose run (March 26, 2012).
* mRNA percentiles for altered genes shown in patient view.

# April 2, 2013

* All TCGA data updated to the February Firehose run (February 22, 2012).

# March 28, 2013

* All TCGA data updated to the January Firehose run (January 16, 2012).
* Data from a new bladder cancer study from MSKCC has been added (97 samples, Iyer et al., JCO in press).

# February 16, 2013

* The cBio Portal now contains mutation data from all provisional TCGA projects. Please adhere to [the TCGA publication guidelines](http://cancergenome.nih.gov/abouttcga/policies/publicationguidelines) when using these and any TCGA data in your publications.
* All data updated to the October Firehose run (October 24, 2012).
* __Sequencing read counts and frequencies__ are now shown in the Mutation Details table when available.
* Improved OncoPrints, resulting in performance improvements.

# November 21, 2012

* Major new feature: Users can now visualize __genomic alterations and clinical data of individual tumors__, including:
    - Summary of __mutations__ and __copy-number alterations__ of interest
    - __Clinical trial__ information
    - TCGA __Pathology Reports__
* New __cancer summary view__ (Example [Endometrial Cancer](study?id=ucec_tcga))
* __Updated drug data__ from KEGG DRUG and NCI Cancer Drugs (aggregated by [PiHelper](https://bitbucket.org/armish/pihelper))

# October 22, 2012

* All data updated to the __Broad Firehose__ run from July 25, 2012.
* __COSMIC data__ added to Mutation Details (via Oncotator).
* All predicted functional impact scores are updated to __Mutation Assessor 2.0__.
* Users can now base queries on genes in recurrent regions of copy-number alteration (from __GISTIC__ via Firehose).
* The [Onco Query Language (OQL)](onco_query_lang_desc.jsp) now supports queries for specific mutations or mutation types.
* Data sets added that match the data of all TCGA publications (GBM, ovarian, colorectal, and lung squamous).

# July 18, 2012

* Mutation data for the TCGA lung squamous cell carcinoma and breast cancer projects (manuscripts in press at Nature).
* All data updated to the __latest Broad Firehose run__ (May 25, 2012).
* __Drug information__ added to the network view (via Drugbank).
* __Improved cross-cancer__ queries: Option to select data types, export of summary graphs.
* Users can now base queries on frequently mutated genes (from __MutSig__ via Firehose).

# May 16, 2012

* All data updated to the __latest Broad Firehose run__ (March 21, 2012).
* Extended __cross-cancer__ functionality, enabling users to query across all cancer studies in our database.
* New __"build a case"__ functionality, enabling users to generate custom case sets, based on one or more clinical attributes.
* New OncoPrint features, including more __compact OncoPrints__, and support for __RPPA__ visualization.

# February 27, 2012

* All data updated to the __latest Broad Firehose run__ (January 24, 2012).
* Validated mutation data for colorectal cancer.
* New feature: __Mutation Diagrams__ that show mutations in the context of protein domains.
<center>![TP53 Mutations in Ovarian Cancer](images/previews/tp53_mutations.png)</center>

# January 30, 2012

* Updated data for several TCGA cancer studies.
* Some small bug-fixes.

# December 22, 2011

* __Fourteen new [TCGA](http://cancergenome.nih.gov/) cancer studies__:  This includes complete data for TCGA Colorectal Carcinoma
and provisional data for thirteen other cancer types in the TCGA production pipeline.  Please note that data from these
thirteen new cancer types are provisional, not final and do not yet include mutation data.
As per NCI guidelines, preliminary mutation data cannot be redistributed until they have been validated.  

<center>![TCGA](http://cancergenome.nih.gov/PublishedContent/Images/SharedItems/Images/tcga_logo.png)</center>

* __Four new data types__:
    * Reverse-phase protein array (RPPA) data.
    * microRNA expression and copy-number (including support for multiple loci)
    * RNA-Seq based expression data.
    * log2 copy-number data.
* Updated TCGA GBM copy-number, expression, and methylation data.
* New __gene symbol validation__ service.  You can now use gene aliases and/or Entrez Gene IDs within your gene sets.
* __Links to IGV__ for visualization of DNA copy-number changes.
* Background information from the [Sanger Cancer Gene Census](http://www.sanger.ac.uk/genetics/CGP/Census/).
* Two __new [Tutorials](tutorial.jsp)__ to get you quickly started in using the portal.
<BR/>
<BR/>

# November 14, 2011

* New and __improved mutation details__, with sorting and filtering capabilities.
* In collaboration with Bilkent University, we have added a __new Network tab__ to our results pages.  The network tab enables users to visualize, analyze and filter cancer genomic data in the context of pathways and interaction networks derived from [Pathway Commons](http://www.pathwaycommons.org).

<center>![GBM Network](images/previews/ova_network.png)</center>
<BR/>
<BR/>

# September 3, 2011

* You can now query across different cancer studies (feature available directly from the home page).
* Our [MATLAB CGDS Cancer Genomics Toolbox](cgds_r.jsp) is now available.  The toolbox enables you to download data from the cBio Portal, and import it directly into MATLAB.
* The code for the cBio Portal has now been fully open sourced, and made available at [Google Code](http://code.google.com/p/cbio-cancer-genomics-portal/).  If you would like to join our open source efforts and make the portal even better, drop us an email.
<BR/>
<BR/>

# March 2, 2011

New plotting features and other improvements:

* Correlation plots that show the relationship between different data types for individual genes.
* Survival analysis - assess survival differences between altered and non-altered patient sets.
* Updated [R Package](cgds_r.jsp) with support for correlation plots and general improvements for retrieving and accessing data in R data frames.
* The [Web Interface](web_api.jsp) now supports basic clinical data, e.g. survival data.
* [Networks](networks.jsp) for pathway analysis are now available for download.

<center>![Survival Analysis](images/previews/ova_survival_60_percent.png)</center>
<BR/>
<BR/>

# December 15, 2010

Several new features, including:

* Redesigned and streamlined user interface, based on user feedback and usability testing.
* Advanced support for gene-specific alterations.  For example, users
can now view mutations within TP53, and ignore copy number alterations, or
only view amplifications of EGFR, and ignore deletions.
* Improved performance.
* [Frequently Asked Questions](faq.jsp) document released.
* Updated [Video Tutorial](video.jsp).
<BR/>
<BR/>

# November 4, 2010

* Enhanced [Oncoprints](faq.jsp#what-are-oncoprints), enabling users to quickly visualize genomic alterations across many cases.  Oncoprints now also work in all major browsers, including Firefox, Chrome, Safari, and Internet Explorer.
* Official release of our [Web Interface](web_api.jsp), enabling programmatic access to all data.
* Official release of our [R Package](cgds_r.jsp), enabling programmatic access to all data from the R platform for statistical computing.

<center>![OncoPrints](images/previews/gbm_oncoprint.png)</center>
